{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "PROJECT_DIR=`git rev-parse --show-toplevel`\n",
    "cd $PROJECT_DIR/single-kmer-match\n",
    "\n",
    "threads=1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "These commands test out fishing for a specific AMR gene kmer in reads, assembling them, and then trying to pull out more kmers to extend the assembled contig until we get an AMR gene match."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Construct k-mer probe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">gb|FJ196385|-|8531-9323|ARO:3000413|sul3 [Escherichia coli] \n",
      "ATGAGCAAGATTTTTGGAATCGTAAATATAACCACCGATAGTTTTTCCGATGGAGGACTTTATTTAGATACAGATAAGGCAATTGAGCATGCTCTGCATTTGGTTGAAGATGGAGCAGATGTGATTGATTTGGGAGCCGCTTCCAGTAATCCTGATACAACTGAAGTGGGCGTTGTGGAAGAAATCAAAAGACTCAAACCTGTCATTAAGGCTTTAAAAGAAAAAGGCATTTCTATTTCTGTTGATACATTTAAACCTGAGGTTCAGAGTTTTTGCATAGAACAAAAGGTTGATTTTATTAATGATATTCAAGGTTTTCCTTATCCTGAGATTTATTCAGGCTTGGCAAAGTCAGATTGCAAACTTGTGTTGATGCACTCCGTTCAGCGAATTGGTGCAGCTACTAAAGTTGAAACGAATCCGGAAGAGGTTTTTACTTCCATGATGGAATTTTTTAAAGAAAGAATTGCTGCTTTAGTTGAGGCTGGTGTAAAGCGTGAACGAATTATTCTTGATCCGGGTATGGGCTTCTTTTTAGGCTCTAATCCAGAAACATCTATTCTTGTTTTGAAGCGTTTCCCTGAAATTCAAGAAGCTTTTAATTTGCAAGTAATGATTGCAGTGTCACGGAAATCATTCTTAGGTAAAATAACTGGAACCGATGTGAAATCTCGTTTAGCACCAACTCTTGCAGCAGAAATGTATGCATACAAAAAAGGTGCAGATTATCTCCGCACCCATGATGTTAAGTCTTTATCAGATGCCTTGAAAATATCCAAAGCCCTAGGTTAG\n",
      "kmer-probe/sul3.fasta 792 bp 1 sequences\n"
     ]
    }
   ],
   "source": [
    "grep 'sul3' kmer-probe/nucleotide_fasta_protein_homolog_model.fasta -A 1 | tee kmer-probe/sul3.fasta\n",
    "bp_seq_length kmer-probe/sul3.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Pick a single kmer from the middle of this gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "31\n",
      ">gb|FJ196385|-|8531-9323|ARO:3000413|sul3 [Escherichia coli] \n",
      "ATGAGCAAGATTTTTGGAATCGTAAATATAACCACCGATAGTTTTTCCGATGGAGGACTTTATTTAGATACAGATAAGGCAATTGAGCATGCTCTGCATTTGGTTGAAGATGGAGCAGATGTGATTGATTTGGGAGCCGCTTCCAGTAATCCTGATACAACTGAAGTGGGCGTTGTGGAAGAAATCAAAAGACTCAAACCTGTCATTAAGGCTTTAAAAGAAAAAGGCATTTCTATTTCTGTTGATACATTTAAACCTGAGGTTCAGAGTTTTTGCATAGAACAAAAGGTTGATTTTATTAATGATATTCAAGGTTTTCCTTATCCTGAGATTTATTCAGGCTTGGCAAAGTCAGATTGCAAACT [TGTGTTGATGCACTCCGTTCAGCGAATTGGT] GCAGCTACTAAAGTTGAAACGAATCCGGAAGAGGTTTTTACTTCCATGATGGAATTTTTTAAAGAAAGAATTGCTGCTTTAGTTGAGGCTGGTGTAAAGCGTGAACGAATTATTCTTGATCCGGGTATGGGCTTCTTTTTAGGCTCTAATCCAGAAACATCTATTCTTGTTTTGAAGCGTTTCCCTGAAATTCAAGAAGCTTTTAATTTGCAAGTAATGATTGCAGTGTCACGGAAATCATTCTTAGGTAAAATAACTGGAACCGATGTGAAATCTCGTTTAGCACCAACTCTTGCAGCAGAAATGTATGCATACAAAAAAGGTGCAGATTATCTCCGCACCCATGATGTTAAGTCTTTATCAGATGCCTTGAAAATATCCAAAGCCCTAGGTTAG\n"
     ]
    }
   ],
   "source": [
    "kmer='TGTGTTGATGCACTCCGTTCAGCGAATTGGT'\n",
    "kmer_length=31\n",
    "echo -n ${kmer} | wc -c\n",
    "sed -e \"s/\\(${kmer}\\)/ [\\1] /\" kmer-probe/sul3.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">single-kmer\n",
      "TGTGTTGATGCACTCCGTTCAGCGAATTGGT\n"
     ]
    }
   ],
   "source": [
    "(echo -e \">single-kmer\\n${kmer}\";) | tee kmer-probe/sul3.single.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Make blast database from this single gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Building a new DB, current time: 01/17/2020 11:14:27\n",
      "New DB name:   /home/CSCScience.ca/apetkau/workspace/kmer-analysis-test/single-kmer-match/kmer-probe/sul3.fasta\n",
      "New DB title:  kmer-probe/sul3.fasta\n",
      "Sequence type: Nucleotide\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 1 sequences in 0.000257969 seconds.\n"
     ]
    }
   ],
   "source": [
    "conda run --name blast makeblastdb -dbtype nucl -in kmer-probe/sul3.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Iteration 1\n",
    "\n",
    "## 1.1 Pull out reads matching kmer(s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Kmer Analysis Toolkit (KAT) V2.4.2\n",
      "\n",
      "Running KAT in filter sequence mode\n",
      "-----------------------------------\n",
      "\n",
      "Input 1216 is a sequence file.  Counting kmers for input 1216 (kmer-probe/sul3.single.fasta) ... done.  Time taken: 0.1s\n",
      "\n",
      "Filtering sequences ...\n",
      "Processed 100000 pairs\n",
      "Processed 200000 pairs\n",
      "Processed 300000 pairs\n",
      "Processed 400000 pairs\n",
      "Processed 500000 pairs\n",
      "Processed 600000 pairs\n",
      "Processed 700000 pairs\n",
      "Processed 800000 pairs\n",
      "Processed 900000 pairs\n",
      "Processed 1000000 pairs\n",
      "Processed 1100000 pairs\n",
      "Processed 1200000 pairs\n",
      "Processed 1300000 pairs\n",
      "Processed 1400000 pairs\n",
      "Processed 1500000 pairs\n",
      "Processed 1600000 pairs\n",
      "Processed 1700000 pairs\n",
      "Processed 1800000 pairs\n",
      "Processed 1900000 pairs\n",
      "Processed 2000000 pairs\n",
      "Processed 2100000 pairs\n",
      "Finished filtering.  Time taken: 108.6s\n",
      "\n",
      "Found 77 / 2147046 to keep\n",
      "\n",
      "KAT filter seq completed.\n",
      "Total runtime: 108.7s\n",
      "\n",
      "\tCommand being timed: \"kat filter seq -t 1 -o SRR1952908.kat.1 -m 31 -T 0.0000001 --seq ../input/SRR1952908_1.fastq --seq2 ../input/SRR1952908_2.fastq kmer-probe/sul3.single.fasta\"\n",
      "\tUser time (seconds): 107.60\n",
      "\tSystem time (seconds): 1.57\n",
      "\tPercent of CPU this job got: 100%\n",
      "\tElapsed (wall clock) time (h:mm:ss or m:ss): 1:48.79\n",
      "\tAverage shared text size (kbytes): 0\n",
      "\tAverage unshared data size (kbytes): 0\n",
      "\tAverage stack size (kbytes): 0\n",
      "\tAverage total size (kbytes): 0\n",
      "\tMaximum resident set size (kbytes): 843712\n",
      "\tAverage resident set size (kbytes): 0\n",
      "\tMajor (requiring I/O) page faults: 0\n",
      "\tMinor (reclaiming a frame) page faults: 314862\n",
      "\tVoluntary context switches: 15\n",
      "\tInvoluntary context switches: 45\n",
      "\tSwaps: 0\n",
      "\tFile system inputs: 0\n",
      "\tFile system outputs: 96\n",
      "\tSocket messages sent: 0\n",
      "\tSocket messages received: 0\n",
      "\tSignals delivered: 0\n",
      "\tPage size (bytes): 4096\n",
      "\tExit status: 0\n"
     ]
    }
   ],
   "source": [
    "conda run --name kat /usr/bin/time -v kat filter seq -t ${threads} -o SRR1952908.kat.1 -m ${kmer_length} -T 0.0000001 \\\n",
    "    --seq ../input/SRR1952908_1.fastq --seq2 ../input/SRR1952908_2.fastq \\\n",
    "    kmer-probe/sul3.single.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20K\tSRR1952908.kat.1.in.R1.fastq\n",
      "20K\tSRR1952908.kat.1.in.R2.fastq\n",
      "40K\ttotal\n"
     ]
    }
   ],
   "source": [
    "du -ch *kat.1.*.fastq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.2 Assemble reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "skesa --cores 1 --fastq SRR1952908.kat.1.in.R1.fastq,SRR1952908.kat.1.in.R2.fastq --contigs_out SRR1952908.kat.1.contigs.fasta --vector_percent 1 \n",
      "\n",
      "Total mates: 154 Paired reads: 77\n",
      "Reads acquired in  0.000441s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Adapters clip is disabled\n",
      "\n",
      "Kmer len: 21\n",
      "Raw kmers: 12239 Memory needed (GB): 0.000234988 Memory available (GB): 30 1 cycle(s) will be performed\n",
      "Distinct kmers: 689\n",
      "Kmer count in  0.006960s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Uniq kmers merging in  0.000921s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (n/a%)\n",
      "Kmers branching in  0.001992s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "\n",
      "Average read length: 99\n",
      "Genome size estimate: 289\n",
      "\n",
      "Kmer: 21 Graph size: 689 Contigs in: 0\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.000004s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmers in multiple/single contigs: 0 0\n",
      "Fragments before: 1 667\n",
      "Fragments after: 1 667\n",
      "New seeds: 1\n",
      "New seeds in  0.001451s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Fragments before: 2 42\n",
      "Fragments after: 2 42\n",
      "Connectors: 0 Extenders: 2\n",
      "Connections and extensions in  0.000951s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Contigs out: 1 Genome: 625 N50: 625 L50: 1\n",
      "Assembled in  0.002457s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "\n",
      "\n",
      "Average count: 4.88235 Max kmer: 21\n",
      "Connected: 73 ambiguously connected: 0 from 77 mate pairs\n",
      "Connect pairs in  0.045289s wall, 0.040000s user + 0.000000s system = 0.040000s CPU (88.3%)\n",
      "\n",
      "N50 for inserts: 253\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 58\n",
      "Reads for connection: 34\n",
      "Internal reads: 7\n",
      "Reads cleaned in  0.004065s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "WARNING: iterations are disabled\n",
      "\n",
      "Connecting mate pairs using kmer length: 21\n",
      "Connected: 14 ambiguously connected: 0 from 17 mate pairs\n",
      "Connect pairs in  0.008247s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (121.3%)\n",
      "Totally connected: 14\n",
      "Added notconnected: 0\n",
      "\n",
      "Kmer len: 25\n",
      "Raw kmers: 7152 Memory needed (GB): 0.000137318 Memory available (GB): 30 1 cycle(s) will be performed\n",
      "Distinct kmers: 663\n",
      "Kmer count in  0.004543s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (220.1%)\n",
      "Uniq kmers merging in  0.000832s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmers branching in  0.002037s wall, 0.000000s user + 0.010000s system = 0.010000s CPU (491.0%)\n",
      "Kmer: 25 Graph size: 663 Contigs in: 1\n",
      "Valley: 5\n",
      "\n",
      "Mark used kmers in  0.000499s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmers in multiple/single contigs: 0 601\n",
      "Fragments before: 1 31\n",
      "Fragments after: 1 31\n",
      "New seeds: 0\n",
      "New seeds in  0.000495s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Fragments before: 2 62\n",
      "Fragments after: 2 62\n",
      "Connectors: 0 Extenders: 2\n",
      "Connections and extensions in  0.001116s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Contigs out: 1 Genome: 631 N50: 631 L50: 1\n",
      "Assembled in  0.002221s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "\n",
      "\n",
      "Kmer len: 139\n",
      "Raw kmers: 4758 Memory needed (GB): 0.00027406 Memory available (GB): 30 1 cycle(s) will be performed\n",
      "Distinct kmers: 549\n",
      "Kmer count in  0.005239s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Uniq kmers merging in  0.001060s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmers branching in  0.002486s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmer: 139 Graph size: 549 Contigs in: 1\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.000587s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmers in multiple/single contigs: 0 493\n",
      "Fragments before: 2 56\n",
      "Fragments after: 2 56\n",
      "New seeds: 0\n",
      "New seeds in  0.000637s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Fragments before: 2 56\n",
      "Fragments after: 2 56\n",
      "Connectors: 0 Extenders: 2\n",
      "Connections and extensions in  0.001487s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Contigs out: 1 Genome: 631 N50: 631 L50: 1\n",
      "Assembled in  0.002817s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "\n",
      "\n",
      "Kmer len: 253\n",
      "Raw kmers: 2430 Memory needed (GB): 0.000209952 Memory available (GB): 29.9999 1 cycle(s) will be performed\n",
      "Distinct kmers: 353\n",
      "Kmer count in  0.003868s wall, 0.000000s user + 0.010000s system = 0.010000s CPU (258.5%)\n",
      "Uniq kmers merging in  0.000958s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (n/a%)\n",
      "Kmers branching in  0.002148s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmer: 253 Graph size: 353 Contigs in: 1\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.000508s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmers in multiple/single contigs: 0 297\n",
      "Fragments before: 2 56\n",
      "Fragments after: 2 56\n",
      "New seeds: 0\n",
      "New seeds in  0.000748s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Fragments before: 2 56\n",
      "Fragments after: 2 56\n",
      "Connectors: 0 Extenders: 2\n",
      "Connections and extensions in  0.002007s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Contigs out: 1 Genome: 631 N50: 631 L50: 1\n",
      "Assembled in  0.003368s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "\n",
      "DONE\n",
      "\tCommand being timed: \"skesa --cores 1 --fastq SRR1952908.kat.1.in.R1.fastq,SRR1952908.kat.1.in.R2.fastq --contigs_out SRR1952908.kat.1.contigs.fasta --vector_percent 1\"\n",
      "\tUser time (seconds): 0.08\n",
      "\tSystem time (seconds): 0.02\n",
      "\tPercent of CPU this job got: 101%\n",
      "\tElapsed (wall clock) time (h:mm:ss or m:ss): 0:00.10\n",
      "\tAverage shared text size (kbytes): 0\n",
      "\tAverage unshared data size (kbytes): 0\n",
      "\tAverage stack size (kbytes): 0\n",
      "\tAverage total size (kbytes): 0\n",
      "\tMaximum resident set size (kbytes): 5992\n",
      "\tAverage resident set size (kbytes): 0\n",
      "\tMajor (requiring I/O) page faults: 0\n",
      "\tMinor (reclaiming a frame) page faults: 388\n",
      "\tVoluntary context switches: 248\n",
      "\tInvoluntary context switches: 1\n",
      "\tSwaps: 0\n",
      "\tFile system inputs: 0\n",
      "\tFile system outputs: 8\n",
      "\tSocket messages sent: 0\n",
      "\tSocket messages received: 0\n",
      "\tSignals delivered: 0\n",
      "\tPage size (bytes): 4096\n",
      "\tExit status: 0\n"
     ]
    }
   ],
   "source": [
    "conda run --name skesa /usr/bin/time -v skesa --cores 1 \\\n",
    "    --fastq SRR1952908.kat.1.in.R1.fastq,SRR1952908.kat.1.in.R2.fastq \\\n",
    "    --contigs_out SRR1952908.kat.1.contigs.fasta --vector_percent 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.3 Examine assembled sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">Contig_1_19.1064\n",
      "GGAGATAATCTGCACCTTTTTTGTATGCATACATTTCTGCTGCAAGAGTTGGTGCTAAACGAGATTTCACATCGGTTCCAGTTATTTTACCTAAGAATGATTTCCGTGACACTGCAATCATTACTTGCAAATTAAAAGCTTCTTGAATTTCAGGGAAACGCTTCAAAACAAGAATAGATGTTTCTGGATTAGAGCCTAAAAAGAAGCCCATACCCGGATCAAGAATAATTCGTTCACGCTTTACACCAGCCTCAACTAAAGCAGCAATTCTTTCTTTAAAAAATTCCATCATGGAAGTAAAAACCTCTTCCGGATTCGTTTCAACTTTAGTAGCTGCACCAATTCGCTGAACGGAGTGCATCAACACAAGTTTGCAATCTGACTTTGCCAAGCCTGAATAAATCTCAGGATAAGGAAAACCTTGAATATCATTAATAAAATCAACCTTTTGTTCTATGCAAAAACTCTGAACCTCAGGTTTAAATGTATCAACAGAAATAGAAATGCCTTTTTCTTTTAAAGCCTTAATGACAGGTTTGAGTCTTTTGATTTCTTCCACAACGCCCACTTCAGTTGTATCAGGATTACTGGAAGCGGCTCCCAAATCAATCACATCTGCTCCATCTTCAAC\n",
      "SRR1952908.kat.1.contigs.fasta 631 bp 1 sequences\n"
     ]
    }
   ],
   "source": [
    "cat SRR1952908.kat.1.contigs.fasta\n",
    "bp_seq_length SRR1952908.kat.1.contigs.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BLASTN 2.9.0+\n",
      "\n",
      "\n",
      "Reference: Zheng Zhang, Scott Schwartz, Lukas Wagner, and Webb\n",
      "Miller (2000), \"A greedy algorithm for aligning DNA sequences\", J\n",
      "Comput Biol 2000; 7(1-2):203-14.\n",
      "\n",
      "\n",
      "\n",
      "Database: kmer-probe/sul3.fasta\n",
      "           1 sequences; 792 total letters\n",
      "\n",
      "\n",
      "\n",
      "Query= Contig_1_19.1064\n",
      "\n",
      "Length=631\n",
      "                                                                      Score        E\n",
      "Sequences producing significant alignments:                          (Bits)     Value\n",
      "\n",
      "gb|FJ196385|-|8531-9323|ARO:3000413|sul3 [Escherichia coli]           1166       0.0  \n",
      "\n",
      "\n",
      ">gb|FJ196385|-|8531-9323|ARO:3000413|sul3 [Escherichia coli] \n",
      "Length=792\n",
      "\n",
      " Score = 1166 bits (631),  Expect = 0.0\n",
      " Identities = 631/631 (100%), Gaps = 0/631 (0%)\n",
      " Strand=Plus/Minus\n",
      "\n",
      "Query  1    GGAGATAATCTGCACCTTTTTTGTATGCATACATTTCTGCTGCAAGAGTTGGTGCTAAAC  60\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  733  GGAGATAATCTGCACCTTTTTTGTATGCATACATTTCTGCTGCAAGAGTTGGTGCTAAAC  674\n",
      "\n",
      "Query  61   GAGATTTCACATCGGTTCCAGTTATTTTACCTAAGAATGATTTCCGTGACACTGCAATCA  120\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  673  GAGATTTCACATCGGTTCCAGTTATTTTACCTAAGAATGATTTCCGTGACACTGCAATCA  614\n",
      "\n",
      "Query  121  TTACTTGCAAATTAAAAGCTTCTTGAATTTCAGGGAAACGCTTCAAAACAAGAATAGATG  180\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  613  TTACTTGCAAATTAAAAGCTTCTTGAATTTCAGGGAAACGCTTCAAAACAAGAATAGATG  554\n",
      "\n",
      "Query  181  TTTCTGGATTAGAGCCTAAAAAGAAGCCCATACCCGGATCAAGAATAATTCGTTCACGCT  240\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  553  TTTCTGGATTAGAGCCTAAAAAGAAGCCCATACCCGGATCAAGAATAATTCGTTCACGCT  494\n",
      "\n",
      "Query  241  TTACACCAGCCTCAACTAAAGCAGCAATTCTTTCTTTAAAAAATTCCATCATGGAAGTAA  300\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  493  TTACACCAGCCTCAACTAAAGCAGCAATTCTTTCTTTAAAAAATTCCATCATGGAAGTAA  434\n",
      "\n",
      "Query  301  AAACCTCTTCCGGATTCGTTTCAACTTTAGTAGCTGCACCAATTCGCTGAACGGAGTGCA  360\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  433  AAACCTCTTCCGGATTCGTTTCAACTTTAGTAGCTGCACCAATTCGCTGAACGGAGTGCA  374\n",
      "\n",
      "Query  361  TCAACACAAGTTTGCAATCTGACTTTGCCAAGCCTGAATAAATCTCAGGATAAGGAAAAC  420\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  373  TCAACACAAGTTTGCAATCTGACTTTGCCAAGCCTGAATAAATCTCAGGATAAGGAAAAC  314\n",
      "\n",
      "Query  421  CTTGAATATCATTAATAAAATCAACCTTTTGTTCTATGCAAAAACTCTGAACCTCAGGTT  480\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  313  CTTGAATATCATTAATAAAATCAACCTTTTGTTCTATGCAAAAACTCTGAACCTCAGGTT  254\n",
      "\n",
      "Query  481  TAAATGTATCAACAGAAATAGAAATGCCTTTTTCTTTTAAAGCCTTAATGACAGGTTTGA  540\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  253  TAAATGTATCAACAGAAATAGAAATGCCTTTTTCTTTTAAAGCCTTAATGACAGGTTTGA  194\n",
      "\n",
      "Query  541  GTCTTTTGATTTCTTCCACAACGCCCACTTCAGTTGTATCAGGATTACTGGAAGCGGCTC  600\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  193  GTCTTTTGATTTCTTCCACAACGCCCACTTCAGTTGTATCAGGATTACTGGAAGCGGCTC  134\n",
      "\n",
      "Query  601  CCAAATCAATCACATCTGCTCCATCTTCAAC  631\n",
      "            |||||||||||||||||||||||||||||||\n",
      "Sbjct  133  CCAAATCAATCACATCTGCTCCATCTTCAAC  103\n",
      "\n",
      "\n",
      "\n",
      "Lambda      K        H\n",
      "    1.33    0.621     1.12 \n",
      "\n",
      "Gapped\n",
      "Lambda      K        H\n",
      "    1.28    0.460    0.850 \n",
      "\n",
      "Effective search space used: 482820\n",
      "\n",
      "\n",
      "  Database: kmer-probe/sul3.fasta\n",
      "    Posted date:  Jan 17, 2020  11:14 AM\n",
      "  Number of letters in database: 792\n",
      "  Number of sequences in database:  1\n",
      "\n",
      "\n",
      "\n",
      "Matrix: blastn matrix 1 -2\n",
      "Gap Penalties: Existence: 0, Extension: 2.5\n",
      "\tCommand being timed: \"blastn -db kmer-probe/sul3.fasta -query SRR1952908.kat.1.contigs.fasta\"\n",
      "\tUser time (seconds): 0.04\n",
      "\tSystem time (seconds): 0.01\n",
      "\tPercent of CPU this job got: 88%\n",
      "\tElapsed (wall clock) time (h:mm:ss or m:ss): 0:00.07\n",
      "\tAverage shared text size (kbytes): 0\n",
      "\tAverage unshared data size (kbytes): 0\n",
      "\tAverage stack size (kbytes): 0\n",
      "\tAverage total size (kbytes): 0\n",
      "\tMaximum resident set size (kbytes): 36984\n",
      "\tAverage resident set size (kbytes): 0\n",
      "\tMajor (requiring I/O) page faults: 17\n",
      "\tMinor (reclaiming a frame) page faults: 1950\n",
      "\tVoluntary context switches: 19\n",
      "\tInvoluntary context switches: 3\n",
      "\tSwaps: 0\n",
      "\tFile system inputs: 3536\n",
      "\tFile system outputs: 0\n",
      "\tSocket messages sent: 0\n",
      "\tSocket messages received: 0\n",
      "\tSignals delivered: 0\n",
      "\tPage size (bytes): 4096\n",
      "\tExit status: 0\n"
     ]
    }
   ],
   "source": [
    "conda run --name blast /usr/bin/time -v blastn -db kmer-probe/sul3.fasta -query SRR1952908.kat.1.contigs.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We matched 631 bp at 100%, but only matched 631/792 bp. Let's try another iteration.\n",
    "\n",
    "# Iteration 2\n",
    "\n",
    "## 2.0 Get kmers from ends of contigs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">Contig_1_19.1064\n",
      "GGAGATAATCTGCACCTTTTTTGTATGCATACATTTCTGCTGCAAGAGTTGGTGCTAAACGAGATTTCACATCGGTTCCAGTTATTTTACCTAAGAATGATTTCCGTGACACTGCAATCATTACTTGCAAATTAAAAGCTTCTTGAATTTCAGGGAAACGCTTCAAAACAAGAATAGATGTTTCTGGATTAGAGCCTAAAAAGAAGCCCATACCCGGATCAAGAATAATTCGTTCACGCTTTACACCAGCCTCAACTAAAGCAGCAATTCTTTCTTTAAAAAATTCCATCATGGAAGTAAAAACCTCTTCCGGATTCGTTTCAACTTTAGTAGCTGCACCAATTCGCTGAACGGAGTGCATCAACACAAGTTTGCAATCTGACTTTGCCAAGCCTGAATAAATCTCAGGATAAGGAAAACCTTGAATATCATTAATAAAATCAACCTTTTGTTCTATGCAAAAACTCTGAACCTCAGGTTTAAATGTATCAACAGAAATAGAAATGCCTTTTTCTTTTAAAGCCTTAATGACAGGTTTGAGTCTTTTGATTTCTTCCACAACGCCCACTTCAGTTGTATCAGGATTACTGGAAGCGGCTCCCAAATCAATCACATCTGCTCCATCTTCAAC\n"
     ]
    }
   ],
   "source": [
    "cat SRR1952908.kat.1.contigs.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TGTGTTGATGCACTCCGTTCAGCGAATTGGT\n",
      "31\n",
      "31\n",
      "31\n",
      ">Contig_1_19.1064\n",
      " [GGAGATAATCTGCACCTTTTTTGTATGCATA] CATTTCTGCTGCAAGAGTTGGTGCTAAACGAGATTTCACATCGGTTCCAGTTATTTTACCTAAGAATGATTTCCGTGACACTGCAATCATTACTTGCAAATTAAAAGCTTCTTGAATTTCAGGGAAACGCTTCAAAACAAGAATAGATGTTTCTGGATTAGAGCCTAAAAAGAAGCCCATACCCGGATCAAGAATAATTCGTTCACGCTTTACACCAGCCTCAACTAAAGCAGCAATTCTTTCTTTAAAAAATTCCATCATGGAAGTAAAAACCTCTTCCGGATTCGTTTCAACTTTAGTAGCTGCACCAATTCGCTGAACGGAGTGCATCAACACAAGTTTGCAATCTGACTTTGCCAAGCCTGAATAAATCTCAGGATAAGGAAAACCTTGAATATCATTAATAAAATCAACCTTTTGTTCTATGCAAAAACTCTGAACCTCAGGTTTAAATGTATCAACAGAAATAGAAATGCCTTTTTCTTTTAAAGCCTTAATGACAGGTTTGAGTCTTTTGATTTCTTCCACAACGCCCACTTCAGTTGTATCAGGATTACTGGAAGCGGCTC [CCAAATCAATCACATCTGCTCCATCTTCAAC] \n",
      ">kmer_orig\n",
      "TGTGTTGATGCACTCCGTTCAGCGAATTGGT\n",
      ">kmer1\n",
      "GGAGATAATCTGCACCTTTTTTGTATGCATA\n",
      ">kmer2\n",
      "CCAAATCAATCACATCTGCTCCATCTTCAAC\n"
     ]
    }
   ],
   "source": [
    "kmer_orig=${kmer}\n",
    "echo ${kmer_orig}\n",
    "echo -n ${kmer_orig} | wc -c\n",
    "kmer1='GGAGATAATCTGCACCTTTTTTGTATGCATA'\n",
    "echo -n ${kmer1} | wc -c\n",
    "kmer2='CCAAATCAATCACATCTGCTCCATCTTCAAC'\n",
    "echo -n ${kmer2} | wc -c\n",
    "sed -re \"s/(${kmer_orig}|${kmer1}|${kmer2})/ [\\1] /g\" SRR1952908.kat.1.contigs.fasta\n",
    "(echo -e \">kmer_orig\\n${kmer_orig}\\n>kmer1\\n${kmer1}\\n>kmer2\\n${kmer2}\";) | tee kmer-probe/sul3.2.kmer.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "kmer_orig isn't here in the middle (likely because we need to search for the reverse complement). Let's do this now."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">Contig_1_19.1064\n",
      "GGAGATAATCTGCACCTTTTTTGTATGCATACATTTCTGCTGCAAGAGTTGGTGCTAAACGAGATTTCACATCGGTTCCAGTTATTTTACCTAAGAATGATTTCCGTGACACTGCAATCATTACTTGCAAATTAAAAGCTTCTTGAATTTCAGGGAAACGCTTCAAAACAAGAATAGATGTTTCTGGATTAGAGCCTAAAAAGAAGCCCATACCCGGATCAAGAATAATTCGTTCACGCTTTACACCAGCCTCAACTAAAGCAGCAATTCTTTCTTTAAAAAATTCCATCATGGAAGTAAAAACCTCTTCCGGATTCGTTTCAACTTTAGTAGCTGC [ACCAATTCGCTGAACGGAGTGCATCAACACA] AGTTTGCAATCTGACTTTGCCAAGCCTGAATAAATCTCAGGATAAGGAAAACCTTGAATATCATTAATAAAATCAACCTTTTGTTCTATGCAAAAACTCTGAACCTCAGGTTTAAATGTATCAACAGAAATAGAAATGCCTTTTTCTTTTAAAGCCTTAATGACAGGTTTGAGTCTTTTGATTTCTTCCACAACGCCCACTTCAGTTGTATCAGGATTACTGGAAGCGGCTCCCAAATCAATCACATCTGCTCCATCTTCAAC\n"
     ]
    }
   ],
   "source": [
    "kmer_orig_rc='ACCAATTCGCTGAACGGAGTGCATCAACACA'\n",
    "sed -re \"s/(${kmer_orig_rc})/ [\\1] /g\" SRR1952908.kat.1.contigs.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There it is. So now let's pull out reads matching these 3 kmers (and their reverse complements)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.1 Pull out reads matching kmer(s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Kmer Analysis Toolkit (KAT) V2.4.2\n",
      "\n",
      "Running KAT in filter sequence mode\n",
      "-----------------------------------\n",
      "\n",
      "Input 38080 is a sequence file.  Counting kmers for input 38080 (kmer-probe/sul3.2.kmer.fasta) ... done.  Time taken: 0.1s\n",
      "\n",
      "Filtering sequences ...\n",
      "Processed 100000 pairs\n",
      "Processed 200000 pairs\n",
      "Processed 300000 pairs\n",
      "Processed 400000 pairs\n",
      "Processed 500000 pairs\n",
      "Processed 600000 pairs\n",
      "Processed 700000 pairs\n",
      "Processed 800000 pairs\n",
      "Processed 900000 pairs\n",
      "Processed 1000000 pairs\n",
      "Processed 1100000 pairs\n",
      "Processed 1200000 pairs\n",
      "Processed 1300000 pairs\n",
      "Processed 1400000 pairs\n",
      "Processed 1500000 pairs\n",
      "Processed 1600000 pairs\n",
      "Processed 1700000 pairs\n",
      "Processed 1800000 pairs\n",
      "Processed 1900000 pairs\n",
      "Processed 2000000 pairs\n",
      "Processed 2100000 pairs\n",
      "Finished filtering.  Time taken: 109.5s\n",
      "\n",
      "Found 286 / 2147046 to keep\n",
      "\n",
      "KAT filter seq completed.\n",
      "Total runtime: 109.6s\n",
      "\n",
      "\tCommand being timed: \"kat filter seq -t 1 -o SRR1952908.kat.2 -m 31 -T 0.0000001 --seq ../input/SRR1952908_1.fastq --seq2 ../input/SRR1952908_2.fastq kmer-probe/sul3.2.kmer.fasta\"\n",
      "\tUser time (seconds): 108.22\n",
      "\tSystem time (seconds): 1.80\n",
      "\tPercent of CPU this job got: 100%\n",
      "\tElapsed (wall clock) time (h:mm:ss or m:ss): 1:49.65\n",
      "\tAverage shared text size (kbytes): 0\n",
      "\tAverage unshared data size (kbytes): 0\n",
      "\tAverage stack size (kbytes): 0\n",
      "\tAverage total size (kbytes): 0\n",
      "\tMaximum resident set size (kbytes): 843728\n",
      "\tAverage resident set size (kbytes): 0\n",
      "\tMajor (requiring I/O) page faults: 0\n",
      "\tMinor (reclaiming a frame) page faults: 419718\n",
      "\tVoluntary context switches: 6\n",
      "\tInvoluntary context switches: 57\n",
      "\tSwaps: 0\n",
      "\tFile system inputs: 0\n",
      "\tFile system outputs: 376\n",
      "\tSocket messages sent: 0\n",
      "\tSocket messages received: 0\n",
      "\tSignals delivered: 0\n",
      "\tPage size (bytes): 4096\n",
      "\tExit status: 0\n"
     ]
    }
   ],
   "source": [
    "conda run --name kat /usr/bin/time -v kat filter seq -t ${threads} -o SRR1952908.kat.2 -m ${kmer_length} -T 0.0000001 \\\n",
    "    --seq ../input/SRR1952908_1.fastq --seq2 ../input/SRR1952908_2.fastq \\\n",
    "    kmer-probe/sul3.2.kmer.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2.2 Assemble reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "skesa --cores 1 --fastq SRR1952908.kat.2.in.R1.fastq,SRR1952908.kat.2.in.R2.fastq --contigs_out SRR1952908.kat.2.contigs.fasta --vector_percent 1 \n",
      "\n",
      "Total mates: 572 Paired reads: 286\n",
      "Reads acquired in  0.003861s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Adapters clip is disabled\n",
      "\n",
      "Kmer len: 21\n",
      "Raw kmers: 45509 Memory needed (GB): 0.000873772 Memory available (GB): 30 1 cycle(s) will be performed\n",
      "Distinct kmers: 1427\n",
      "Kmer count in  0.023129s wall, 0.020000s user + 0.000000s system = 0.020000s CPU (86.5%)\n",
      "Uniq kmers merging in  0.001074s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "WARNING: --min_count changed from 2 to 3 because of high coverage for genome size 384\n",
      "WARNING: --max_kmer_count 10 to 15 because of high coverage for genome size 384\n",
      "Kmers branching in  0.003712s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (269.4%)\n",
      "\n",
      "Average read length: 100\n",
      "Genome size estimate: 170\n",
      "\n",
      "Kmer: 21 Graph size: 1213 Contigs in: 0\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.000007s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmers in multiple/single contigs: 0 0\n",
      "Fragments before: 2 1145\n",
      "Fragments after: 2 1145\n",
      "New seeds: 1\n",
      "New seeds in  0.002529s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Fragments before: 2 42\n",
      "Fragments after: 2 42\n",
      "Connectors: 0 Extenders: 2\n",
      "Connections and extensions in  0.001536s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Contigs out: 1 Genome: 1057 N50: 1057 L50: 1\n",
      "Assembled in  0.004197s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "\n",
      "\n",
      "Average count: 3.37647 Max kmer: 21\n",
      "Connected: 271 ambiguously connected: 0 from 286 mate pairs\n",
      "Connect pairs in  0.170280s wall, 0.170000s user + 0.000000s system = 0.170000s CPU (99.8%)\n",
      "\n",
      "N50 for inserts: 248\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 454\n",
      "Reads for connection: 420\n",
      "Internal reads: 6\n",
      "Reads cleaned in  0.021880s wall, 0.020000s user + 0.000000s system = 0.020000s CPU (91.4%)\n",
      "WARNING: iterations are disabled\n",
      "\n",
      "Connecting mate pairs using kmer length: 21\n",
      "Connected: 197 ambiguously connected: 0 from 210 mate pairs\n",
      "Connect pairs in  0.121954s wall, 0.130000s user + 0.000000s system = 0.130000s CPU (106.6%)\n",
      "Totally connected: 197\n",
      "Added notconnected: 0\n",
      "\n",
      "Kmer len: 25\n",
      "Raw kmers: 50315 Memory needed (GB): 0.000966048 Memory available (GB): 29.9999 1 cycle(s) will be performed\n",
      "Distinct kmers: 1095\n",
      "Kmer count in  0.026072s wall, 0.020000s user + 0.010000s system = 0.030000s CPU (115.1%)\n",
      "Uniq kmers merging in  0.000895s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmers branching in  0.003411s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmer: 25 Graph size: 1095 Contigs in: 1\n",
      "Valley: 6\n",
      "\n",
      "Mark used kmers in  0.000767s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmers in multiple/single contigs: 0 1033\n",
      "Fragments before: 2 62\n",
      "Fragments after: 2 62\n",
      "New seeds: 0\n",
      "New seeds in  0.000722s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (n/a%)\n",
      "Fragments before: 2 62\n",
      "Fragments after: 2 62\n",
      "Connectors: 0 Extenders: 2\n",
      "Connections and extensions in  0.001376s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Contigs out: 1 Genome: 1057 N50: 1057 L50: 1\n",
      "Assembled in  0.003012s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (332.0%)\n",
      "\n",
      "\n",
      "Kmer len: 137\n",
      "Raw kmers: 27764 Memory needed (GB): 0.00159921 Memory available (GB): 29.9999 1 cycle(s) will be performed\n",
      "Distinct kmers: 983\n",
      "Kmer count in  0.023563s wall, 0.010000s user + 0.010000s system = 0.020000s CPU (84.9%)\n",
      "Uniq kmers merging in  0.000885s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmers branching in  0.004674s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (214.0%)\n",
      "Kmer: 137 Graph size: 983 Contigs in: 1\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.000953s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmers in multiple/single contigs: 0 921\n",
      "Fragments before: 2 62\n",
      "Fragments after: 2 62\n",
      "New seeds: 0\n",
      "New seeds in  0.000873s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Fragments before: 2 62\n",
      "Fragments after: 2 62\n",
      "Connectors: 0 Extenders: 2\n",
      "Connections and extensions in  0.002038s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Contigs out: 1 Genome: 1057 N50: 1057 L50: 1\n",
      "Assembled in  0.004004s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "\n",
      "\n",
      "Kmer len: 247\n",
      "Raw kmers: 10599 Memory needed (GB): 0.000915753 Memory available (GB): 29.9999 1 cycle(s) will be performed\n",
      "Distinct kmers: 708\n",
      "Kmer count in  0.013542s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (73.8%)\n",
      "Uniq kmers merging in  0.000975s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmers branching in  0.004093s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (244.3%)\n",
      "Kmer: 247 Graph size: 708 Contigs in: 1\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.000963s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmers in multiple/single contigs: 0 646\n",
      "Fragments before: 2 62\n",
      "Fragments after: 2 62\n",
      "New seeds: 0\n",
      "New seeds in  0.000812s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Fragments before: 2 62\n",
      "Fragments after: 2 62\n",
      "Connectors: 0 Extenders: 2\n",
      "Connections and extensions in  0.002686s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Contigs out: 1 Genome: 1057 N50: 1057 L50: 1\n",
      "Assembled in  0.004575s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "\n",
      "DONE\n",
      "\tCommand being timed: \"skesa --cores 1 --fastq SRR1952908.kat.2.in.R1.fastq,SRR1952908.kat.2.in.R2.fastq --contigs_out SRR1952908.kat.2.contigs.fasta --vector_percent 1\"\n",
      "\tUser time (seconds): 0.42\n",
      "\tSystem time (seconds): 0.02\n",
      "\tPercent of CPU this job got: 101%\n",
      "\tElapsed (wall clock) time (h:mm:ss or m:ss): 0:00.45\n",
      "\tAverage shared text size (kbytes): 0\n",
      "\tAverage unshared data size (kbytes): 0\n",
      "\tAverage stack size (kbytes): 0\n",
      "\tAverage total size (kbytes): 0\n",
      "\tMaximum resident set size (kbytes): 7104\n",
      "\tAverage resident set size (kbytes): 0\n",
      "\tMajor (requiring I/O) page faults: 0\n",
      "\tMinor (reclaiming a frame) page faults: 1301\n",
      "\tVoluntary context switches: 528\n",
      "\tInvoluntary context switches: 1\n",
      "\tSwaps: 0\n",
      "\tFile system inputs: 0\n",
      "\tFile system outputs: 8\n",
      "\tSocket messages sent: 0\n",
      "\tSocket messages received: 0\n",
      "\tSignals delivered: 0\n",
      "\tPage size (bytes): 4096\n",
      "\tExit status: 0\n"
     ]
    }
   ],
   "source": [
    "conda run --name skesa /usr/bin/time -v skesa --cores 1 \\\n",
    "    --fastq SRR1952908.kat.2.in.R1.fastq,SRR1952908.kat.2.in.R2.fastq \\\n",
    "    --contigs_out SRR1952908.kat.2.contigs.fasta --vector_percent 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.3 Examine assembled sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">Contig_1_41.9267\n",
      "AGTTTAAGTCCTCTTGATTATAGTGTAGGTATTGGAACAGGTTTAGAATTAGGACATTTCCAGTTTATGTTTGGCTATGATTACCCTCTTAGTAACAGTTCGAAAGATCCAGACCTAAGTCAAGTACGCCAACACAACTTCAGAGCAAGTATAGCATATTCATTTAGAAAGCTAAAAAAATAAAACTAACCTAGGGCTTTGGATATTTTCAAGGCATCTGATAAAGACTTAACATCATGGGTGCGGAGATAATCTGCACCTTTTTTGTATGCATACATTTCTGCTGCAAGAGTTGGTGCTAAACGAGATTTCACATCGGTTCCAGTTATTTTACCTAAGAATGATTTCCGTGACACTGCAATCATTACTTGCAAATTAAAAGCTTCTTGAATTTCAGGGAAACGCTTCAAAACAAGAATAGATGTTTCTGGATTAGAGCCTAAAAAGAAGCCCATACCCGGATCAAGAATAATTCGTTCACGCTTTACACCAGCCTCAACTAAAGCAGCAATTCTTTCTTTAAAAAATTCCATCATGGAAGTAAAAACCTCTTCCGGATTCGTTTCAACTTTAGTAGCTGCACCAATTCGCTGAACGGAGTGCATCAACACAAGTTTGCAATCTGACTTTGCCAAGCCTGAATAAATCTCAGGATAAGGAAAACCTTGAATATCATTAATAAAATCAACCTTTTGTTCTATGCAAAAACTCTGAACCTCAGGTTTAAATGTATCAACAGAAATAGAAATGCCTTTTTCTTTTAAAGCCTTAATGACAGGTTTGAGTCTTTTGATTTCTTCCACAACGCCCACTTCAGTTGTATCAGGATTACTGGAAGCGGCTCCCAAATCAATCACATCTGCTCCATCTTCAACCAAATGCAGAGCATGCTCAATTGCCTTATCTGTATCTAAATAAAGTCCTCCATCGGAAAAACTATCGGTGGTTATATTTACGATTCCAAAAATCTTGCTCATATCATTTTGCTTTGATTACAAAAGTAGCATTTATCTTAATTGACTAATTATCACTTACGCAAATTGACACATTTTCCCTA\n",
      "SRR1952908.kat.2.contigs.fasta 1057 bp 1 sequences\n"
     ]
    }
   ],
   "source": [
    "cat SRR1952908.kat.2.contigs.fasta\n",
    "bp_seq_length SRR1952908.kat.2.contigs.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BLASTN 2.9.0+\n",
      "\n",
      "\n",
      "Reference: Zheng Zhang, Scott Schwartz, Lukas Wagner, and Webb\n",
      "Miller (2000), \"A greedy algorithm for aligning DNA sequences\", J\n",
      "Comput Biol 2000; 7(1-2):203-14.\n",
      "\n",
      "\n",
      "\n",
      "Database: kmer-probe/sul3.fasta\n",
      "           1 sequences; 792 total letters\n",
      "\n",
      "\n",
      "\n",
      "Query= Contig_1_41.9267\n",
      "\n",
      "Length=1057\n",
      "                                                                      Score        E\n",
      "Sequences producing significant alignments:                          (Bits)     Value\n",
      "\n",
      "gb|FJ196385|-|8531-9323|ARO:3000413|sul3 [Escherichia coli]           1463       0.0  \n",
      "\n",
      "\n",
      ">gb|FJ196385|-|8531-9323|ARO:3000413|sul3 [Escherichia coli] \n",
      "Length=792\n",
      "\n",
      " Score = 1463 bits (792),  Expect = 0.0\n",
      " Identities = 792/792 (100%), Gaps = 0/792 (0%)\n",
      " Strand=Plus/Minus\n",
      "\n",
      "Query  186  CTAACCTAGGGCTTTGGATATTTTCAAGGCATCTGATAAAGACTTAACATCATGGGTGCG  245\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  792  CTAACCTAGGGCTTTGGATATTTTCAAGGCATCTGATAAAGACTTAACATCATGGGTGCG  733\n",
      "\n",
      "Query  246  GAGATAATCTGCACCTTTTTTGTATGCATACATTTCTGCTGCAAGAGTTGGTGCTAAACG  305\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  732  GAGATAATCTGCACCTTTTTTGTATGCATACATTTCTGCTGCAAGAGTTGGTGCTAAACG  673\n",
      "\n",
      "Query  306  AGATTTCACATCGGTTCCAGTTATTTTACCTAAGAATGATTTCCGTGACACTGCAATCAT  365\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  672  AGATTTCACATCGGTTCCAGTTATTTTACCTAAGAATGATTTCCGTGACACTGCAATCAT  613\n",
      "\n",
      "Query  366  TACTTGCAAATTAAAAGCTTCTTGAATTTCAGGGAAACGCTTCAAAACAAGAATAGATGT  425\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  612  TACTTGCAAATTAAAAGCTTCTTGAATTTCAGGGAAACGCTTCAAAACAAGAATAGATGT  553\n",
      "\n",
      "Query  426  TTCTGGATTAGAGCCTAAAAAGAAGCCCATACCCGGATCAAGAATAATTCGTTCACGCTT  485\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  552  TTCTGGATTAGAGCCTAAAAAGAAGCCCATACCCGGATCAAGAATAATTCGTTCACGCTT  493\n",
      "\n",
      "Query  486  TACACCAGCCTCAACTAAAGCAGCAATTCTTTCTTTAAAAAATTCCATCATGGAAGTAAA  545\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  492  TACACCAGCCTCAACTAAAGCAGCAATTCTTTCTTTAAAAAATTCCATCATGGAAGTAAA  433\n",
      "\n",
      "Query  546  AACCTCTTCCGGATTCGTTTCAACTTTAGTAGCTGCACCAATTCGCTGAACGGAGTGCAT  605\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  432  AACCTCTTCCGGATTCGTTTCAACTTTAGTAGCTGCACCAATTCGCTGAACGGAGTGCAT  373\n",
      "\n",
      "Query  606  CAACACAAGTTTGCAATCTGACTTTGCCAAGCCTGAATAAATCTCAGGATAAGGAAAACC  665\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  372  CAACACAAGTTTGCAATCTGACTTTGCCAAGCCTGAATAAATCTCAGGATAAGGAAAACC  313\n",
      "\n",
      "Query  666  TTGAATATCATTAATAAAATCAACCTTTTGTTCTATGCAAAAACTCTGAACCTCAGGTTT  725\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  312  TTGAATATCATTAATAAAATCAACCTTTTGTTCTATGCAAAAACTCTGAACCTCAGGTTT  253\n",
      "\n",
      "Query  726  AAATGTATCAACAGAAATAGAAATGCCTTTTTCTTTTAAAGCCTTAATGACAGGTTTGAG  785\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  252  AAATGTATCAACAGAAATAGAAATGCCTTTTTCTTTTAAAGCCTTAATGACAGGTTTGAG  193\n",
      "\n",
      "Query  786  TCTTTTGATTTCTTCCACAACGCCCACTTCAGTTGTATCAGGATTACTGGAAGCGGCTCC  845\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  192  TCTTTTGATTTCTTCCACAACGCCCACTTCAGTTGTATCAGGATTACTGGAAGCGGCTCC  133\n",
      "\n",
      "Query  846  CAAATCAATCACATCTGCTCCATCTTCAACCAAATGCAGAGCATGCTCAATTGCCTTATC  905\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  132  CAAATCAATCACATCTGCTCCATCTTCAACCAAATGCAGAGCATGCTCAATTGCCTTATC  73\n",
      "\n",
      "Query  906  TGTATCTAAATAAAGTCCTCCATCGGAAAAACTATCGGTGGTTATATTTACGATTCCAAA  965\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  72   TGTATCTAAATAAAGTCCTCCATCGGAAAAACTATCGGTGGTTATATTTACGATTCCAAA  13\n",
      "\n",
      "Query  966  AATCTTGCTCAT  977\n",
      "            ||||||||||||\n",
      "Sbjct  12   AATCTTGCTCAT  1\n",
      "\n",
      "\n",
      "\n",
      "Lambda      K        H\n",
      "    1.33    0.621     1.12 \n",
      "\n",
      "Gapped\n",
      "Lambda      K        H\n",
      "    1.28    0.460    0.850 \n",
      "\n",
      "Effective search space used: 813276\n",
      "\n",
      "\n",
      "  Database: kmer-probe/sul3.fasta\n",
      "    Posted date:  Jan 17, 2020  11:14 AM\n",
      "  Number of letters in database: 792\n",
      "  Number of sequences in database:  1\n",
      "\n",
      "\n",
      "\n",
      "Matrix: blastn matrix 1 -2\n",
      "Gap Penalties: Existence: 0, Extension: 2.5\n",
      "\tCommand being timed: \"blastn -db kmer-probe/sul3.fasta -query SRR1952908.kat.2.contigs.fasta\"\n",
      "\tUser time (seconds): 0.05\n",
      "\tSystem time (seconds): 0.02\n",
      "\tPercent of CPU this job got: 95%\n",
      "\tElapsed (wall clock) time (h:mm:ss or m:ss): 0:00.08\n",
      "\tAverage shared text size (kbytes): 0\n",
      "\tAverage unshared data size (kbytes): 0\n",
      "\tAverage stack size (kbytes): 0\n",
      "\tAverage total size (kbytes): 0\n",
      "\tMaximum resident set size (kbytes): 37068\n",
      "\tAverage resident set size (kbytes): 0\n",
      "\tMajor (requiring I/O) page faults: 2\n",
      "\tMinor (reclaiming a frame) page faults: 1954\n",
      "\tVoluntary context switches: 3\n",
      "\tInvoluntary context switches: 1\n",
      "\tSwaps: 0\n",
      "\tFile system inputs: 248\n",
      "\tFile system outputs: 0\n",
      "\tSocket messages sent: 0\n",
      "\tSocket messages received: 0\n",
      "\tSignals delivered: 0\n",
      "\tPage size (bytes): 4096\n",
      "\tExit status: 0\n"
     ]
    }
   ],
   "source": [
    "conda run --name blast /usr/bin/time -v blastn -db kmer-probe/sul3.fasta -query SRR1952908.kat.2.contigs.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Awesome. We matched 792/792 bp at 100% identity.\n",
    "\n",
    "Let's continue on extending.\n",
    "\n",
    "# Iteration 3\n",
    "\n",
    "## 3.0: Get kmers from ends"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">Contig_1_41.9267\n",
      "AGTTTAAGTCCTCTTGATTATAGTGTAGGTATTGGAACAGGTTTAGAATTAGGACATTTCCAGTTTATGTTTGGCTATGATTACCCTCTTAGTAACAGTTCGAAAGATCCAGACCTAAGTCAAGTACGCCAACACAACTTCAGAGCAAGTATAGCATATTCATTTAGAAAGCTAAAAAAATAAAACTAACCTAGGGCTTTGGATATTTTCAAGGCATCTGATAAAGACTTAACATCATGGGTGCGGAGATAATCTGCACCTTTTTTGTATGCATACATTTCTGCTGCAAGAGTTGGTGCTAAACGAGATTTCACATCGGTTCCAGTTATTTTACCTAAGAATGATTTCCGTGACACTGCAATCATTACTTGCAAATTAAAAGCTTCTTGAATTTCAGGGAAACGCTTCAAAACAAGAATAGATGTTTCTGGATTAGAGCCTAAAAAGAAGCCCATACCCGGATCAAGAATAATTCGTTCACGCTTTACACCAGCCTCAACTAAAGCAGCAATTCTTTCTTTAAAAAATTCCATCATGGAAGTAAAAACCTCTTCCGGATTCGTTTCAACTTTAGTAGCTGCACCAATTCGCTGAACGGAGTGCATCAACACAAGTTTGCAATCTGACTTTGCCAAGCCTGAATAAATCTCAGGATAAGGAAAACCTTGAATATCATTAATAAAATCAACCTTTTGTTCTATGCAAAAACTCTGAACCTCAGGTTTAAATGTATCAACAGAAATAGAAATGCCTTTTTCTTTTAAAGCCTTAATGACAGGTTTGAGTCTTTTGATTTCTTCCACAACGCCCACTTCAGTTGTATCAGGATTACTGGAAGCGGCTCCCAAATCAATCACATCTGCTCCATCTTCAACCAAATGCAGAGCATGCTCAATTGCCTTATCTGTATCTAAATAAAGTCCTCCATCGGAAAAACTATCGGTGGTTATATTTACGATTCCAAAAATCTTGCTCATATCATTTTGCTTTGATTACAAAAGTAGCATTTATCTTAATTGACTAATTATCACTTACGCAAATTGACACATTTTCCCTA\n",
      "SRR1952908.kat.2.contigs.fasta 1057 bp 1 sequences\n"
     ]
    }
   ],
   "source": [
    "cat SRR1952908.kat.2.contigs.fasta\n",
    "bp_seq_length SRR1952908.kat.2.contigs.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TGTGTTGATGCACTCCGTTCAGCGAATTGGT\n",
      "31\n",
      "31\n",
      "31\n",
      "31\n",
      "31\n",
      ">Contig_1_41.9267\n",
      " [AGTTTAAGTCCTCTTGATTATAGTGTAGGTA] TTGGAACAGGTTTAGAATTAGGACATTTCCAGTTTATGTTTGGCTATGATTACCCTCTTAGTAACAGTTCGAAAGATCCAGACCTAAGTCAAGTACGCCAACACAACTTCAGAGCAAGTATAGCATATTCATTTAGAAAGCTAAAAAAATAAAACTAACCTAGGGCTTTGGATATTTTCAAGGCATCTGATAAAGACTTAACATCATGGGTGC [GGAGATAATCTGCACCTTTTTTGTATGCATA] CATTTCTGCTGCAAGAGTTGGTGCTAAACGAGATTTCACATCGGTTCCAGTTATTTTACCTAAGAATGATTTCCGTGACACTGCAATCATTACTTGCAAATTAAAAGCTTCTTGAATTTCAGGGAAACGCTTCAAAACAAGAATAGATGTTTCTGGATTAGAGCCTAAAAAGAAGCCCATACCCGGATCAAGAATAATTCGTTCACGCTTTACACCAGCCTCAACTAAAGCAGCAATTCTTTCTTTAAAAAATTCCATCATGGAAGTAAAAACCTCTTCCGGATTCGTTTCAACTTTAGTAGCTGC [ACCAATTCGCTGAACGGAGTGCATCAACACA] AGTTTGCAATCTGACTTTGCCAAGCCTGAATAAATCTCAGGATAAGGAAAACCTTGAATATCATTAATAAAATCAACCTTTTGTTCTATGCAAAAACTCTGAACCTCAGGTTTAAATGTATCAACAGAAATAGAAATGCCTTTTTCTTTTAAAGCCTTAATGACAGGTTTGAGTCTTTTGATTTCTTCCACAACGCCCACTTCAGTTGTATCAGGATTACTGGAAGCGGCTC [CCAAATCAATCACATCTGCTCCATCTTCAAC] CAAATGCAGAGCATGCTCAATTGCCTTATCTGTATCTAAATAAAGTCCTCCATCGGAAAAACTATCGGTGGTTATATTTACGATTCCAAAAATCTTGCTCATATCATTTTGCTTTGATTACAAAAGTAGCATTTATCTTAATTGACTAATT [ATCACTTACGCAAATTGACACATTTTCCCTA] \n",
      ">kmer_orig\n",
      "TGTGTTGATGCACTCCGTTCAGCGAATTGGT\n",
      ">kmer1\n",
      "GGAGATAATCTGCACCTTTTTTGTATGCATA\n",
      ">kmer2\n",
      "CCAAATCAATCACATCTGCTCCATCTTCAAC\n",
      ">kmer3\n",
      "AGTTTAAGTCCTCTTGATTATAGTGTAGGTA\n",
      ">kmer4\n",
      "ATCACTTACGCAAATTGACACATTTTCCCTA\n"
     ]
    }
   ],
   "source": [
    "kmer_orig=${kmer}\n",
    "kmer_orig_rc='ACCAATTCGCTGAACGGAGTGCATCAACACA'\n",
    "echo ${kmer_orig}\n",
    "echo -n ${kmer_orig} | wc -c\n",
    "kmer1='GGAGATAATCTGCACCTTTTTTGTATGCATA'\n",
    "echo -n ${kmer1} | wc -c\n",
    "kmer2='CCAAATCAATCACATCTGCTCCATCTTCAAC'\n",
    "echo -n ${kmer2} | wc -c\n",
    "kmer3='AGTTTAAGTCCTCTTGATTATAGTGTAGGTA'\n",
    "echo -n ${kmer3} | wc -c\n",
    "kmer4='ATCACTTACGCAAATTGACACATTTTCCCTA'\n",
    "echo -n ${kmer4} | wc -c\n",
    "sed -re \"s/(${kmer_orig_rc}|${kmer1}|${kmer2}|${kmer3}|${kmer4})/ [\\1] /g\" SRR1952908.kat.2.contigs.fasta\n",
    "(echo -e \">kmer_orig\\n${kmer_orig}\\n>kmer1\\n${kmer1}\\n>kmer2\\n${kmer2}\\n>kmer3\\n${kmer3}\\n>kmer4\\n${kmer4}\";) \\\n",
    "    | tee kmer-probe/sul3.3.kmer.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3.1 Pull out reads matching kmer(s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Kmer Analysis Toolkit (KAT) V2.4.2\n",
      "\n",
      "Running KAT in filter sequence mode\n",
      "-----------------------------------\n",
      "\n",
      "Input 38080 is a sequence file.  Counting kmers for input 38080 (kmer-probe/sul3.3.kmer.fasta) ... done.  Time taken: 0.1s\n",
      "\n",
      "Filtering sequences ...\n",
      "Processed 100000 pairs\n",
      "Processed 200000 pairs\n",
      "Processed 300000 pairs\n",
      "Processed 400000 pairs\n",
      "Processed 500000 pairs\n",
      "Processed 600000 pairs\n",
      "Processed 700000 pairs\n",
      "Processed 800000 pairs\n",
      "Processed 900000 pairs\n",
      "Processed 1000000 pairs\n",
      "Processed 1100000 pairs\n",
      "Processed 1200000 pairs\n",
      "Processed 1300000 pairs\n",
      "Processed 1400000 pairs\n",
      "Processed 1500000 pairs\n",
      "Processed 1600000 pairs\n",
      "Processed 1700000 pairs\n",
      "Processed 1800000 pairs\n",
      "Processed 1900000 pairs\n",
      "Processed 2000000 pairs\n",
      "Processed 2100000 pairs\n",
      "Finished filtering.  Time taken: 108.1s\n",
      "\n",
      "Found 475 / 2147046 to keep\n",
      "\n",
      "KAT filter seq completed.\n",
      "Total runtime: 108.3s\n",
      "\n",
      "\tCommand being timed: \"kat filter seq -t 1 -o SRR1952908.kat.3 -m 31 -T 0.0000001 --seq ../input/SRR1952908_1.fastq --seq2 ../input/SRR1952908_2.fastq kmer-probe/sul3.3.kmer.fasta\"\n",
      "\tUser time (seconds): 107.11\n",
      "\tSystem time (seconds): 1.59\n",
      "\tPercent of CPU this job got: 100%\n",
      "\tElapsed (wall clock) time (h:mm:ss or m:ss): 1:48.32\n",
      "\tAverage shared text size (kbytes): 0\n",
      "\tAverage unshared data size (kbytes): 0\n",
      "\tAverage stack size (kbytes): 0\n",
      "\tAverage total size (kbytes): 0\n",
      "\tMaximum resident set size (kbytes): 843704\n",
      "\tAverage resident set size (kbytes): 0\n",
      "\tMajor (requiring I/O) page faults: 0\n",
      "\tMinor (reclaiming a frame) page faults: 262431\n",
      "\tVoluntary context switches: 8\n",
      "\tInvoluntary context switches: 59\n",
      "\tSwaps: 0\n",
      "\tFile system inputs: 0\n",
      "\tFile system outputs: 464\n",
      "\tSocket messages sent: 0\n",
      "\tSocket messages received: 0\n",
      "\tSignals delivered: 0\n",
      "\tPage size (bytes): 4096\n",
      "\tExit status: 0\n"
     ]
    }
   ],
   "source": [
    "conda run --name kat /usr/bin/time -v kat filter seq -t ${threads} -o SRR1952908.kat.3 -m ${kmer_length} -T 0.0000001 \\\n",
    "    --seq ../input/SRR1952908_1.fastq --seq2 ../input/SRR1952908_2.fastq \\\n",
    "    kmer-probe/sul3.3.kmer.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3.2 Assemble reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "skesa --cores 1 --fastq SRR1952908.kat.3.in.R1.fastq,SRR1952908.kat.3.in.R2.fastq --contigs_out SRR1952908.kat.3.contigs.fasta --vector_percent 1 \n",
      "\n",
      "Total mates: 950 Paired reads: 475\n",
      "Reads acquired in  0.006307s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Adapters clip is disabled\n",
      "\n",
      "Kmer len: 21\n",
      "Raw kmers: 75364 Memory needed (GB): 0.00144699 Memory available (GB): 29.9999 1 cycle(s) will be performed\n",
      "Distinct kmers: 2008\n",
      "Kmer count in  0.037942s wall, 0.030000s user + 0.010000s system = 0.040000s CPU (105.4%)\n",
      "Uniq kmers merging in  0.001272s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "WARNING: --min_count changed from 2 to 9 because of high coverage for genome size 215\n",
      "WARNING: --max_kmer_count 10 to 44 because of high coverage for genome size 215\n",
      "Kmers branching in  0.004549s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "\n",
      "Average read length: 99\n",
      "Genome size estimate: 120\n",
      "\n",
      "Kmer: 21 Graph size: 1515 Contigs in: 0\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.000007s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmers in multiple/single contigs: 0 0\n",
      "Fragments before: 3 1493\n",
      "Fragments after: 3 1493\n",
      "New seeds: 1\n",
      "New seeds in  0.003260s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Fragments before: 2 42\n",
      "Fragments after: 2 42\n",
      "Connectors: 0 Extenders: 2\n",
      "Connections and extensions in  0.001891s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (528.7%)\n",
      "Contigs out: 1 Genome: 1344 N50: 1344 L50: 1\n",
      "Assembled in  0.005308s wall, 0.010000s user + 0.010000s system = 0.020000s CPU (376.8%)\n",
      "\n",
      "\n",
      "Average count: 9.55 Max kmer: 21\n",
      "Connected: 419 ambiguously connected: 0 from 475 mate pairs\n",
      "Connect pairs in  0.290945s wall, 0.290000s user + 0.000000s system = 0.290000s CPU (99.7%)\n",
      "\n",
      "N50 for inserts: 258\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 278\n",
      "Reads for connection: 162\n",
      "Internal reads: 50\n",
      "Reads cleaned in  0.013190s wall, 0.020000s user + 0.000000s system = 0.020000s CPU (151.6%)\n",
      "WARNING: iterations are disabled\n",
      "\n",
      "Connecting mate pairs using kmer length: 21\n",
      "Connected: 34 ambiguously connected: 0 from 81 mate pairs\n",
      "Connect pairs in  0.039308s wall, 0.040000s user + 0.000000s system = 0.040000s CPU (101.8%)\n",
      "Totally connected: 34\n",
      "Added notconnected: 0\n",
      "\n",
      "Kmer len: 25\n",
      "Raw kmers: 20744 Memory needed (GB): 0.000398284 Memory available (GB): 30 1 cycle(s) will be performed\n",
      "Distinct kmers: 634\n",
      "Kmer count in  0.010890s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (91.8%)\n",
      "Uniq kmers merging in  0.000930s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmers branching in  0.001893s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmer: 25 Graph size: 634 Contigs in: 1\n",
      "Valley: 15\n",
      "\n",
      "Mark used kmers in  0.000892s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmers in multiple/single contigs: 0 613\n",
      "Fragments before: 1 21\n",
      "Fragments after: 1 21\n",
      "New seeds: 0\n",
      "New seeds in  0.000508s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Fragments before: 1 21\n",
      "Fragments after: 1 21\n",
      "Connectors: 0 Extenders: 1\n",
      "Connections and extensions in  0.001621s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Contigs out: 1 Genome: 1344 N50: 1344 L50: 1\n",
      "Assembled in  0.003141s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "\n",
      "\n",
      "Kmer len: 141\n",
      "Raw kmers: 11259 Memory needed (GB): 0.000648518 Memory available (GB): 29.9999 1 cycle(s) will be performed\n",
      "Distinct kmers: 387\n",
      "Kmer count in  0.010649s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (93.9%)\n",
      "Uniq kmers merging in  0.001090s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (917.4%)\n",
      "Kmers branching in  0.001984s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmer: 141 Graph size: 387 Contigs in: 1\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.001101s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmers in multiple/single contigs: 0 366\n",
      "Fragments before: 1 21\n",
      "Fragments after: 1 21\n",
      "New seeds: 0\n",
      "New seeds in  0.000458s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Fragments before: 1 21\n",
      "Fragments after: 1 21\n",
      "Connectors: 0 Extenders: 1\n",
      "Connections and extensions in  0.001750s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Contigs out: 1 Genome: 1344 N50: 1344 L50: 1\n",
      "Assembled in  0.003420s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "\n",
      "\n",
      "Kmer len: 257\n",
      "Raw kmers: 3857 Memory needed (GB): 0.000370272 Memory available (GB): 29.9999 1 cycle(s) will be performed\n",
      "Distinct kmers: 103\n",
      "Kmer count in  0.005761s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (173.6%)\n",
      "Uniq kmers merging in  0.000972s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmers branching in  0.000773s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmer: 257 Graph size: 103 Contigs in: 1\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.001114s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmers in multiple/single contigs: 0 82\n",
      "Fragments before: 1 21\n",
      "Fragments after: 1 21\n",
      "New seeds: 0\n",
      "New seeds in  0.000441s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Fragments before: 1 21\n",
      "Fragments after: 1 21\n",
      "Connectors: 0 Extenders: 1\n",
      "Connections and extensions in  0.001912s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Contigs out: 1 Genome: 1344 N50: 1344 L50: 1\n",
      "Assembled in  0.003573s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "\n",
      "DONE\n",
      "\tCommand being timed: \"skesa --cores 1 --fastq SRR1952908.kat.3.in.R1.fastq,SRR1952908.kat.3.in.R2.fastq --contigs_out SRR1952908.kat.3.contigs.fasta --vector_percent 1\"\n",
      "\tUser time (seconds): 0.43\n",
      "\tSystem time (seconds): 0.02\n",
      "\tPercent of CPU this job got: 101%\n",
      "\tElapsed (wall clock) time (h:mm:ss or m:ss): 0:00.45\n",
      "\tAverage shared text size (kbytes): 0\n",
      "\tAverage unshared data size (kbytes): 0\n",
      "\tAverage stack size (kbytes): 0\n",
      "\tAverage total size (kbytes): 0\n",
      "\tMaximum resident set size (kbytes): 6264\n",
      "\tAverage resident set size (kbytes): 0\n",
      "\tMajor (requiring I/O) page faults: 0\n",
      "\tMinor (reclaiming a frame) page faults: 869\n",
      "\tVoluntary context switches: 529\n",
      "\tInvoluntary context switches: 0\n",
      "\tSwaps: 0\n",
      "\tFile system inputs: 0\n",
      "\tFile system outputs: 8\n",
      "\tSocket messages sent: 0\n",
      "\tSocket messages received: 0\n",
      "\tSignals delivered: 0\n",
      "\tPage size (bytes): 4096\n",
      "\tExit status: 0\n"
     ]
    }
   ],
   "source": [
    "conda run --name skesa /usr/bin/time -v skesa --cores 1 \\\n",
    "    --fastq SRR1952908.kat.3.in.R1.fastq,SRR1952908.kat.3.in.R2.fastq \\\n",
    "    --contigs_out SRR1952908.kat.3.contigs.fasta --vector_percent 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3.3 Examine assembled contig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">Contig_1_52.4456\n",
      "CCCTGCGGGTACTTACTCCTTTGACATACATTTCGGCGATGGCGAGCTTGAGAGCACGCTCACTGCGCATGCCTTTCTCGATGCAGCTGGGATAAAACTCCATCCCGCGAACTTGTGGTACTTTCAAGAGGACTTTCCCGGTCGCAAGATTGAATATTGGATTATTTGGATTTACCAACAAACACTACTATTTAGCTACGAAAACTCAAGAGAAGTACACAAAACCTTTTGGAAGTGATGGAAGTTTAAGTCCTCTTGATTATAGTGTAGGTATTGGAACAGGTTTAGAATTAGGACATTTCCAGTTTATGTTTGGCTATGATTACCCTCTTAGTAACAGTTCGAAAGATCCAGACCTAAGTCAAGTACGCCAACACAACTTCAGAGCAAGTATAGCATATTCATTTAGAAAGCTAAAAAAATAAAACTAACCTAGGGCTTTGGATATTTTCAAGGCATCTGATAAAGACTTAACATCATGGGTGCGGAGATAATCTGCACCTTTTTTGTATGCATACATTTCTGCTGCAAGAGTTGGTGCTAAACGAGATTTCACATCGGTTCCAGTTATTTTACCTAAGAATGATTTCCGTGACACTGCAATCATTACTTGCAAATTAAAAGCTTCTTGAATTTCAGGGAAACGCTTCAAAACAAGAATAGATGTTTCTGGATTAGAGCCTAAAAAGAAGCCCATACCCGGATCAAGAATAATTCGTTCACGCTTTACACCAGCCTCAACTAAAGCAGCAATTCTTTCTTTAAAAAATTCCATCATGGAAGTAAAAACCTCTTCCGGATTCGTTTCAACTTTAGTAGCTGCACCAATTCGCTGAACGGAGTGCATCAACACAAGTTTGCAATCTGACTTTGCCAAGCCTGAATAAATCTCAGGATAAGGAAAACCTTGAATATCATTAATAAAATCAACCTTTTGTTCTATGCAAAAACTCTGAACCTCAGGTTTAAATGTATCAACAGAAATAGAAATGCCTTTTTCTTTTAAAGCCTTAATGACAGGTTTGAGTCTTTTGATTTCTTCCACAACGCCCACTTCAGTTGTATCAGGATTACTGGAAGCGGCTCCCAAATCAATCACATCTGCTCCATCTTCAACCAAATGCAGAGCATGCTCAATTGCCTTATCTGTATCTAAATAAAGTCCTCCATCGGAAAAACTATCGGTGGTTATATTTACGATTCCAAAAATCTTGCTCATATCATTTTGCTTTGATTACAAAAGTAGCATTTATCTTAATTGACTAATTATCACTTACGCAAATTGACACATTTTCCCTATCTTTGCAGCCTAATTTTTTTGTTTTAATTCATTATGAAAACTGA\n",
      "SRR1952908.kat.3.contigs.fasta 1344 bp 1 sequences\n"
     ]
    }
   ],
   "source": [
    "cat SRR1952908.kat.3.contigs.fasta\n",
    "bp_seq_length SRR1952908.kat.3.contigs.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Align to original gene to identify flanking regions\n",
    "\n",
    "## Align iteration 1\n",
    "\n",
    "First let's reverse complement the fasta file (so I can easily copy/past alignment in correct orientation)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reverse and complement a nucleotide sequence\n",
      ">>Contig_1_19.1064 Reversed:\n",
      "GTTGAAGATGGAGCAGATGTGATTGATTTGGGAGCCGCTTCCAGTAATCCTGATACAACTGAAGTGGGCGTTGTGGAAGAAATCAAAAGACTCAAACCTGTCATTAAGGCTTTAAAAGAAAAAGGCATTTCTATTTCTGTTGATACATTTAAACCTGAGGTTCAGAGTTTTTGCATAGAACAAAAGGTTGATTTTATTAATGATATTCAAGGTTTTCCTTATCCTGAGATTTATTCAGGCTTGGCAAAGTCAGATTGCAAACTTGTGTTGATGCACTCCGTTCAGCGAATTGGTGCAGCTACTAAAGTTGAAACGAATCCGGAAGAGGTTTTTACTTCCATGATGGAATTTTTTAAAGAAAGAATTGCTGCTTTAGTTGAGGCTGGTGTAAAGCGTGAACGAATTATTCTTGATCCGGGTATGGGCTTCTTTTTAGGCTCTAATCCAGAAACATCTATTCTTGTTTTGAAGCGTTTCCCTGAAATTCAAGAAGCTTTTAATTTGCAAGTAATGATTGCAGTGTCACGGAAATCATTCTTAGGTAAAATAACTGGAACCGATGTGAAATCTCGTTTAGCACCAACTCTTGCAGCAGAAATGTATGCATACAAAAAAGGTGCAGATTATCTCC\n",
      "SRR1952908.kat.1.contigs.rc.single.fasta 631 bp 1 sequences\n"
     ]
    }
   ],
   "source": [
    "revseq -sequence SRR1952908.kat.1.contigs.fasta -outseq SRR1952908.kat.1.contigs.rc.fasta\n",
    "header=`head -n 1 SRR1952908.kat.1.contigs.rc.fasta`\n",
    "seq_one_line=`tail -n +2 SRR1952908.kat.1.contigs.rc.fasta | tr -d '\\n'`\n",
    "echo -e \">$header\\n${seq_one_line}\" | tee SRR1952908.kat.1.contigs.rc.single.fasta\n",
    "bp_seq_length SRR1952908.kat.1.contigs.rc.single.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Blasting."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BLASTN 2.9.0+\n",
      "\n",
      "\n",
      "Reference: Zheng Zhang, Scott Schwartz, Lukas Wagner, and Webb\n",
      "Miller (2000), \"A greedy algorithm for aligning DNA sequences\", J\n",
      "Comput Biol 2000; 7(1-2):203-14.\n",
      "\n",
      "\n",
      "\n",
      "Database: kmer-probe/sul3.fasta\n",
      "           1 sequences; 792 total letters\n",
      "\n",
      "\n",
      "\n",
      "Query= >Contig_1_19.1064 Reversed:\n",
      "\n",
      "Length=631\n",
      "                                                                      Score        E\n",
      "Sequences producing significant alignments:                          (Bits)     Value\n",
      "\n",
      "gb|FJ196385|-|8531-9323|ARO:3000413|sul3 [Escherichia coli]           1166       0.0  \n",
      "\n",
      "\n",
      ">gb|FJ196385|-|8531-9323|ARO:3000413|sul3 [Escherichia coli] \n",
      "Length=792\n",
      "\n",
      " Score = 1166 bits (631),  Expect = 0.0\n",
      " Identities = 631/631 (100%), Gaps = 0/631 (0%)\n",
      " Strand=Plus/Plus\n",
      "\n",
      "Query  1    GTTGAAGATGGAGCAGATGTGATTGATTTGGGAGCCGCTTCCAGTAATCCTGATACAACT  60\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  103  GTTGAAGATGGAGCAGATGTGATTGATTTGGGAGCCGCTTCCAGTAATCCTGATACAACT  162\n",
      "\n",
      "Query  61   GAAGTGGGCGTTGTGGAAGAAATCAAAAGACTCAAACCTGTCATTAAGGCTTTAAAAGAA  120\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  163  GAAGTGGGCGTTGTGGAAGAAATCAAAAGACTCAAACCTGTCATTAAGGCTTTAAAAGAA  222\n",
      "\n",
      "Query  121  AAAGGCATTTCTATTTCTGTTGATACATTTAAACCTGAGGTTCAGAGTTTTTGCATAGAA  180\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  223  AAAGGCATTTCTATTTCTGTTGATACATTTAAACCTGAGGTTCAGAGTTTTTGCATAGAA  282\n",
      "\n",
      "Query  181  CAAAAGGTTGATTTTATTAATGATATTCAAGGTTTTCCTTATCCTGAGATTTATTCAGGC  240\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  283  CAAAAGGTTGATTTTATTAATGATATTCAAGGTTTTCCTTATCCTGAGATTTATTCAGGC  342\n",
      "\n",
      "Query  241  TTGGCAAAGTCAGATTGCAAACTTGTGTTGATGCACTCCGTTCAGCGAATTGGTGCAGCT  300\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  343  TTGGCAAAGTCAGATTGCAAACTTGTGTTGATGCACTCCGTTCAGCGAATTGGTGCAGCT  402\n",
      "\n",
      "Query  301  ACTAAAGTTGAAACGAATCCGGAAGAGGTTTTTACTTCCATGATGGAATTTTTTAAAGAA  360\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  403  ACTAAAGTTGAAACGAATCCGGAAGAGGTTTTTACTTCCATGATGGAATTTTTTAAAGAA  462\n",
      "\n",
      "Query  361  AGAATTGCTGCTTTAGTTGAGGCTGGTGTAAAGCGTGAACGAATTATTCTTGATCCGGGT  420\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  463  AGAATTGCTGCTTTAGTTGAGGCTGGTGTAAAGCGTGAACGAATTATTCTTGATCCGGGT  522\n",
      "\n",
      "Query  421  ATGGGCTTCTTTTTAGGCTCTAATCCAGAAACATCTATTCTTGTTTTGAAGCGTTTCCCT  480\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  523  ATGGGCTTCTTTTTAGGCTCTAATCCAGAAACATCTATTCTTGTTTTGAAGCGTTTCCCT  582\n",
      "\n",
      "Query  481  GAAATTCAAGAAGCTTTTAATTTGCAAGTAATGATTGCAGTGTCACGGAAATCATTCTTA  540\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  583  GAAATTCAAGAAGCTTTTAATTTGCAAGTAATGATTGCAGTGTCACGGAAATCATTCTTA  642\n",
      "\n",
      "Query  541  GGTAAAATAACTGGAACCGATGTGAAATCTCGTTTAGCACCAACTCTTGCAGCAGAAATG  600\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  643  GGTAAAATAACTGGAACCGATGTGAAATCTCGTTTAGCACCAACTCTTGCAGCAGAAATG  702\n",
      "\n",
      "Query  601  TATGCATACAAAAAAGGTGCAGATTATCTCC  631\n",
      "            |||||||||||||||||||||||||||||||\n",
      "Sbjct  703  TATGCATACAAAAAAGGTGCAGATTATCTCC  733\n",
      "\n",
      "\n",
      "\n",
      "Lambda      K        H\n",
      "    1.33    0.621     1.12 \n",
      "\n",
      "Gapped\n",
      "Lambda      K        H\n",
      "    1.28    0.460    0.850 \n",
      "\n",
      "Effective search space used: 482820\n",
      "\n",
      "\n",
      "  Database: kmer-probe/sul3.fasta\n",
      "    Posted date:  Jan 17, 2020  11:14 AM\n",
      "  Number of letters in database: 792\n",
      "  Number of sequences in database:  1\n",
      "\n",
      "\n",
      "\n",
      "Matrix: blastn matrix 1 -2\n",
      "Gap Penalties: Existence: 0, Extension: 2.5\n",
      "\tCommand being timed: \"blastn -db kmer-probe/sul3.fasta -query SRR1952908.kat.1.contigs.rc.single.fasta\"\n",
      "\tUser time (seconds): 0.05\n",
      "\tSystem time (seconds): 0.00\n",
      "\tPercent of CPU this job got: 98%\n",
      "\tElapsed (wall clock) time (h:mm:ss or m:ss): 0:00.06\n",
      "\tAverage shared text size (kbytes): 0\n",
      "\tAverage unshared data size (kbytes): 0\n",
      "\tAverage stack size (kbytes): 0\n",
      "\tAverage total size (kbytes): 0\n",
      "\tMaximum resident set size (kbytes): 36836\n",
      "\tAverage resident set size (kbytes): 0\n",
      "\tMajor (requiring I/O) page faults: 0\n",
      "\tMinor (reclaiming a frame) page faults: 1948\n",
      "\tVoluntary context switches: 1\n",
      "\tInvoluntary context switches: 2\n",
      "\tSwaps: 0\n",
      "\tFile system inputs: 0\n",
      "\tFile system outputs: 0\n",
      "\tSocket messages sent: 0\n",
      "\tSocket messages received: 0\n",
      "\tSignals delivered: 0\n",
      "\tPage size (bytes): 4096\n",
      "\tExit status: 0\n"
     ]
    }
   ],
   "source": [
    "conda run --name blast /usr/bin/time -v blastn -db kmer-probe/sul3.fasta -query SRR1952908.kat.1.contigs.rc.single.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Align to iteration 2\n",
    "\n",
    "Reverse complement."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reverse and complement a nucleotide sequence\n",
      ">>Contig_1_41.9267 Reversed:\n",
      "TAGGGAAAATGTGTCAATTTGCGTAAGTGATAATTAGTCAATTAAGATAAATGCTACTTTTGTAATCAAAGCAAAATGATATGAGCAAGATTTTTGGAATCGTAAATATAACCACCGATAGTTTTTCCGATGGAGGACTTTATTTAGATACAGATAAGGCAATTGAGCATGCTCTGCATTTGGTTGAAGATGGAGCAGATGTGATTGATTTGGGAGCCGCTTCCAGTAATCCTGATACAACTGAAGTGGGCGTTGTGGAAGAAATCAAAAGACTCAAACCTGTCATTAAGGCTTTAAAAGAAAAAGGCATTTCTATTTCTGTTGATACATTTAAACCTGAGGTTCAGAGTTTTTGCATAGAACAAAAGGTTGATTTTATTAATGATATTCAAGGTTTTCCTTATCCTGAGATTTATTCAGGCTTGGCAAAGTCAGATTGCAAACTTGTGTTGATGCACTCCGTTCAGCGAATTGGTGCAGCTACTAAAGTTGAAACGAATCCGGAAGAGGTTTTTACTTCCATGATGGAATTTTTTAAAGAAAGAATTGCTGCTTTAGTTGAGGCTGGTGTAAAGCGTGAACGAATTATTCTTGATCCGGGTATGGGCTTCTTTTTAGGCTCTAATCCAGAAACATCTATTCTTGTTTTGAAGCGTTTCCCTGAAATTCAAGAAGCTTTTAATTTGCAAGTAATGATTGCAGTGTCACGGAAATCATTCTTAGGTAAAATAACTGGAACCGATGTGAAATCTCGTTTAGCACCAACTCTTGCAGCAGAAATGTATGCATACAAAAAAGGTGCAGATTATCTCCGCACCCATGATGTTAAGTCTTTATCAGATGCCTTGAAAATATCCAAAGCCCTAGGTTAGTTTTATTTTTTTAGCTTTCTAAATGAATATGCTATACTTGCTCTGAAGTTGTGTTGGCGTACTTGACTTAGGTCTGGATCTTTCGAACTGTTACTAAGAGGGTAATCATAGCCAAACATAAACTGGAAATGTCCTAATTCTAAACCTGTTCCAATACCTACACTATAATCAAGAGGACTTAAACT\n",
      "SRR1952908.kat.2.contigs.rc.single.fasta 1057 bp 1 sequences\n"
     ]
    }
   ],
   "source": [
    "revseq -sequence SRR1952908.kat.2.contigs.fasta -outseq SRR1952908.kat.2.contigs.rc.fasta\n",
    "header=`head -n 1 SRR1952908.kat.2.contigs.rc.fasta`\n",
    "seq_one_line=`tail -n +2 SRR1952908.kat.2.contigs.rc.fasta | tr -d '\\n'`\n",
    "echo -e \">$header\\n${seq_one_line}\" | tee SRR1952908.kat.2.contigs.rc.single.fasta\n",
    "bp_seq_length SRR1952908.kat.2.contigs.rc.single.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's do a BLAST alignment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BLASTN 2.9.0+\n",
      "\n",
      "\n",
      "Reference: Zheng Zhang, Scott Schwartz, Lukas Wagner, and Webb\n",
      "Miller (2000), \"A greedy algorithm for aligning DNA sequences\", J\n",
      "Comput Biol 2000; 7(1-2):203-14.\n",
      "\n",
      "\n",
      "\n",
      "Database: kmer-probe/sul3.fasta\n",
      "           1 sequences; 792 total letters\n",
      "\n",
      "\n",
      "\n",
      "Query= >Contig_1_41.9267 Reversed:\n",
      "\n",
      "Length=1057\n",
      "                                                                      Score        E\n",
      "Sequences producing significant alignments:                          (Bits)     Value\n",
      "\n",
      "gb|FJ196385|-|8531-9323|ARO:3000413|sul3 [Escherichia coli]           1463       0.0  \n",
      "\n",
      "\n",
      ">gb|FJ196385|-|8531-9323|ARO:3000413|sul3 [Escherichia coli] \n",
      "Length=792\n",
      "\n",
      " Score = 1463 bits (792),  Expect = 0.0\n",
      " Identities = 792/792 (100%), Gaps = 0/792 (0%)\n",
      " Strand=Plus/Plus\n",
      "\n",
      "Query  81   ATGAGCAAGATTTTTGGAATCGTAAATATAACCACCGATAGTTTTTCCGATGGAGGACTT  140\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  1    ATGAGCAAGATTTTTGGAATCGTAAATATAACCACCGATAGTTTTTCCGATGGAGGACTT  60\n",
      "\n",
      "Query  141  TATTTAGATACAGATAAGGCAATTGAGCATGCTCTGCATTTGGTTGAAGATGGAGCAGAT  200\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  61   TATTTAGATACAGATAAGGCAATTGAGCATGCTCTGCATTTGGTTGAAGATGGAGCAGAT  120\n",
      "\n",
      "Query  201  GTGATTGATTTGGGAGCCGCTTCCAGTAATCCTGATACAACTGAAGTGGGCGTTGTGGAA  260\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  121  GTGATTGATTTGGGAGCCGCTTCCAGTAATCCTGATACAACTGAAGTGGGCGTTGTGGAA  180\n",
      "\n",
      "Query  261  GAAATCAAAAGACTCAAACCTGTCATTAAGGCTTTAAAAGAAAAAGGCATTTCTATTTCT  320\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  181  GAAATCAAAAGACTCAAACCTGTCATTAAGGCTTTAAAAGAAAAAGGCATTTCTATTTCT  240\n",
      "\n",
      "Query  321  GTTGATACATTTAAACCTGAGGTTCAGAGTTTTTGCATAGAACAAAAGGTTGATTTTATT  380\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  241  GTTGATACATTTAAACCTGAGGTTCAGAGTTTTTGCATAGAACAAAAGGTTGATTTTATT  300\n",
      "\n",
      "Query  381  AATGATATTCAAGGTTTTCCTTATCCTGAGATTTATTCAGGCTTGGCAAAGTCAGATTGC  440\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  301  AATGATATTCAAGGTTTTCCTTATCCTGAGATTTATTCAGGCTTGGCAAAGTCAGATTGC  360\n",
      "\n",
      "Query  441  AAACTTGTGTTGATGCACTCCGTTCAGCGAATTGGTGCAGCTACTAAAGTTGAAACGAAT  500\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  361  AAACTTGTGTTGATGCACTCCGTTCAGCGAATTGGTGCAGCTACTAAAGTTGAAACGAAT  420\n",
      "\n",
      "Query  501  CCGGAAGAGGTTTTTACTTCCATGATGGAATTTTTTAAAGAAAGAATTGCTGCTTTAGTT  560\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  421  CCGGAAGAGGTTTTTACTTCCATGATGGAATTTTTTAAAGAAAGAATTGCTGCTTTAGTT  480\n",
      "\n",
      "Query  561  GAGGCTGGTGTAAAGCGTGAACGAATTATTCTTGATCCGGGTATGGGCTTCTTTTTAGGC  620\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  481  GAGGCTGGTGTAAAGCGTGAACGAATTATTCTTGATCCGGGTATGGGCTTCTTTTTAGGC  540\n",
      "\n",
      "Query  621  TCTAATCCAGAAACATCTATTCTTGTTTTGAAGCGTTTCCCTGAAATTCAAGAAGCTTTT  680\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  541  TCTAATCCAGAAACATCTATTCTTGTTTTGAAGCGTTTCCCTGAAATTCAAGAAGCTTTT  600\n",
      "\n",
      "Query  681  AATTTGCAAGTAATGATTGCAGTGTCACGGAAATCATTCTTAGGTAAAATAACTGGAACC  740\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  601  AATTTGCAAGTAATGATTGCAGTGTCACGGAAATCATTCTTAGGTAAAATAACTGGAACC  660\n",
      "\n",
      "Query  741  GATGTGAAATCTCGTTTAGCACCAACTCTTGCAGCAGAAATGTATGCATACAAAAAAGGT  800\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  661  GATGTGAAATCTCGTTTAGCACCAACTCTTGCAGCAGAAATGTATGCATACAAAAAAGGT  720\n",
      "\n",
      "Query  801  GCAGATTATCTCCGCACCCATGATGTTAAGTCTTTATCAGATGCCTTGAAAATATCCAAA  860\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  721  GCAGATTATCTCCGCACCCATGATGTTAAGTCTTTATCAGATGCCTTGAAAATATCCAAA  780\n",
      "\n",
      "Query  861  GCCCTAGGTTAG  872\n",
      "            ||||||||||||\n",
      "Sbjct  781  GCCCTAGGTTAG  792\n",
      "\n",
      "\n",
      "\n",
      "Lambda      K        H\n",
      "    1.33    0.621     1.12 \n",
      "\n",
      "Gapped\n",
      "Lambda      K        H\n",
      "    1.28    0.460    0.850 \n",
      "\n",
      "Effective search space used: 813276\n",
      "\n",
      "\n",
      "  Database: kmer-probe/sul3.fasta\n",
      "    Posted date:  Jan 17, 2020  11:14 AM\n",
      "  Number of letters in database: 792\n",
      "  Number of sequences in database:  1\n",
      "\n",
      "\n",
      "\n",
      "Matrix: blastn matrix 1 -2\n",
      "Gap Penalties: Existence: 0, Extension: 2.5\n",
      "\tCommand being timed: \"blastn -db kmer-probe/sul3.fasta -query SRR1952908.kat.2.contigs.rc.single.fasta\"\n",
      "\tUser time (seconds): 0.05\n",
      "\tSystem time (seconds): 0.01\n",
      "\tPercent of CPU this job got: 98%\n",
      "\tElapsed (wall clock) time (h:mm:ss or m:ss): 0:00.06\n",
      "\tAverage shared text size (kbytes): 0\n",
      "\tAverage unshared data size (kbytes): 0\n",
      "\tAverage stack size (kbytes): 0\n",
      "\tAverage total size (kbytes): 0\n",
      "\tMaximum resident set size (kbytes): 37336\n",
      "\tAverage resident set size (kbytes): 0\n",
      "\tMajor (requiring I/O) page faults: 0\n",
      "\tMinor (reclaiming a frame) page faults: 1964\n",
      "\tVoluntary context switches: 1\n",
      "\tInvoluntary context switches: 1\n",
      "\tSwaps: 0\n",
      "\tFile system inputs: 0\n",
      "\tFile system outputs: 0\n",
      "\tSocket messages sent: 0\n",
      "\tSocket messages received: 0\n",
      "\tSignals delivered: 0\n",
      "\tPage size (bytes): 4096\n",
      "\tExit status: 0\n"
     ]
    }
   ],
   "source": [
    "conda run --name blast /usr/bin/time -v blastn -db kmer-probe/sul3.fasta -query SRR1952908.kat.2.contigs.rc.single.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's see where this is in the query (assembled) contig."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">>Contig_1_41.9267 Reversed:\n",
      "TAGGGAAAATGTGTCAATTTGCGTAAGTGATAATTAGTCAATTAAGATAAATGCTACTTTTGTAATCAAAGCAAAATGAT [ATGAGCAAGATTTTTGGAATCGTAAATATAACCACCGATAGTTTTTCCGATGGAGGACTT] TATTTAGATACAGATAAGGCAATTGAGCATGCTCTGCATTTGGTTGAAGATGGAGCAGATGTGATTGATTTGGGAGCCGCTTCCAGTAATCCTGATACAACTGAAGTGGGCGTTGTGGAAGAAATCAAAAGACTCAAACCTGTCATTAAGGCTTTAAAAGAAAAAGGCATTTCTATTTCTGTTGATACATTTAAACCTGAGGTTCAGAGTTTTTGCATAGAACAAAAGGTTGATTTTATTAATGATATTCAAGGTTTTCCTTATCCTGAGATTTATTCAGGCTTGGCAAAGTCAGATTGCAAACTTGTGTTGATGCACTCCGTTCAGCGAATTGGTGCAGCTACTAAAGTTGAAACGAATCCGGAAGAGGTTTTTACTTCCATGATGGAATTTTTTAAAGAAAGAATTGCTGCTTTAGTTGAGGCTGGTGTAAAGCGTGAACGAATTATTCTTGATCCGGGTATGGGCTTCTTTTTAGGCTCTAATCCAGAAACATCTATTCTTGTTTTGAAGCGTTTCCCTGAAATTCAAGAAGCTTTTAATTTGCAAGTAATGATTGCAGTGTCACGGAAATCATTCTTAGGTAAAATAACTGGAACCGATGTGAAATCTCGTTTAGCACCAACTCTTGCAGCAGAAATGTATGCATACAAAAAAGGTGCAGATTATCTCC [GCACCCATGATGTTAAGTCTTTATCAGATGCCTTGAAAATATCCAAAGCCCTAGGTTAG] TTTTATTTTTTTAGCTTTCTAAATGAATATGCTATACTTGCTCTGAAGTTGTGTTGGCGTACTTGACTTAGGTCTGGATCTTTCGAACTGTTACTAAGAGGGTAATCATAGCCAAACATAAACTGGAAATGTCCTAATTCTAAACCTGTTCCAATACCTACACTATAATCAAGAGGACTTAAACT\n"
     ]
    }
   ],
   "source": [
    "query_start='ATGAGCAAGATTTTTGGAATCGTAAATATAACCACCGATAGTTTTTCCGATGGAGGACTT'\n",
    "query_end='GCACCCATGATGTTAAGTCTTTATCAGATGCCTTGAAAATATCCAAAGCCCTAGGTTAG'\n",
    "sed -re \"s/(${query_start}|${query_end})/ [\\1] /g\" SRR1952908.kat.2.contigs.rc.single.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's count the flanking sequence bp."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "80\n",
      "185\n"
     ]
    }
   ],
   "source": [
    "echo -n 'TAGGGAAAATGTGTCAATTTGCGTAAGTGATAATTAGTCAATTAAGATAAATGCTACTTTTGTAATCAAAGCAAAATGAT' | wc -c\n",
    "echo -n 'TTTTATTTTTTTAGCTTTCTAAATGAATATGCTATACTTGCTCTGAAGTTGTGTTGGCGTACTTGACTTAGGTCTGGATCTTTCGAACTGTTACTAAGAGGGTAATCATAGCCAAACATAAACTGGAAATGTCCTAATTCTAAACCTGTTCCAATACCTACACTATAATCAAGAGGACTTAAACT' | wc -c"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Align to iteration 3\n",
    "\n",
    "Reverse complement."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reverse and complement a nucleotide sequence\n",
      ">>Contig_1_52.4456 Reversed:\n",
      "TCAGTTTTCATAATGAATTAAAACAAAAAAATTAGGCTGCAAAGATAGGGAAAATGTGTCAATTTGCGTAAGTGATAATTAGTCAATTAAGATAAATGCTACTTTTGTAATCAAAGCAAAATGATATGAGCAAGATTTTTGGAATCGTAAATATAACCACCGATAGTTTTTCCGATGGAGGACTTTATTTAGATACAGATAAGGCAATTGAGCATGCTCTGCATTTGGTTGAAGATGGAGCAGATGTGATTGATTTGGGAGCCGCTTCCAGTAATCCTGATACAACTGAAGTGGGCGTTGTGGAAGAAATCAAAAGACTCAAACCTGTCATTAAGGCTTTAAAAGAAAAAGGCATTTCTATTTCTGTTGATACATTTAAACCTGAGGTTCAGAGTTTTTGCATAGAACAAAAGGTTGATTTTATTAATGATATTCAAGGTTTTCCTTATCCTGAGATTTATTCAGGCTTGGCAAAGTCAGATTGCAAACTTGTGTTGATGCACTCCGTTCAGCGAATTGGTGCAGCTACTAAAGTTGAAACGAATCCGGAAGAGGTTTTTACTTCCATGATGGAATTTTTTAAAGAAAGAATTGCTGCTTTAGTTGAGGCTGGTGTAAAGCGTGAACGAATTATTCTTGATCCGGGTATGGGCTTCTTTTTAGGCTCTAATCCAGAAACATCTATTCTTGTTTTGAAGCGTTTCCCTGAAATTCAAGAAGCTTTTAATTTGCAAGTAATGATTGCAGTGTCACGGAAATCATTCTTAGGTAAAATAACTGGAACCGATGTGAAATCTCGTTTAGCACCAACTCTTGCAGCAGAAATGTATGCATACAAAAAAGGTGCAGATTATCTCCGCACCCATGATGTTAAGTCTTTATCAGATGCCTTGAAAATATCCAAAGCCCTAGGTTAGTTTTATTTTTTTAGCTTTCTAAATGAATATGCTATACTTGCTCTGAAGTTGTGTTGGCGTACTTGACTTAGGTCTGGATCTTTCGAACTGTTACTAAGAGGGTAATCATAGCCAAACATAAACTGGAAATGTCCTAATTCTAAACCTGTTCCAATACCTACACTATAATCAAGAGGACTTAAACTTCCATCACTTCCAAAAGGTTTTGTGTACTTCTCTTGAGTTTTCGTAGCTAAATAGTAGTGTTTGTTGGTAAATCCAAATAATCCAATATTCAATCTTGCGACCGGGAAAGTCCTCTTGAAAGTACCACAAGTTCGCGGGATGGAGTTTTATCCCAGCTGCATCGAGAAAGGCATGCGCAGTGAGCGTGCTCTCAAGCTCGCCATCGCCGAAATGTATGTCAAAGGAGTAAGTACCCGCAGGG\n",
      "SRR1952908.kat.3.contigs.rc.single.fasta 1344 bp 1 sequences\n"
     ]
    }
   ],
   "source": [
    "revseq -sequence SRR1952908.kat.3.contigs.fasta -outseq SRR1952908.kat.3.contigs.rc.fasta\n",
    "header=`head -n 1 SRR1952908.kat.3.contigs.rc.fasta`\n",
    "seq_one_line=`tail -n +2 SRR1952908.kat.3.contigs.rc.fasta | tr -d '\\n'`\n",
    "echo -e \">$header\\n${seq_one_line}\" | tee SRR1952908.kat.3.contigs.rc.single.fasta\n",
    "bp_seq_length SRR1952908.kat.3.contigs.rc.single.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Blasting."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BLASTN 2.9.0+\n",
      "\n",
      "\n",
      "Reference: Zheng Zhang, Scott Schwartz, Lukas Wagner, and Webb\n",
      "Miller (2000), \"A greedy algorithm for aligning DNA sequences\", J\n",
      "Comput Biol 2000; 7(1-2):203-14.\n",
      "\n",
      "\n",
      "\n",
      "Database: kmer-probe/sul3.fasta\n",
      "           1 sequences; 792 total letters\n",
      "\n",
      "\n",
      "\n",
      "Query= >Contig_1_52.4456 Reversed:\n",
      "\n",
      "Length=1344\n",
      "                                                                      Score        E\n",
      "Sequences producing significant alignments:                          (Bits)     Value\n",
      "\n",
      "gb|FJ196385|-|8531-9323|ARO:3000413|sul3 [Escherichia coli]           1463       0.0  \n",
      "\n",
      "\n",
      ">gb|FJ196385|-|8531-9323|ARO:3000413|sul3 [Escherichia coli] \n",
      "Length=792\n",
      "\n",
      " Score = 1463 bits (792),  Expect = 0.0\n",
      " Identities = 792/792 (100%), Gaps = 0/792 (0%)\n",
      " Strand=Plus/Plus\n",
      "\n",
      "Query  126  ATGAGCAAGATTTTTGGAATCGTAAATATAACCACCGATAGTTTTTCCGATGGAGGACTT  185\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  1    ATGAGCAAGATTTTTGGAATCGTAAATATAACCACCGATAGTTTTTCCGATGGAGGACTT  60\n",
      "\n",
      "Query  186  TATTTAGATACAGATAAGGCAATTGAGCATGCTCTGCATTTGGTTGAAGATGGAGCAGAT  245\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  61   TATTTAGATACAGATAAGGCAATTGAGCATGCTCTGCATTTGGTTGAAGATGGAGCAGAT  120\n",
      "\n",
      "Query  246  GTGATTGATTTGGGAGCCGCTTCCAGTAATCCTGATACAACTGAAGTGGGCGTTGTGGAA  305\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  121  GTGATTGATTTGGGAGCCGCTTCCAGTAATCCTGATACAACTGAAGTGGGCGTTGTGGAA  180\n",
      "\n",
      "Query  306  GAAATCAAAAGACTCAAACCTGTCATTAAGGCTTTAAAAGAAAAAGGCATTTCTATTTCT  365\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  181  GAAATCAAAAGACTCAAACCTGTCATTAAGGCTTTAAAAGAAAAAGGCATTTCTATTTCT  240\n",
      "\n",
      "Query  366  GTTGATACATTTAAACCTGAGGTTCAGAGTTTTTGCATAGAACAAAAGGTTGATTTTATT  425\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  241  GTTGATACATTTAAACCTGAGGTTCAGAGTTTTTGCATAGAACAAAAGGTTGATTTTATT  300\n",
      "\n",
      "Query  426  AATGATATTCAAGGTTTTCCTTATCCTGAGATTTATTCAGGCTTGGCAAAGTCAGATTGC  485\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  301  AATGATATTCAAGGTTTTCCTTATCCTGAGATTTATTCAGGCTTGGCAAAGTCAGATTGC  360\n",
      "\n",
      "Query  486  AAACTTGTGTTGATGCACTCCGTTCAGCGAATTGGTGCAGCTACTAAAGTTGAAACGAAT  545\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  361  AAACTTGTGTTGATGCACTCCGTTCAGCGAATTGGTGCAGCTACTAAAGTTGAAACGAAT  420\n",
      "\n",
      "Query  546  CCGGAAGAGGTTTTTACTTCCATGATGGAATTTTTTAAAGAAAGAATTGCTGCTTTAGTT  605\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  421  CCGGAAGAGGTTTTTACTTCCATGATGGAATTTTTTAAAGAAAGAATTGCTGCTTTAGTT  480\n",
      "\n",
      "Query  606  GAGGCTGGTGTAAAGCGTGAACGAATTATTCTTGATCCGGGTATGGGCTTCTTTTTAGGC  665\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  481  GAGGCTGGTGTAAAGCGTGAACGAATTATTCTTGATCCGGGTATGGGCTTCTTTTTAGGC  540\n",
      "\n",
      "Query  666  TCTAATCCAGAAACATCTATTCTTGTTTTGAAGCGTTTCCCTGAAATTCAAGAAGCTTTT  725\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  541  TCTAATCCAGAAACATCTATTCTTGTTTTGAAGCGTTTCCCTGAAATTCAAGAAGCTTTT  600\n",
      "\n",
      "Query  726  AATTTGCAAGTAATGATTGCAGTGTCACGGAAATCATTCTTAGGTAAAATAACTGGAACC  785\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  601  AATTTGCAAGTAATGATTGCAGTGTCACGGAAATCATTCTTAGGTAAAATAACTGGAACC  660\n",
      "\n",
      "Query  786  GATGTGAAATCTCGTTTAGCACCAACTCTTGCAGCAGAAATGTATGCATACAAAAAAGGT  845\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  661  GATGTGAAATCTCGTTTAGCACCAACTCTTGCAGCAGAAATGTATGCATACAAAAAAGGT  720\n",
      "\n",
      "Query  846  GCAGATTATCTCCGCACCCATGATGTTAAGTCTTTATCAGATGCCTTGAAAATATCCAAA  905\n",
      "            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Sbjct  721  GCAGATTATCTCCGCACCCATGATGTTAAGTCTTTATCAGATGCCTTGAAAATATCCAAA  780\n",
      "\n",
      "Query  906  GCCCTAGGTTAG  917\n",
      "            ||||||||||||\n",
      "Sbjct  781  GCCCTAGGTTAG  792\n",
      "\n",
      "\n",
      "\n",
      "Lambda      K        H\n",
      "    1.33    0.621     1.12 \n",
      "\n",
      "Gapped\n",
      "Lambda      K        H\n",
      "    1.28    0.460    0.850 \n",
      "\n",
      "Effective search space used: 1036849\n",
      "\n",
      "\n",
      "  Database: kmer-probe/sul3.fasta\n",
      "    Posted date:  Jan 17, 2020  11:14 AM\n",
      "  Number of letters in database: 792\n",
      "  Number of sequences in database:  1\n",
      "\n",
      "\n",
      "\n",
      "Matrix: blastn matrix 1 -2\n",
      "Gap Penalties: Existence: 0, Extension: 2.5\n",
      "\tCommand being timed: \"blastn -db kmer-probe/sul3.fasta -query SRR1952908.kat.3.contigs.rc.single.fasta\"\n",
      "\tUser time (seconds): 0.04\n",
      "\tSystem time (seconds): 0.01\n",
      "\tPercent of CPU this job got: 100%\n",
      "\tElapsed (wall clock) time (h:mm:ss or m:ss): 0:00.06\n",
      "\tAverage shared text size (kbytes): 0\n",
      "\tAverage unshared data size (kbytes): 0\n",
      "\tAverage stack size (kbytes): 0\n",
      "\tAverage total size (kbytes): 0\n",
      "\tMaximum resident set size (kbytes): 37368\n",
      "\tAverage resident set size (kbytes): 0\n",
      "\tMajor (requiring I/O) page faults: 0\n",
      "\tMinor (reclaiming a frame) page faults: 1962\n",
      "\tVoluntary context switches: 1\n",
      "\tInvoluntary context switches: 2\n",
      "\tSwaps: 0\n",
      "\tFile system inputs: 0\n",
      "\tFile system outputs: 0\n",
      "\tSocket messages sent: 0\n",
      "\tSocket messages received: 0\n",
      "\tSignals delivered: 0\n",
      "\tPage size (bytes): 4096\n",
      "\tExit status: 0\n"
     ]
    }
   ],
   "source": [
    "conda run --name blast /usr/bin/time -v blastn -db kmer-probe/sul3.fasta -query SRR1952908.kat.3.contigs.rc.single.fasta"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
