{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get signature/canonical k-mers from MLST scheme"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'aspA_1': {'GA', 'GC', 'CC', 'AA', 'AT', 'CG', 'TA', 'AG', 'CA', 'AC'}}\n",
      "{'AT': {'aspA_1'}, 'CA': {'aspA_1'}, 'GA': {'aspA_1'}, 'TA': {'aspA_1'}, 'AG': {'aspA_1'}, 'CC': {'aspA_1'}, 'AC': {'aspA_1'}, 'AA': {'aspA_1'}, 'GC': {'aspA_1'}, 'CG': {'aspA_1'}}\n"
     ]
    }
   ],
   "source": [
    "from pathlib import Path\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import reverse_complement\n",
    "import pandas as pd\n",
    "\n",
    "def canonical_kmer(kmer):\n",
    "    kmer_upper = kmer.upper()\n",
    "    kmer_rc = reverse_complement(kmer_upper)\n",
    "    if kmer_upper < kmer_rc:\n",
    "        return kmer_upper\n",
    "    else:\n",
    "        return kmer_rc\n",
    "\n",
    "def load_kmers(file_path: Path, kmer_size):\n",
    "    allele_table = {}\n",
    "    kmer_table = {}\n",
    "        \n",
    "    with open(file_path, \"r\") as file:\n",
    "        for record in SeqIO.parse(file, 'fasta'):\n",
    "            allele = str(record.id)\n",
    "            sequence = str(record.seq)\n",
    "            seq_length = len(sequence)\n",
    "            \n",
    "            max_start = seq_length - kmer_size + 1\n",
    "            \n",
    "            if max_start <= 0:\n",
    "                raise Exception(f\"Cannot extract any kmers from ({file}, {allele}): kmer size {kmer_size} longer than allele length {seq_length}\")\n",
    "\n",
    "            for start in range(0, max_start):\n",
    "                stop = start + kmer_size\n",
    "\n",
    "                kmer = canonical_kmer(sequence[start : stop])\n",
    "                \n",
    "                allele_table.setdefault(allele, set()).add(kmer)\n",
    "                kmer_table.setdefault(kmer, set()).add(allele)\n",
    "                \n",
    "    return (allele_table, kmer_table)\n",
    "                \n",
    "(allele_table, kmer_table) = load_kmers('aspA-1.fas', 2)\n",
    "print(allele_table)\n",
    "print(kmer_table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Kmers in (1,2) = (457, 457)\n",
      "Kmers in union = 543\n",
      "Kmers in intersection = 371\n",
      "Unique kmers aspA_1 = 86\n",
      "Unique kmers aspA_2 = 86\n",
      "Unique kmers in aspA_1 = {'CTGCTATTGGAACGGGAATTA', 'AGCAAAAGTTACAGTTACATC', 'CAGGATGAGAATTAATTCCCG', 'CGATCAAATCCTCAGCCACAG', 'AAAAGTTACAGTTACATCTGC', 'GACCACATTTTGGACCACTGC', 'AGATGTAACTGTAACTTTTGC', 'TTACATCTGCTCCAATAACAA', 'CAAAAGTTACAGTTACATCTG', 'CTGAGGATTTGATCGAGGCGA', 'ATTGGAGCAGATGTAACTGTA', 'ATTTTGTTATTGGAGCAGATG', 'GACCACTGCTTAAAAGTCTTA', 'TACATCTGCTCCAATAACAAA', 'ACTGCTTAAAAGTCTTAAGTC', 'CCACTGCTTAAAAGTCTTAAG', 'CTGCTTAAAAGTCTTAAGTCA', 'CAAGCAAAAGTTACAGTTACA', 'ACCACTGCTTAAAAGTCTTAA', 'AAGCAAAAGTTACAGTTACAT', 'AATTCCCGTTCCAATAGCAGT', 'ATTAATTCCCGTTCCAATAGC', 'CATTTTGGACCACTGCTTAAA', 'TGAGAATTAATTCCCGTTCCA', 'ACTGTGGCTGAGGATTTGATC', 'ACAGTTACATCTGCTCCAATA', 'GACTTTTAAGCAGTGGTCCAA', 'AATGACTTAAGACTTTTAAGC', 'TTACAGTTACATCTGCTCCAA', 'AAAGTTACAGTTACATCTGCT', 'AAGCAGTGGTCCAAAATGTGG', 'AGTCGCCTCGATCAAATCCTC', 'ACCACATTTTGGACCACTGCT', 'ATGAGAATTAATTCCCGTTCC', 'AGACTTTTAAGCAGTGGTCCA', 'ATACTGTGGCTGAGGATTTGA', 'GGATTTGATCGAGGCGACTCA', 'AACTGCTATTGGAACGGGAAT', 'ATTTGATCGAGGCGACTCAAG', 'GCAACTGGTTCAAAAACATTA', 'CCGTTCCAATAGCAGTTCCAC', 'CACATTTTGGACCACTGCTTA', 'TATCTTGAGTCGCCTCGATCA', 'CCTCGATCAAATCCTCAGCCA', 'GATCGAGGCGACTCAAGATAC', 'ATTTTGGACCACTGCTTAAAA', 'AGTTACATCTGCTCCAATAAC', 'GATTTGATCGAGGCGACTCAA', 'TCTTGAGTCGCCTCGATCAAA', 'CAGTTACATCTGCTCCAATAA', 'GGAACTGCTATTGGAACGGGA', 'ACTTAAGACTTTTAAGCAGTG', 'AGAATTAATTCCCGTTCCAAT', 'AATGTTTTTGAACCAGTTGCA', 'AGGATTTGATCGAGGCGACTC', 'ACATCTGCTCCAATAACAAAA', 'CTTTTAAGCAGTGGTCCAAAA', 'ACTTTTAAGCAGTGGTCCAAA', 'AACGGGAATTAATTCTCATCC', 'GAGAATTAATTCCCGTTCCAA', 'CTCGATCAAATCCTCAGCCAC', 'GTCGCCTCGATCAAATCCTCA', 'GGGAATTAATTCTCATCCTGA', 'GTTACATCTGCTCCAATAACA', 'ATCAAATCCTCAGCCACAGTA', 'ACGGGAATTAATTCTCATCCT', 'ATGACTTAAGACTTTTAAGCA', 'CGCCTCGATCAAATCCTCAGC', 'CGTTCCAATAGCAGTTCCACC', 'ATCTTGAGTCGCCTCGATCAA', 'TCGATCAAATCCTCAGCCACA', 'CAAATCCTCAGCCACAGTATA', 'TGCTATTGGAACGGGAATTAA', 'AAGTTACAGTTACATCTGCTC', 'AGTTACAGTTACATCTGCTCC', 'GTTACAGTTACATCTGCTCCA', 'CCCGTTCCAATAGCAGTTCCA', 'ATCTGCTCCAATAACAAAATA', 'GAACGGGAATTAATTCTCATC', 'AGACCACATTTTGGACCACTG', 'GCCTCGATCAAATCCTCAGCC', 'GAACTGCTATTGGAACGGGAA', 'GAATTAATTCCCGTTCCAATA', 'ACATTTTGGACCACTGCTTAA', 'AAGACTTTTAAGCAGTGGTCC', 'AATTAATTCCCGTTCCAATAG'}\n"
     ]
    }
   ],
   "source": [
    "(allele_kmers, kmer_table) = load_kmers('aspA-2.fas', 21)\n",
    "aspA_1 = allele_kmers['aspA_1']\n",
    "aspA_2 = allele_kmers['aspA_2']\n",
    "\n",
    "print(f\"Kmers in (1,2) = ({len(aspA_1)}, {len(aspA_2)})\")\n",
    "print(f\"Kmers in union = {len(aspA_1 | aspA_2)}\")\n",
    "print(f\"Kmers in intersection = {len(aspA_1 & aspA_2)}\")\n",
    "print(f\"Unique kmers aspA_1 = {len(aspA_1 - aspA_2)}\")\n",
    "print(f\"Unique kmers aspA_2 = {len(aspA_2 - aspA_1)}\")\n",
    "print(f\"Unique kmers in aspA_1 = {aspA_1 - aspA_2}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [],
   "source": [
    "(allele_kmers, kmer_table) = load_kmers('aspA-10.fas', 21)\n",
    "#(allele_kmers, kmer_table) = load_kmers('mlst-campy/aspA.fas', 21)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Unique kmers among all pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "aspA_1     aspA_2\n",
      "aspA_10    aspA_2\n",
      "aspA_2     aspA_5\n",
      "aspA_3     aspA_5\n",
      "aspA_4     aspA_2\n",
      "aspA_5     aspA_2\n",
      "aspA_6     aspA_5\n",
      "aspA_7     aspA_5\n",
      "aspA_8     aspA_5\n",
      "aspA_9     aspA_5\n",
      "dtype: object\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>aspA_1</th>\n",
       "      <th>aspA_10</th>\n",
       "      <th>aspA_2</th>\n",
       "      <th>aspA_3</th>\n",
       "      <th>aspA_4</th>\n",
       "      <th>aspA_5</th>\n",
       "      <th>aspA_6</th>\n",
       "      <th>aspA_7</th>\n",
       "      <th>aspA_8</th>\n",
       "      <th>aspA_9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>aspA_1</th>\n",
       "      <td>0</td>\n",
       "      <td>30</td>\n",
       "      <td>86</td>\n",
       "      <td>21</td>\n",
       "      <td>51</td>\n",
       "      <td>70</td>\n",
       "      <td>65</td>\n",
       "      <td>65</td>\n",
       "      <td>44</td>\n",
       "      <td>65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_10</th>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>116</td>\n",
       "      <td>51</td>\n",
       "      <td>21</td>\n",
       "      <td>40</td>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>74</td>\n",
       "      <td>95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_2</th>\n",
       "      <td>86</td>\n",
       "      <td>116</td>\n",
       "      <td>0</td>\n",
       "      <td>65</td>\n",
       "      <td>137</td>\n",
       "      <td>154</td>\n",
       "      <td>63</td>\n",
       "      <td>21</td>\n",
       "      <td>42</td>\n",
       "      <td>63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_3</th>\n",
       "      <td>21</td>\n",
       "      <td>51</td>\n",
       "      <td>65</td>\n",
       "      <td>0</td>\n",
       "      <td>72</td>\n",
       "      <td>91</td>\n",
       "      <td>44</td>\n",
       "      <td>44</td>\n",
       "      <td>65</td>\n",
       "      <td>86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_4</th>\n",
       "      <td>51</td>\n",
       "      <td>21</td>\n",
       "      <td>137</td>\n",
       "      <td>72</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>116</td>\n",
       "      <td>116</td>\n",
       "      <td>95</td>\n",
       "      <td>116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_5</th>\n",
       "      <td>70</td>\n",
       "      <td>40</td>\n",
       "      <td>154</td>\n",
       "      <td>91</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>121</td>\n",
       "      <td>133</td>\n",
       "      <td>112</td>\n",
       "      <td>121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_6</th>\n",
       "      <td>65</td>\n",
       "      <td>95</td>\n",
       "      <td>63</td>\n",
       "      <td>44</td>\n",
       "      <td>116</td>\n",
       "      <td>121</td>\n",
       "      <td>0</td>\n",
       "      <td>42</td>\n",
       "      <td>63</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_7</th>\n",
       "      <td>65</td>\n",
       "      <td>95</td>\n",
       "      <td>21</td>\n",
       "      <td>44</td>\n",
       "      <td>116</td>\n",
       "      <td>133</td>\n",
       "      <td>42</td>\n",
       "      <td>0</td>\n",
       "      <td>63</td>\n",
       "      <td>84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_8</th>\n",
       "      <td>44</td>\n",
       "      <td>74</td>\n",
       "      <td>42</td>\n",
       "      <td>65</td>\n",
       "      <td>95</td>\n",
       "      <td>112</td>\n",
       "      <td>63</td>\n",
       "      <td>63</td>\n",
       "      <td>0</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_9</th>\n",
       "      <td>65</td>\n",
       "      <td>95</td>\n",
       "      <td>63</td>\n",
       "      <td>86</td>\n",
       "      <td>116</td>\n",
       "      <td>121</td>\n",
       "      <td>42</td>\n",
       "      <td>84</td>\n",
       "      <td>21</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         aspA_1  aspA_10  aspA_2  aspA_3  aspA_4  aspA_5  aspA_6  aspA_7  \\\n",
       "aspA_1        0       30      86      21      51      70      65      65   \n",
       "aspA_10      30        0     116      51      21      40      95      95   \n",
       "aspA_2       86      116       0      65     137     154      63      21   \n",
       "aspA_3       21       51      65       0      72      91      44      44   \n",
       "aspA_4       51       21     137      72       0      19     116     116   \n",
       "aspA_5       70       40     154      91      19       0     121     133   \n",
       "aspA_6       65       95      63      44     116     121       0      42   \n",
       "aspA_7       65       95      21      44     116     133      42       0   \n",
       "aspA_8       44       74      42      65      95     112      63      63   \n",
       "aspA_9       65       95      63      86     116     121      42      84   \n",
       "\n",
       "         aspA_8  aspA_9  \n",
       "aspA_1       44      65  \n",
       "aspA_10      74      95  \n",
       "aspA_2       42      63  \n",
       "aspA_3       65      86  \n",
       "aspA_4       95     116  \n",
       "aspA_5      112     121  \n",
       "aspA_6       63      42  \n",
       "aspA_7       63      84  \n",
       "aspA_8        0      21  \n",
       "aspA_9       21       0  "
      ]
     },
     "execution_count": 145,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ordered_labels = sorted(allele_kmers.keys())\n",
    "\n",
    "data = []\n",
    "for allele1 in ordered_labels:\n",
    "    kmers1 = allele_kmers[allele1]\n",
    "    \n",
    "    row = [len(kmers1 - allele_kmers[allele2]) for allele2 in ordered_labels]\n",
    "    data.append(row)\n",
    "    \n",
    "df = pd.DataFrame(data, columns=ordered_labels, index=ordered_labels)\n",
    "print(df.idxmax())\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Unique kmers for any one over all others"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>aspA_1</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_10</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_2</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_3</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_4</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_5</th>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_6</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_7</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_8</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_9</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          0\n",
       "aspA_1    0\n",
       "aspA_10   0\n",
       "aspA_2    0\n",
       "aspA_3    0\n",
       "aspA_4    0\n",
       "aspA_5   19\n",
       "aspA_6    0\n",
       "aspA_7    0\n",
       "aspA_8    0\n",
       "aspA_9    0"
      ]
     },
     "execution_count": 146,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ordered_labels = sorted(allele_kmers.keys())\n",
    "\n",
    "data = []\n",
    "for allele1 in ordered_labels:\n",
    "    unique_kmers1 = allele_kmers[allele1]\n",
    "    \n",
    "    for allele2 in ordered_labels:\n",
    "        if allele2 != allele1 and len(unique_kmers1) > 0:\n",
    "            unique_kmers1 = unique_kmers1 - allele_kmers[allele2]\n",
    "            \n",
    "    data.append(len(unique_kmers1))\n",
    "    \n",
    "df2 = pd.DataFrame(data, index=ordered_labels)\n",
    "df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AAAAACGTTAAGTTGTAATTG\t{'aspA_5'}\n",
      "AAAACGTTAAGTTGTAATTGT\t{'aspA_5'}\n",
      "AAACGTTAAGTTGTAATTGTC\t{'aspA_5'}\n",
      "AACGTTAAGTTGTAATTGTCC\t{'aspA_5'}\n",
      "AACGTTTTTGAACCAGTTGCA\t{'aspA_5'}\n",
      "AACTGGTTCAAAAACGTTAAG\t{'aspA_5'}\n",
      "AACTTAACGTTTTTGAACCAG\t{'aspA_5'}\n",
      "AATTACAACTTAACGTTTTTG\t{'aspA_5'}\n",
      "ACAACTTAACGTTTTTGAACC\t{'aspA_5'}\n",
      "ACGTTAAGTTGTAATTGTCCA\t{'aspA_5'}\n",
      "ACTGGTTCAAAAACGTTAAGT\t{'aspA_5'}\n",
      "ATTACAACTTAACGTTTTTGA\t{'aspA_5'}\n",
      "CAACTGGTTCAAAAACGTTAA\t{'aspA_5'}\n",
      "CAACTTAACGTTTTTGAACCA\t{'aspA_5'}\n",
      "CGTTAAGTTGTAATTGTCCAC\t{'aspA_5'}\n",
      "GCAACTGGTTCAAAAACGTTA\t{'aspA_5'}\n",
      "GGTGGACAATTACAACTTAAC\t{'aspA_5'}\n",
      "GTTCAAAAACGTTAAGTTGTA\t{'aspA_5'}\n",
      "TTACAACTTAACGTTTTTGAA\t{'aspA_5'}\n"
     ]
    }
   ],
   "source": [
    "for kmer in sorted(kmer_table.keys()):\n",
    "    if len(kmer_table[kmer]) == 1:\n",
    "        print(f\"{kmer}\\t{kmer_table[kmer]}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Minimal k-mers differentiating one genome from all others"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>aspA_1</th>\n",
       "      <th>aspA_10</th>\n",
       "      <th>aspA_2</th>\n",
       "      <th>aspA_3</th>\n",
       "      <th>aspA_4</th>\n",
       "      <th>aspA_5</th>\n",
       "      <th>aspA_6</th>\n",
       "      <th>aspA_7</th>\n",
       "      <th>aspA_8</th>\n",
       "      <th>aspA_9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>457</td>\n",
       "      <td>457</td>\n",
       "      <td>457</td>\n",
       "      <td>457</td>\n",
       "      <td>457</td>\n",
       "      <td>457</td>\n",
       "      <td>457</td>\n",
       "      <td>457</td>\n",
       "      <td>457</td>\n",
       "      <td>457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>163</td>\n",
       "      <td>163</td>\n",
       "      <td>163</td>\n",
       "      <td>163</td>\n",
       "      <td>163</td>\n",
       "      <td>163</td>\n",
       "      <td>163</td>\n",
       "      <td>163</td>\n",
       "      <td>163</td>\n",
       "      <td>163</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   aspA_1  aspA_10  aspA_2  aspA_3  aspA_4  aspA_5  aspA_6  aspA_7  aspA_8  \\\n",
       "0     457      457     457     457     457     457     457     457     457   \n",
       "1     163      163     163     163     163     163     163     163     163   \n",
       "\n",
       "   aspA_9  \n",
       "0     457  \n",
       "1     163  "
      ]
     },
     "execution_count": 148,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ordered_labels = sorted(allele_kmers.keys()) \n",
    "\n",
    "data_orig = []\n",
    "data_union = []\n",
    "for allele1 in ordered_labels:\n",
    "    kmers1 = allele_kmers[allele1]\n",
    "    kmers1_unique_union = set()\n",
    "    \n",
    "    for allele2 in ordered_labels:\n",
    "        kmers1_unique_union = kmers1_unique_union | (kmers1 - allele_kmers[allele2])\n",
    "            \n",
    "    data_orig.append(len(kmers1))\n",
    "    data_union.append(len(kmers1_unique_union))\n",
    "\n",
    "df3 = pd.DataFrame([data_orig, data_union], columns=ordered_labels)\n",
    "df3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "457\n",
      "294\n",
      "[163, 163, 163, 163, 163, 163, 163, 163, 163, 163]\n",
      "{'ATGCAACCAGGTAGTTCTATC', 'CTGCTATTGGAACGGGAATTA', 'AATCAATTTTCTAGCTTCTAA', 'AGCAAAAGTTACAGTTACATC', 'CAGGATGAGAATTAATTCCCG', 'CAAATTGATCTCTAAAATCAA', 'CGATCAAATCCTCAGCCACAG', 'AAAAGTTACAGTTACATCTGC', 'GGGTGGACAATTACAACTTAA', 'CCCAAATTGATCTCTAAAATC', 'GTAGTTCTATCATGCCAGGTA', 'AAATTGATTTTAGAGATCAAT', 'GACCACATTTTGGACCACTGC', 'AGATGTAACTGTAACTTTTGC', 'TTACATCTGCTCCAATAACAA', 'CAAAAGTTACAGTTACATCTG', 'CTGAGGATTTGATCGAGGCGA', 'AGGGTGGACAATTACAACTTA', 'AATACTCTTTGTATATCTTCA', 'ATTGGAGCAGATGTAACTGTA', 'TATTAGAAGCTAGAAAATTGA', 'CTAGAAAATTGATTTTAGAGA', 'GACCACTGCTTAAAAGTCTTA', 'ATTTTGTTATTGGAGCAGATG', 'ATTACAACTTAATGTTTTTGA', 'TACATCTGCTCCAATAACAAA', 'ACTGCTTAAAAGTCTTAAGTC', 'GGTGAAGATATACAAAGAGTA', 'ATGATAGGTGAAGATATACAA', 'CCACTGCTTAAAAGTCTTAAG', 'CTGCTTAAAAGTCTTAAGTCA', 'CAAGCAAAAGTTACAGTTACA', 'CAACCAGGTAGTTCTATCATG', 'ACCACTGCTTAAAAGTCTTAA', 'TGATAGAACTACCTGGTTGCA', 'AAGCAAAAGTTACAGTTACAT', 'ATGATAGAACTACCTGGTTGC', 'AATTCCCGTTCCAATAGCAGT', 'ATTAATTCCCGTTCCAATAGC', 'CATTTTGGACCACTGCTTAAA', 'TGAGAATTAATTCCCGTTCCA', 'AATTGATCTCTAAAATCAATT', 'ACTGTGGCTGAGGATTTGATC', 'CATGCCAGGTAAGGTAAATCC', 'CCAGGTAGTTCTATCATGCCA', 'ACAGTTACATCTGCTCCAATA', 'GACTTTTAAGCAGTGGTCCAA', 'AATGACTTAAGACTTTTAAGC', 'ATCTCTAAAATCAATTTTCTA', 'TTACAGTTACATCTGCTCCAA', 'ACCTGGCATGATAGAACTACC', 'GATTTACCTTACCTGGCATGA', 'AAAGTTACAGTTACATCTGCT', 'GCTTGTGAGGGTGGACAATTA', 'CTATCATGCCAGGTAAGGTAA', 'CAGGTAGTTCTATCATGCCAG', 'AAGCTAGAAAATTGATTTTAG', 'ACCTTACCTGGCATGATAGAA', 'AAACATTAAGTTGTAATTGTC', 'AGTTCTATCATGCCAGGTAAG', 'AAGCAGTGGTCCAAAATGTGG', 'AGTCGCCTCGATCAAATCCTC', 'ACCACATTTTGGACCACTGCT', 'ATGAGAATTAATTCCCGTTCC', 'CTCTAAAATCAATTTTCTAGC', 'TGTGAGGGTGGACAATTACAA', 'AGACTTTTAAGCAGTGGTCCA', 'CAATTTTCTAGCTTCTAATAC', 'ACCAGGTAGTTCTATCATGCC', 'AATTACAACTTAATGTTTTTG', 'AAATTGATCTCTAAAATCAAT', 'AAAATCAATTTTCTAGCTTCT', 'ATACTGTGGCTGAGGATTTGA', 'GGATTTGATCGAGGCGACTCA', 'AACTGCTATTGGAACGGGAAT', 'ATTTGATCGAGGCGACTCAAG', 'TTACAACTTAATGTTTTTGAA', 'CCGTTCCAATAGCAGTTCCAC', 'GCAACTGGTTCAAAAACATTA', 'CACATTTTGGACCACTGCTTA', 'TAGTTCTATCATGCCAGGTAA', 'CTTGTGAGGGTGGACAATTAC', 'TATCTTGAGTCGCCTCGATCA', 'CCTCGATCAAATCCTCAGCCA', 'GATCGAGGCGACTCAAGATAC', 'AAAAACATTAAGTTGTAATTG', 'ATTTTGGACCACTGCTTAAAA', 'ATACTCTTTGTATATCTTCAC', 'AGTTACATCTGCTCCAATAAC', 'GAAGCTAGAAAATTGATTTTA', 'GATTTGATCGAGGCGACTCAA', 'TCTTGAGTCGCCTCGATCAAA', 'AACATTAAGTTGTAATTGTCC', 'AACTTAATGTTTTTGAACCAG', 'GAAAATTGATTTTAGAGATCA', 'CAGTTACATCTGCTCCAATAA', 'AACTGGTTCAAAAACATTAAG', 'GGAACTGCTATTGGAACGGGA', 'ACTTAAGACTTTTAAGCAGTG', 'AGAATTAATTCCCGTTCCAAT', 'AACCAGGTAGTTCTATCATGC', 'AATGTTTTTGAACCAGTTGCA', 'ATTAAGTTGTAATTGTCCACC', 'ATAGGTGAAGATATACAAAGA', 'AGGATTTGATCGAGGCGACTC', 'AAAACATTAAGTTGTAATTGT', 'ACATCTGCTCCAATAACAAAA', 'TACCTTACCTGGCATGATAGA', 'TGTAATTGTCCACCCTCACAA', 'CTTTTAAGCAGTGGTCCAAAA', 'AGGTAGTTCTATCATGCCAGG', 'CTCTTTGTATATCTTCACCTA', 'ACTTTTAAGCAGTGGTCCAAA', 'AAGTTGTAATTGTCCACCCTC', 'AACGGGAATTAATTCTCATCC', 'GAGAATTAATTCCCGTTCCAA', 'CCAAATTGATCTCTAAAATCA', 'CTCGATCAAATCCTCAGCCAC', 'AGAAAATTGATTTTAGAGATC', 'GTCGCCTCGATCAAATCCTCA', 'GGGAATTAATTCTCATCCTGA', 'TATCATGCCAGGTAAGGTAAA', 'GTTACATCTGCTCCAATAACA', 'AAAATTGATTTTAGAGATCAA', 'ATCAAATCCTCAGCCACAGTA', 'GTTCAAAAACATTAAGTTGTA', 'AGCTAGAAAATTGATTTTAGA', 'ACGGGAATTAATTCTCATCCT', 'ATGACTTAAGACTTTTAAGCA', 'CGCCTCGATCAAATCCTCAGC', 'ATCATGCCAGGTAAGGTAAAT', 'CAACTTAATGTTTTTGAACCA', 'ACTGGTTCAAAAACATTAAGT', 'AAATCAATTTTCTAGCTTCTA', 'CGTTCCAATAGCAGTTCCACC', 'ATCTTGAGTCGCCTCGATCAA', 'TCGATCAAATCCTCAGCCACA', 'CAAATCCTCAGCCACAGTATA', 'TGCTATTGGAACGGGAATTAA', 'CTTTGTATATCTTCACCTATC', 'AAGTTACAGTTACATCTGCTC', 'AGTTACAGTTACATCTGCTCC', 'GTTACAGTTACATCTGCTCCA', 'CCCGTTCCAATAGCAGTTCCA', 'ACAACTTAATGTTTTTGAACC', 'TGATAGGTGAAGATATACAAA', 'CAACTGGTTCAAAAACATTAA', 'ATCTGCTCCAATAACAAAATA', 'ACATTAAGTTGTAATTGTCCA', 'GCCTCGATCAAATCCTCAGCC', 'GTGAGGGTGGACAATTACAAC', 'GAACGGGAATTAATTCTCATC', 'AGACCACATTTTGGACCACTG', 'GAACTGCTATTGGAACGGGAA', 'AGTTGTAATTGTCCACCCTCA', 'CATTAAGTTGTAATTGTCCAC', 'ATCAATTTTCTAGCTTCTAAT', 'GAATTAATTCCCGTTCCAATA', 'ACATTTTGGACCACTGCTTAA', 'ACTCTTTGTATATCTTCACCT', 'AAGACTTTTAAGCAGTGGTCC', 'AATTAATTCCCGTTCCAATAG', 'CCTTACCTGGCATGATAGAAC'}\n"
     ]
    }
   ],
   "source": [
    "ordered_labels = sorted(allele_kmers.keys()) \n",
    "\n",
    "def get_unique_kmers(kmers_alleles_table):\n",
    "    allele_kmer_values = list(kmers_alleles_table.values())\n",
    "    intersection_all = allele_kmer_values.pop()\n",
    "    print(len(intersection_all))\n",
    "    for other in allele_kmer_values:\n",
    "        intersection_all = intersection_all & other\n",
    "\n",
    "    print(len(intersection_all))\n",
    "\n",
    "    data = [kmers_alleles_table[allele1] - intersection_all for allele1 in ordered_labels]\n",
    "    print(f\"{[len(d) for d in data]}\")\n",
    "    print(data[0])\n",
    "    \n",
    "get_unique_kmers(allele_kmers)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Minimal after encoding non-kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[set(), set(), set(), set(), set(), {'ACGTTAAGTTGTAATTGTCCA', 'AATTACAACTTAACGTTTTTG', '~CAACTTAATGTTTTTGAACCA', 'AACGTTAAGTTGTAATTGTCC', '~ACTGGTTCAAAAACATTAAGT', 'GGTGGACAATTACAACTTAAC', 'GCAACTGGTTCAAAAACGTTA', 'CAACTTAACGTTTTTGAACCA', 'ACAACTTAACGTTTTTGAACC', 'AAAACGTTAAGTTGTAATTGT', 'AACTTAACGTTTTTGAACCAG', 'AACTGGTTCAAAAACGTTAAG', 'AAACGTTAAGTTGTAATTGTC', '~AACTTAATGTTTTTGAACCAG', '~CAACTGGTTCAAAAACATTAA', 'AAAAACGTTAAGTTGTAATTG', 'AACGTTTTTGAACCAGTTGCA', 'CGTTAAGTTGTAATTGTCCAC', 'TTACAACTTAACGTTTTTGAA', '~AACTGGTTCAAAAACATTAAG', 'CAACTGGTTCAAAAACGTTAA', 'ACTGGTTCAAAAACGTTAAGT', 'ATTACAACTTAACGTTTTTGA', 'GTTCAAAAACGTTAAGTTGTA'}, set(), set(), set(), set()]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>aspA_1</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_10</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_2</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_3</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_4</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_5</th>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_6</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_7</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_8</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_9</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          0\n",
       "aspA_1    0\n",
       "aspA_10   0\n",
       "aspA_2    0\n",
       "aspA_3    0\n",
       "aspA_4    0\n",
       "aspA_5   24\n",
       "aspA_6    0\n",
       "aspA_7    0\n",
       "aspA_8    0\n",
       "aspA_9    0"
      ]
     },
     "execution_count": 150,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def insert_non_kmers(kmers_alleles_table):\n",
    "    allele_non_kmers = {}\n",
    "    \n",
    "    union_all = set()\n",
    "    for allele in kmers_alleles_table:\n",
    "        union_all = union_all | kmers_alleles_table[allele]\n",
    "\n",
    "    for allele in kmers_alleles_table:\n",
    "        non_kmers = union_all - kmers_alleles_table[allele]\n",
    "        allele_non_set = kmers_alleles_table[allele].copy()\n",
    "        for k in non_kmers:\n",
    "            allele_non_set.add(f\"~{k}\")\n",
    "\n",
    "        allele_non_kmers[allele] = allele_non_set\n",
    "        \n",
    "    return allele_non_kmers\n",
    "\n",
    "def get_unique_kmers(kmers_alleles_table):\n",
    "    ordered_labels = sorted(kmers_alleles_table.keys())\n",
    "    \n",
    "    data = []\n",
    "    for allele1 in ordered_labels:\n",
    "        unique_kmers1 = kmers_alleles_table[allele1]\n",
    "\n",
    "        for allele2 in ordered_labels:\n",
    "            if allele2 != allele1 and len(unique_kmers1) > 0:\n",
    "                unique_kmers1 = unique_kmers1 - kmers_alleles_table[allele2]\n",
    "\n",
    "        data.append(unique_kmers1)\n",
    "        \n",
    "    return (data, ordered_labels)\n",
    "    \n",
    "allele_non_kmers = insert_non_kmers(allele_kmers)\n",
    "#print(allele_non_kmers['aspA_1'] - allele_non_kmers['aspA_2'])\n",
    "(data, labels) = get_unique_kmers(allele_non_kmers)\n",
    "print(data)\n",
    "data = [len(d) for d in data]\n",
    "\n",
    "df4 = pd.DataFrame(data, index=labels)\n",
    "df4"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Jaccard distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>aspA_1</th>\n",
       "      <th>aspA_10</th>\n",
       "      <th>aspA_2</th>\n",
       "      <th>aspA_3</th>\n",
       "      <th>aspA_4</th>\n",
       "      <th>aspA_5</th>\n",
       "      <th>aspA_6</th>\n",
       "      <th>aspA_7</th>\n",
       "      <th>aspA_8</th>\n",
       "      <th>aspA_9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>aspA_1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.123203</td>\n",
       "      <td>0.316759</td>\n",
       "      <td>0.087866</td>\n",
       "      <td>0.200787</td>\n",
       "      <td>0.265655</td>\n",
       "      <td>0.249042</td>\n",
       "      <td>0.249042</td>\n",
       "      <td>0.175649</td>\n",
       "      <td>0.249042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_10</th>\n",
       "      <td>0.123203</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.404887</td>\n",
       "      <td>0.200787</td>\n",
       "      <td>0.087866</td>\n",
       "      <td>0.160966</td>\n",
       "      <td>0.344203</td>\n",
       "      <td>0.344203</td>\n",
       "      <td>0.278719</td>\n",
       "      <td>0.344203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_2</th>\n",
       "      <td>0.316759</td>\n",
       "      <td>0.404887</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.249042</td>\n",
       "      <td>0.461279</td>\n",
       "      <td>0.504092</td>\n",
       "      <td>0.242308</td>\n",
       "      <td>0.087866</td>\n",
       "      <td>0.168337</td>\n",
       "      <td>0.242308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_3</th>\n",
       "      <td>0.087866</td>\n",
       "      <td>0.200787</td>\n",
       "      <td>0.249042</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.272212</td>\n",
       "      <td>0.332117</td>\n",
       "      <td>0.175649</td>\n",
       "      <td>0.175649</td>\n",
       "      <td>0.249042</td>\n",
       "      <td>0.316759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_4</th>\n",
       "      <td>0.200787</td>\n",
       "      <td>0.087866</td>\n",
       "      <td>0.461279</td>\n",
       "      <td>0.272212</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.079832</td>\n",
       "      <td>0.404887</td>\n",
       "      <td>0.404887</td>\n",
       "      <td>0.344203</td>\n",
       "      <td>0.404887</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_5</th>\n",
       "      <td>0.265655</td>\n",
       "      <td>0.160966</td>\n",
       "      <td>0.504092</td>\n",
       "      <td>0.332117</td>\n",
       "      <td>0.079832</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.418685</td>\n",
       "      <td>0.450847</td>\n",
       "      <td>0.393673</td>\n",
       "      <td>0.418685</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_6</th>\n",
       "      <td>0.249042</td>\n",
       "      <td>0.344203</td>\n",
       "      <td>0.242308</td>\n",
       "      <td>0.175649</td>\n",
       "      <td>0.404887</td>\n",
       "      <td>0.418685</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.168337</td>\n",
       "      <td>0.242308</td>\n",
       "      <td>0.168337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_7</th>\n",
       "      <td>0.249042</td>\n",
       "      <td>0.344203</td>\n",
       "      <td>0.087866</td>\n",
       "      <td>0.175649</td>\n",
       "      <td>0.404887</td>\n",
       "      <td>0.450847</td>\n",
       "      <td>0.168337</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.242308</td>\n",
       "      <td>0.310536</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_8</th>\n",
       "      <td>0.175649</td>\n",
       "      <td>0.278719</td>\n",
       "      <td>0.168337</td>\n",
       "      <td>0.249042</td>\n",
       "      <td>0.344203</td>\n",
       "      <td>0.393673</td>\n",
       "      <td>0.242308</td>\n",
       "      <td>0.242308</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.087866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_9</th>\n",
       "      <td>0.249042</td>\n",
       "      <td>0.344203</td>\n",
       "      <td>0.242308</td>\n",
       "      <td>0.316759</td>\n",
       "      <td>0.404887</td>\n",
       "      <td>0.418685</td>\n",
       "      <td>0.168337</td>\n",
       "      <td>0.310536</td>\n",
       "      <td>0.087866</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           aspA_1   aspA_10    aspA_2    aspA_3    aspA_4    aspA_5    aspA_6  \\\n",
       "aspA_1   0.000000  0.123203  0.316759  0.087866  0.200787  0.265655  0.249042   \n",
       "aspA_10  0.123203  0.000000  0.404887  0.200787  0.087866  0.160966  0.344203   \n",
       "aspA_2   0.316759  0.404887  0.000000  0.249042  0.461279  0.504092  0.242308   \n",
       "aspA_3   0.087866  0.200787  0.249042  0.000000  0.272212  0.332117  0.175649   \n",
       "aspA_4   0.200787  0.087866  0.461279  0.272212  0.000000  0.079832  0.404887   \n",
       "aspA_5   0.265655  0.160966  0.504092  0.332117  0.079832  0.000000  0.418685   \n",
       "aspA_6   0.249042  0.344203  0.242308  0.175649  0.404887  0.418685  0.000000   \n",
       "aspA_7   0.249042  0.344203  0.087866  0.175649  0.404887  0.450847  0.168337   \n",
       "aspA_8   0.175649  0.278719  0.168337  0.249042  0.344203  0.393673  0.242308   \n",
       "aspA_9   0.249042  0.344203  0.242308  0.316759  0.404887  0.418685  0.168337   \n",
       "\n",
       "           aspA_7    aspA_8    aspA_9  \n",
       "aspA_1   0.249042  0.175649  0.249042  \n",
       "aspA_10  0.344203  0.278719  0.344203  \n",
       "aspA_2   0.087866  0.168337  0.242308  \n",
       "aspA_3   0.175649  0.249042  0.316759  \n",
       "aspA_4   0.404887  0.344203  0.404887  \n",
       "aspA_5   0.450847  0.393673  0.418685  \n",
       "aspA_6   0.168337  0.242308  0.168337  \n",
       "aspA_7   0.000000  0.242308  0.310536  \n",
       "aspA_8   0.242308  0.000000  0.087866  \n",
       "aspA_9   0.310536  0.087866  0.000000  "
      ]
     },
     "execution_count": 154,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ordered_labels = sorted(allele_kmers.keys())\n",
    "\n",
    "data = []\n",
    "for allele1 in ordered_labels:\n",
    "    row = [1 - len(allele_kmers[allele1] & allele_kmers[allele2])/len(allele_kmers[allele1] | allele_kmers[allele2]) for allele2 in ordered_labels]\n",
    "            \n",
    "    data.append(row)\n",
    "    \n",
    "df5 = pd.DataFrame(data, columns=ordered_labels, index=ordered_labels)\n",
    "df5"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Kmer table as colored de Bruijn graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AAAAACATTAAGTTGCAATTG\t{'aspA_9', 'aspA_6'}\n",
      "AAAAACATTAAGTTGTAATTG\t{'aspA_2', 'aspA_3', 'aspA_10', 'aspA_4', 'aspA_7', 'aspA_1', 'aspA_8'}\n",
      "AAAAACGTTAAGTTGTAATTG\t{'aspA_5'}\n",
      "AAAAATAAGAGAAGTGACAGG\t{'aspA_2', 'aspA_3', 'aspA_10', 'aspA_6', 'aspA_4', 'aspA_7', 'aspA_1', 'aspA_9', 'aspA_8', 'aspA_5'}\n",
      "AAAAATGCAACCAGGTAGTTC\t{'aspA_2', 'aspA_3', 'aspA_10', 'aspA_6', 'aspA_4', 'aspA_7', 'aspA_1', 'aspA_9', 'aspA_8', 'aspA_5'}\n",
      "AAAACACCTGAAATTTGTACA\t{'aspA_2', 'aspA_3', 'aspA_10', 'aspA_6', 'aspA_4', 'aspA_7', 'aspA_1', 'aspA_9', 'aspA_8', 'aspA_5'}\n",
      "AAAACATTAAGTTGCAATTGT\t{'aspA_9', 'aspA_6'}\n",
      "AAAACATTAAGTTGTAATTGT\t{'aspA_2', 'aspA_3', 'aspA_10', 'aspA_4', 'aspA_7', 'aspA_1', 'aspA_8'}\n"
     ]
    }
   ],
   "source": [
    "for kmer in sorted(kmer_table.keys())[0:8]:\n",
    "    print(f\"{kmer}\\t{kmer_table[kmer]}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To treat this table as a colored de Bruijn graph, we take some initial k-mer from some genome, shift it over by 1, and look to see if this new k-mer exists in the table. We try every possible nucleotide `{'A', 'T', 'C', 'G'}`. If any exist, then this represents an edge in the de Bruijn graph.\n",
    "\n",
    "Let's first look at the sequence for some allele and pull out the first kmer."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "aspA_1 ATGATAGGTGAAGATATACAAAGAGTATTAGAAGCTAGAAAATTGATTTTAGAGATCAATTTGGGTGGAACTGCTATTGGAACGGGAATTAATTCTCATCCTGATTATCCGAAGGTTGTAGAAAGAAAAATAAGAGAAGTGACAGGTTTTGAATATACTGTGGCTGAGGATTTGATCGAGGCGACTCAAGATACGGGAGCTTATGTACAAATTTCAGGTGTTTTAAAACGTGTTGCAACAAAACTTTCTAAAGTATGTAATGACTTAAGACTTTTAAGCAGTGGTCCAAAATGTGGTCTTAATGAGATTAATCTTCCAAAAATGCAACCAGGTAGTTCTATCATGCCAGGTAAGGTAAATCCTGTTATTCCTGAAGTAGTTAATCAAGTTTGTTATTTTGTTATTGGAGCAGATGTAACTGTAACTTTTGCTTGTGAGGGTGGACAATTACAACTTAATGTTTTTGAACCAGTTGCA\n",
      "start kmer: ATGATAGGTGAAGATATACAA\n"
     ]
    }
   ],
   "source": [
    "original_seq_aspA_1 = ''\n",
    "with open('aspA-1.fas', 'r') as f:\n",
    "    for record in SeqIO.parse(f, 'fasta'):\n",
    "        print(f\"{record.id} {record.seq}\")\n",
    "        print(f\"start kmer: {record.seq[0:21]}\")\n",
    "        original_seq_aspA_1 = record.seq\n",
    "        break"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, let's start from this kmer and walk the de Bruijn graph to recover the original sequence (we ignore possibility of multiple paths)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATGATAGGTGAAGATATACAAAGAGTATTAGAAGCTAGAAAATTGATTTTAGAGATCAATTTGGGTGGAACTGCTATTGGAACGGGAATTAATTCTCATCCTGATTATCCGAAGGTTGTAGAAAGAAAAATAAGAGAAGTGACAGGTTTTGAATATACTGTGGCTGAGGATTTGATCGAGGCGACTCAAGATACGGGAGCTTATGTACAAATTTCAGGTGTTTTAAAACGTGTTGCAACAAAACTTTCTAAAGTATGTAATGACTTAAGACTTTTAAGCAGTGGTCCAAAATGTGGTCTTAATGAGATTAATCTTCCAAAAATGCAACCAGGTAGTTCTATCATGCCAGGTAAGGTAAATCCTGTTATTCCTGAAGTAGTTAATCAAGTTTGTTATTTTGTTATTGGAGCAGATGTAACTGTAACTTTTGCTTGTGAGGGTGGACAATTACAACTTAATGTTTTTGAACCAGTTGCA\n",
      "Recovered seq equal original seq? True\n"
     ]
    }
   ],
   "source": [
    "start_kmer = 'ATGATAGGTGAAGATATACAA'\n",
    "start_genome = 'aspA_1'\n",
    "\n",
    "def shift_kmer(kmer):\n",
    "    shift_kmer = kmer[1:]    \n",
    "    return set([shift_kmer + 'A', shift_kmer + 'T', shift_kmer + 'C', shift_kmer + 'G'])\n",
    "\n",
    "def shift_nucleotide_recursive(kmer):\n",
    "    for shifted_nucleotide in ['A', 'T', 'C', 'G']:\n",
    "        shifted_kmer = kmer[1:] + shifted_nucleotide\n",
    "        shifted_kmer_rc = reverse_complement(shifted_kmer)\n",
    "        if (shifted_kmer in kmer_table and start_genome in kmer_table[shifted_kmer]) or (shifted_kmer_rc in kmer_table and start_genome in kmer_table[shifted_kmer_rc]):\n",
    "            return shifted_nucleotide + shift_nucleotide_recursive(shifted_kmer)\n",
    "    return ''\n",
    "            \n",
    "recovered_seq = start_kmer + shift_nucleotide_recursive(start_kmer)\n",
    "print(recovered_seq)\n",
    "print(f\"Recovered seq equal original seq? {recovered_seq == original_seq_aspA_1}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
