{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "PROJECT_DIR=`git rev-parse --show-toplevel`\n",
    "cd $PROJECT_DIR\n",
    "\n",
    "threads=16"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Examine input files\n",
    "\n",
    "Let's take a look at the input files.\n",
    "\n",
    "## CARD protein homolog database\n",
    "\n",
    "The CARD protein homolog database looks like."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">gb|GQ343019|+|132-1023|ARO:3002999|CblA-1 [mixed culture bacterium AX_gF3SD01_15] \n",
      "ATGAAAGCATATTTCATCGCCATACTTACCTTATTCACTTGTATAGCTACCGTCGTCCGGGCGCAGCAAATGTCTGAACTTGAAAACCGGATTGACAGTCTGCTCAATGGCAAGAAAGCCACCGTTGGTATAGCCGTATGGACAGACAAAGGAGACATGCTCCGGTATAACGACCATGTACACTTCCCCTTGCTCAGTGTATTCAAATTCCATGTGGCACTGGCCGTACTGGACAAGATGGATAAGCAAAGCATCAGTCTGGACAGCATTGTTTCCATAAAGGCATCCCAAATGCCGCCCAATACCTACAGCCCCCTGCGGAAGAAGTTTCCCGACCAGGATTTCACGATTACGCTTAGGGAACTGATGCAATACAGCATTTCCCAAAGCGACAACAATGCCTGCGACATCTTGATAGAATATGCAGGAGGCATCAAACATATCAACGACTATATCCACCGGTTGAGTATCGACTCCTTCAACCTCTCGGAAACAGAAGACGGCATGCACTCCAGCTTCGAGGCTGTATACCGCAACTGGAGTACTCCTTCCGCTATGGTCCGACTACTGAGAACGGCTGATGAAAAAGAGTTGTTCTCCAACAAGGAGCTGAAAGACTTCTTGTGGCAGACCATGATAGATACTGAAACCGGTGCCAACAAACTGAAAGGTATGTTGCCAGCCAAAACCGTGGTAGGACACAAGACCGGCTCTTCCGACCGCAATGCCGACGGTATGAAAACTGCAGATAATGATGCCGGCCTCGTTATCCTTCCCGACGGCCGGAAATACTACATTGCCGCCTTCGTCATGGACTCATACGAGACGGATGAGGACAATGCGAACATCATCGCCCGCATATCACGCATGGTATATGATGCGATGAGATGA\n",
      ">gb|HQ845196|+|0-861|ARO:3001109|SHV-52 [Klebsiella pneumoniae] \n",
      "ATGCGTTATATTCGCCTGTGTATTATCTCCCTGTTAGCCGCCCTGCCGCTGGCGGTACACGCCAGCCCGCAGCCGCTTGAGCAAATTAAACAAAGCGAAAGCCAGCTGTCGGGCCGCGTAGGCATGATAGAAATGGATCTGGCCAGCGGCCGCACGCTGACCGCCTGGCGCGCCGATGAACGCTTTCCCATGATAAGCACCTTTAAAGTAGTGCTCTGCGGCGCAGTGCTGGCGCGGGTGGATGCCGGTGACGAACAGCTGGAGCGAAAGATCCACTATCGCCAGCAGGATCTGGTGGACTACTCGCCGGTCAGCGAAAAACACCTTGCCGACGGCATGACGGTCGGCGAACTCTGTGCCGCCGCCATTACCATGAGCGATAACAGCGCCGCCAATTTGCTGCTGGCCATCGTCGGCGGCCCCGCAGGATTGACTGCCTTTTTGCGCCAGATCGGCGACAACGTCACCCGCCTTGACCGCTGGGAAACGGAACTGAATGAGGCGCTTCCCGGCGACGCCCGCGACACCACTACCCCGGCCAGCATGGCCGCGACCCTGCGCAAGCTGCTGACCAGCCAGCGTCTGAGCGCCCGTTCGCAACGGCAGCTGCTGCAGTGGATGGTGGACGATCGGGTCGCCGGACCGTTGATCCGCTCCGTGCTGCCGGCGGGCTGGTTTATCGCCGATAAGACCGGAGCTGGCGAACGGGGTGCGCGCGGGATTGTCGCCCTGCTTGGCCCGAATAACAAAGCAGAGCGCATTGTAGTGATTTATCTGCGGGATACCCCGGCGAGCATGGCCGAGCGAAATCAGCAAATCGCCGGGATCGGCGCGGCGCTGATCGAGCACTGGCAACGCTAA\n",
      ">gb|AF028812|+|392-887|ARO:3002867|dfrF [Enterococcus faecalis] \n",
      "ATGATAGGTTTGATTGTTGCGAGGTCAAAGAATAATGTTATAGGCAAGAATGGTAATATACCATGGAAAATAAAGGGAGAACAAAAGCAATTTAGAGAGTTAACAACGGGTAATGTGGTTATTATGGGGCGAAAGTCTTATGAAGAAATCGGTCATCCGTTGCCTAATAGAATGAATATTGTTGTTTCCACCACAACAGAGTATCAAGGAGATAATTTAGTTTCAGTTAAATCATTAGAAGATGCATTATTATTGGCTAAAGGACGAGATGTATACATATCTGGTGGATATGGACTATTTAAGGAAGCTTTGCAAATAGTAGATAAAATGTATATCACAGAAGTAGATTTAAATATTGAAGATGGAGATACATTCTTTCCAGAATTTGATATCAATGATTTTGAAGTTTTGATAGGGGAAACACTTGGTGAGGAAGTGAAATATACGAGAACATTTTATGTAAGGAAAAATGAATTGAGTAGATTTTGGATTTAG\n",
      ">gb|JX017365|+|244-1120|ARO:3001989|CTX-M-130 [Escherichia coli] \n",
      "ATGGTGACAAAGAGAGTGCAACGGATGATGTTCGCGGCGGCGGCGTGCATTCCGCTGCTGCTGGGCAGCGCGCCGCTTTATGCGCAGACGAGTGCGGTGCAGCAAAAGCTGGCGGCGCTGGAGAAAAGCAGCGGAGGGCGGCTGGGCGTCGCGCTCATCGATACCGCAGATAATACGCAGGTGCTTTATCGCGGTGATGAACGCTTTCCAATGTGCAGTACCAGTAAAGTTATGGCGGCCGCGGCGGTGCTTAAGCAGAGTGAAACGCAAAAGCAGCTGCTTAATCAGCCTGTCGAGATCAAGCCTGCCGATCTGGTTAACTACAATCCGATTGCCGAAAAACACGTCAACGGCACAATGACGCTGGCAGAACTGAGCGCGGCCGCGTTGCAGTACAGCGACAATACCGCCATGAACAAATTGATTGCCCAGCTCGGTGGCCCGGGAGGCGTGACGGCTTTTGCCCGCGCGATCGGCGATGAGACGTTTCGTCTGGATCGCACTGAACCTACGCTGAATACCGCCATTCCCGGCGACCCGAGAGACACCACCACGCCGCGGGCGATGGCGCAGACGTTGCGTCAGCTTACGCTGGGTCATGCGCTGGGCGAAACCCAGCGGGCGCAGTTGGTGACGTGGCTCAAAGGCAATACGACCGGCGCAGCCAGCATTCGGGCCGGCTTACCGACGTCGTGGACTGTGGGTGATAAGACCGGCAGCGGCGACTACGGCACCACCAATGATATTGCGGTGATCTGGCCGCAGGGTCGCGCGCCGCTGGTTCTGGTGACCTATTTTACCCAGCCGCAACAGAACGCAGAGAGACGCCACGATGTGCTGGCTTCAGCGGCGAGAATCATCGCCGAAGGGCTGTAA\n",
      ">gb|JN967644|+|0-813|ARO:3002356|NDM-6 [Escherichia coli] \n",
      "ATGGAATTGCCCAATATTATGCACCCGGTCGCGAAGCTGAGCACCGCATTAGCCGCTGCATTGATGCTGAGCGGGTGCATGCCCGGTGAAATCCGCCCGACGATTGGCCAGCAAATGGAAACTGGCGACCAACGGTTTGGCGATCTGGTTTTCCGCCAGCTCGCACCGAATGTCTGGCAGCACACTTCCTATCTCGACATGCCGGGTTTCGGGGCAGTCGCTTCCAACGGTTTGATCGTCAGGGATGGCGGCCGCGTGCTGGTGGTCGATACCGCCTGGACCGATGACCAGACCGCCCAGATCCTCAACTGGATCAAGCAGGAGATCAACCTGCCGGTCGCGCTGGCGGTGGTGACTCACGCGCATCAGGACAAGATGGGCGGTATGGACGCGCTGCATGCGGCGGGGATTGCGACTTATGCCAATGCGTTGTCGAACCAGCTTGCCCCGCAAGAGGGGATGGTTGCGGCGCAACACAGCCTGACTTTCGCCGCCAATGGCTGGGTCGAACCAGCAACCGCGCCCAACTTTGGCCCGCTCAAGGTATTTTACCCCGGCCCCGGCCACACCAGTGACAATATCACCGTTGGGATCGACGGCACCGACATCGCTTTTGGTGGCTGCCTGATCAAGGACAGCAAGGCCAAGTCGCTCGGCAATCTCGGTGATGCCGACACTGAGCACTACGCCGCGTCAGTGCGCGCGTTTGGTGCGGCGTTCCCCAAGGCCAGCATGATCGTGATGAGCCATTCCGCCCCCGATAGCCGCGCCGCAATCACTCATACGGCCCGCATGGCCGACAAGCTGCGCTGA\n"
     ]
    }
   ],
   "source": [
    "head db/nucleotide_fasta_protein_homolog_model.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sequence reads\n",
    "\n",
    "The sequence reads look like:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rwxr-xr-x 1 apetkau grp_apetkau 474M Jan 15 22:42 \u001b[0m\u001b[01;32minput/SRR1952908_1.fastq\u001b[0m\n",
      "-rwxr-xr-x 1 apetkau grp_apetkau 471M Jan 15 22:42 \u001b[01;32minput/SRR1952908_2.fastq\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "ls -lrth input/*.fastq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Scenario 1: Assemble genome with SKESA, run RGI\n",
    "\n",
    "Let's test out assembling the genome with SKESA and running RGI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "conda run --name skesa /usr/bin/time -v skesa --cores 1 \\\n",
    "    --fastq input/SRR1952908_1.fastq,input/SRR1952908_2.fastq \\\n",
    "    --contigs_out SRR1952908-full-assembly.fasta --vector_percent 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Great. Let's look at the assembled genome and run RGI on it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rw-r--r-- 1 apetkau grp_apetkau 4.6M Jan 16 17:14 SRR1952908-full-assembly.fasta\n"
     ]
    }
   ],
   "source": [
    "ls -lrth SRR1952908-full-assembly.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘rgi-full-assembly-out’: File exists\n",
      "~/workspace/kmer-analysis-test/rgi-full-assembly-out ~/workspace/kmer-analysis-test\n",
      "\tCommand being timed: \"rgi main -i ../SRR1952908-full-assembly.fasta -o out -n 1\"\n",
      "\tUser time (seconds): 241.83\n",
      "\tSystem time (seconds): 4.98\n",
      "\tPercent of CPU this job got: 109%\n",
      "\tElapsed (wall clock) time (h:mm:ss or m:ss): 3:45.30\n",
      "\tAverage shared text size (kbytes): 0\n",
      "\tAverage unshared data size (kbytes): 0\n",
      "\tAverage stack size (kbytes): 0\n",
      "\tAverage total size (kbytes): 0\n",
      "\tMaximum resident set size (kbytes): 134160\n",
      "\tAverage resident set size (kbytes): 0\n",
      "\tMajor (requiring I/O) page faults: 0\n",
      "\tMinor (reclaiming a frame) page faults: 253002\n",
      "\tVoluntary context switches: 286\n",
      "\tInvoluntary context switches: 158828\n",
      "\tSwaps: 0\n",
      "\tFile system inputs: 0\n",
      "\tFile system outputs: 206040\n",
      "\tSocket messages sent: 0\n",
      "\tSocket messages received: 0\n",
      "\tSignals delivered: 0\n",
      "\tPage size (bytes): 4096\n",
      "\tExit status: 0\n",
      "~/workspace/kmer-analysis-test\n"
     ]
    }
   ],
   "source": [
    "mkdir rgi-full-assembly-out\n",
    "pushd rgi-full-assembly-out\n",
    "conda run --name rgi-4.2.2 /usr/bin/time -v rgi main -i ../SRR1952908-full-assembly.fasta -o out -n 1\n",
    "popd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Best_Hit_ARO\n",
      "ugd\n",
      "basS\n",
      "MdtK\n",
      "acrD\n",
      "PmrF\n",
      "TEM-57\n",
      "tet(C)\n",
      "kdpE\n",
      "patA\n",
      "bacA\n",
      "Escherichia coli acrA\n",
      "acrB\n",
      "golS\n",
      "mdsA\n",
      "adeF\n",
      "mdsC\n",
      "cpxA\n",
      "emrR\n",
      "emrA\n",
      "emrB\n",
      "CRP\n",
      "mdtC\n",
      "sdiA\n",
      "AAC(6')-Iaa\n",
      "marA\n",
      "OKP-B-10\n",
      "mdtM\n",
      "Escherichia coli mdfA\n",
      "msbA\n",
      "mef(B)\n",
      "sul3\n",
      "qacH\n",
      "aadA13\n",
      "Salmonella enterica cmlA\n",
      "aadA\n",
      "aadA9\n",
      "mdtB\n",
      "mdtC\n",
      "baeR\n",
      "Escherichia coli GlpT with mutation conferring resistance to fosfomycin\n",
      "Escherichia coli UhpT with mutation conferring resistance to fosfomycin\n",
      "Escherichia coli EF-Tu mutants conferring resistance to Pulvomycin\n",
      "Haemophilus influenzae PBP3 conferring resistance to beta-lactam antibiotics\n",
      "Escherichia coli acrR with mutation conferring multidrug antibiotic resistance\n",
      "Escherichia coli marR mutant conferring antibiotic resistance\n",
      "Escherichia coli soxS with mutation conferring antibiotic resistance\n",
      "Escherichia coli soxR with mutation conferring antibiotic resistance\n"
     ]
    }
   ],
   "source": [
    "cut -f 9 rgi-full-assembly-out/out.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Scenario 2: Prefilter reads with kat, assemble with SKESA, run RGI\n",
    "\n",
    "## 2.1: Prefilter reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Kmer Analysis Toolkit (KAT) V2.4.2\n",
      "\n",
      "Running KAT in filter sequence mode\n",
      "-----------------------------------\n",
      "\n",
      "Input 50320 is a sequence file.  Counting kmers for input 50320 (db/nucleotide_fasta_protein_homolog_model.fasta) ... done.  Time taken: 0.6s\n",
      "\n",
      "Filtering sequences ...\n",
      "Processed 100000 pairs\n",
      "Processed 200000 pairs\n",
      "Processed 300000 pairs\n",
      "Processed 400000 pairs\n",
      "Processed 500000 pairs\n",
      "Processed 600000 pairs\n",
      "Processed 700000 pairs\n",
      "Processed 800000 pairs\n",
      "Processed 900000 pairs\n",
      "Processed 1000000 pairs\n",
      "Processed 1100000 pairs\n",
      "Processed 1200000 pairs\n",
      "Processed 1300000 pairs\n",
      "Processed 1400000 pairs\n",
      "Processed 1500000 pairs\n",
      "Processed 1600000 pairs\n",
      "Processed 1700000 pairs\n",
      "Processed 1800000 pairs\n",
      "Processed 1900000 pairs\n",
      "Processed 2000000 pairs\n",
      "Processed 2100000 pairs\n",
      "Finished filtering.  Time taken: 109.4s\n",
      "\n",
      "Found 10204 / 2147046 to keep\n",
      "\n",
      "KAT filter seq completed.\n",
      "Total runtime: 110.0s\n",
      "\n",
      "\tCommand being timed: \"kat filter seq -t 1 -o SRR1952908.kat -m 31 --seq input/SRR1952908_1.fastq --seq2 input/SRR1952908_2.fastq db/nucleotide_fasta_protein_homolog_model.fasta\"\n",
      "\tUser time (seconds): 108.71\n",
      "\tSystem time (seconds): 1.70\n",
      "\tPercent of CPU this job got: 100%\n",
      "\tElapsed (wall clock) time (h:mm:ss or m:ss): 1:50.04\n",
      "\tAverage shared text size (kbytes): 0\n",
      "\tAverage unshared data size (kbytes): 0\n",
      "\tAverage stack size (kbytes): 0\n",
      "\tAverage total size (kbytes): 0\n",
      "\tMaximum resident set size (kbytes): 843704\n",
      "\tAverage resident set size (kbytes): 0\n",
      "\tMajor (requiring I/O) page faults: 0\n",
      "\tMinor (reclaiming a frame) page faults: 314859\n",
      "\tVoluntary context switches: 6\n",
      "\tInvoluntary context switches: 141\n",
      "\tSwaps: 0\n",
      "\tFile system inputs: 0\n",
      "\tFile system outputs: 9472\n",
      "\tSocket messages sent: 0\n",
      "\tSocket messages received: 0\n",
      "\tSignals delivered: 0\n",
      "\tPage size (bytes): 4096\n",
      "\tExit status: 0\n"
     ]
    }
   ],
   "source": [
    "conda run --name kat /usr/bin/time -v kat filter seq -t 1 -o SRR1952908.kat -m 31 \\\n",
    "    --seq input/SRR1952908_1.fastq --seq2 input/SRR1952908_2.fastq \\\n",
    "    db/nucleotide_fasta_protein_homolog_model.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rw-r--r-- 1 apetkau grp_apetkau 2.3M Jan 16 17:30 SRR1952908.kat.in.R2.fastq\n",
      "-rw-r--r-- 1 apetkau grp_apetkau 2.3M Jan 16 17:30 SRR1952908.kat.in.R1.fastq\n"
     ]
    }
   ],
   "source": [
    "ls -lrth *kat*.fastq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.2 Assemble extracted reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "skesa --cores 1 --fastq SRR1952908.kat.in.R1.fastq,SRR1952908.kat.in.R2.fastq --contigs_out SRR1952908.kat.contigs.fasta --vector_percent 1 \n",
      "\n",
      "Total mates: 20408 Paired reads: 10204\n",
      "Reads acquired in  0.063515s wall, 0.050000s user + 0.000000s system = 0.050000s CPU (78.7%)\n",
      "Adapters clip is disabled\n",
      "\n",
      "Kmer len: 21\n",
      "Raw kmers: 1607679 Memory needed (GB): 0.0308674 Memory available (GB): 29.9988 1 cycle(s) will be performed\n",
      "Distinct kmers: 33311\n",
      "Kmer count in  0.421162s wall, 0.400000s user + 0.020000s system = 0.420000s CPU (99.7%)\n",
      "Uniq kmers merging in  0.004130s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (242.2%)\n",
      "WARNING: --min_count changed from 2 to 4 because of high coverage for genome size 9698\n",
      "WARNING: --max_kmer_count 10 to 21 because of high coverage for genome size 9698\n",
      "Kmers branching in  0.044778s wall, 0.040000s user + 0.010000s system = 0.050000s CPU (111.7%)\n",
      "\n",
      "Average read length: 99\n",
      "Genome size estimate: 9918\n",
      "\n",
      "Kmer: 21 Graph size: 25504 Contigs in: 0\n",
      "Valley: 19\n",
      "\n",
      "Mark used kmers in  0.000007s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmers in multiple/single contigs: 0 0\n",
      "Fragments before: 33 22459\n",
      "Fragments after: 33 22459\n",
      "New seeds: 29\n",
      "New seeds in  0.043043s wall, 0.050000s user + 0.000000s system = 0.050000s CPU (116.2%)\n",
      "Fragments before: 58 1218\n",
      "Fragments after: 58 1218\n",
      "Connectors: 0 Extenders: 58\n",
      "Connections and extensions in  0.011847s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (84.4%)\n",
      "Contigs out: 29 Genome: 21281 N50: 1385 L50: 4\n",
      "Assembled in  0.055422s wall, 0.060000s user + 0.000000s system = 0.060000s CPU (108.3%)\n",
      "\n",
      "\n",
      "Kmer len: 59\n",
      "Raw kmers: 832819 Memory needed (GB): 0.0239852 Memory available (GB): 29.9983 1 cycle(s) will be performed\n",
      "Distinct kmers: 22029\n",
      "Kmer count in  0.305944s wall, 0.290000s user + 0.010000s system = 0.300000s CPU (98.1%)\n",
      "Uniq kmers merging in  0.002918s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "\n",
      "Average count: 41.2572 Max kmer: 59\n",
      "Connected: 7966 ambiguously connected: 1804 from 10204 mate pairs\n",
      "Connect pairs in  12.401576s wall, 12.500000s user + 0.120000s system = 12.620000s CPU (101.8%)\n",
      "\n",
      "N50 for inserts: 261\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 7368\n",
      "Reads for connection: 5364\n",
      "Internal reads: 548\n",
      "Reads cleaned in  0.149493s wall, 0.150000s user + 0.000000s system = 0.150000s CPU (100.3%)\n",
      "\n",
      "Kmer len: 25\n",
      "Raw kmers: 553307 Memory needed (GB): 0.0106235 Memory available (GB): 29.999 1 cycle(s) will be performed\n",
      "Distinct kmers: 16730\n",
      "Kmer count in  0.178685s wall, 0.180000s user + 0.010000s system = 0.190000s CPU (106.3%)\n",
      "Uniq kmers merging in  0.002294s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmers branching in  0.036637s wall, 0.040000s user + 0.000000s system = 0.040000s CPU (109.2%)\n",
      "Kmer: 25 Graph size: 16730 Contigs in: 29\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.009038s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (110.6%)\n",
      "Kmers in multiple/single contigs: 0 12551\n",
      "Fragments before: 79 3281\n",
      "Fragments after: 79 3281\n",
      "New seeds: 2\n",
      "New seeds in  0.024604s wall, 0.030000s user + 0.000000s system = 0.030000s CPU (121.9%)\n",
      "Fragments before: 62 1581\n",
      "Fragments after: 62 1581\n",
      "Connectors: 0 Extenders: 62\n",
      "Connections and extensions in  0.027103s wall, 0.020000s user + 0.000000s system = 0.020000s CPU (73.8%)\n",
      "Contigs out: 31 Genome: 21471 N50: 1385 L50: 4\n",
      "Assembled in  0.061150s wall, 0.060000s user + 0.000000s system = 0.060000s CPU (98.1%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 7368\n",
      "Reads for connection: 5364\n",
      "Internal reads: 548\n",
      "Reads cleaned in  0.140535s wall, 0.150000s user + 0.000000s system = 0.150000s CPU (106.7%)\n",
      "\n",
      "Kmer len: 29\n",
      "Raw kmers: 523835 Memory needed (GB): 0.0100576 Memory available (GB): 29.9987 1 cycle(s) will be performed\n",
      "Distinct kmers: 16358\n",
      "Kmer count in  0.177127s wall, 0.170000s user + 0.010000s system = 0.180000s CPU (101.6%)\n",
      "Uniq kmers merging in  0.002500s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmers branching in  0.030366s wall, 0.030000s user + 0.000000s system = 0.030000s CPU (98.8%)\n",
      "Kmer: 29 Graph size: 16358 Contigs in: 31\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.013278s wall, 0.020000s user + 0.000000s system = 0.020000s CPU (150.6%)\n",
      "Kmers in multiple/single contigs: 0 12455\n",
      "Fragments before: 79 3091\n",
      "Fragments after: 79 3091\n",
      "New seeds: 0\n",
      "New seeds in  0.023398s wall, 0.020000s user + 0.000000s system = 0.020000s CPU (85.5%)\n",
      "Fragments before: 62 1680\n",
      "Fragments after: 61 1680\n",
      "Connectors: 1 Extenders: 60\n",
      "Connections and extensions in  0.026612s wall, 0.030000s user + 0.000000s system = 0.030000s CPU (112.7%)\n",
      "Contigs out: 30 Genome: 21514 N50: 1385 L50: 4\n",
      "Assembled in  0.063688s wall, 0.070000s user + 0.000000s system = 0.070000s CPU (109.9%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 7142\n",
      "Reads for connection: 5104\n",
      "Internal reads: 580\n",
      "Reads cleaned in  0.108943s wall, 0.110000s user + 0.000000s system = 0.110000s CPU (101.0%)\n",
      "\n",
      "Kmer len: 31\n",
      "Raw kmers: 493367 Memory needed (GB): 0.00947265 Memory available (GB): 29.9985 1 cycle(s) will be performed\n",
      "Distinct kmers: 15972\n",
      "Kmer count in  0.171107s wall, 0.170000s user + 0.000000s system = 0.170000s CPU (99.4%)\n",
      "Uniq kmers merging in  0.002453s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmers branching in  0.031414s wall, 0.040000s user + 0.000000s system = 0.040000s CPU (127.3%)\n",
      "Kmer: 31 Graph size: 15972 Contigs in: 30\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.013786s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (72.5%)\n",
      "Kmers in multiple/single contigs: 0 12207\n",
      "Fragments before: 77 2974\n",
      "Fragments after: 77 2974\n",
      "New seeds: 0\n",
      "New seeds in  0.022755s wall, 0.030000s user + 0.000000s system = 0.030000s CPU (131.8%)\n",
      "Fragments before: 60 1649\n",
      "Fragments after: 60 1649\n",
      "Connectors: 0 Extenders: 60\n",
      "Connections and extensions in  0.024681s wall, 0.020000s user + 0.000000s system = 0.020000s CPU (81.0%)\n",
      "Contigs out: 30 Genome: 21514 N50: 1385 L50: 4\n",
      "Assembled in  0.061616s wall, 0.060000s user + 0.000000s system = 0.060000s CPU (97.4%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 7142\n",
      "Reads for connection: 5104\n",
      "Internal reads: 580\n",
      "Reads cleaned in  0.125667s wall, 0.130000s user + 0.000000s system = 0.130000s CPU (103.4%)\n",
      "\n",
      "Kmer len: 35\n",
      "Raw kmers: 464799 Memory needed (GB): 0.0133862 Memory available (GB): 29.9982 1 cycle(s) will be performed\n",
      "Distinct kmers: 15577\n",
      "Kmer count in  0.198249s wall, 0.200000s user + 0.010000s system = 0.210000s CPU (105.9%)\n",
      "Uniq kmers merging in  0.002944s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmers branching in  0.036919s wall, 0.040000s user + 0.000000s system = 0.040000s CPU (108.3%)\n",
      "Kmer: 35 Graph size: 15577 Contigs in: 30\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.014080s wall, 0.010000s user + 0.010000s system = 0.020000s CPU (142.0%)\n",
      "Kmers in multiple/single contigs: 0 11952\n",
      "Fragments before: 66 2393\n",
      "Fragments after: 66 2393\n",
      "New seeds: 0\n",
      "New seeds in  0.017143s wall, 0.020000s user + 0.000000s system = 0.020000s CPU (116.7%)\n",
      "Fragments before: 58 1652\n",
      "Fragments after: 57 1652\n",
      "Connectors: 1 Extenders: 56\n",
      "Connections and extensions in  0.028915s wall, 0.030000s user + 0.000000s system = 0.030000s CPU (103.8%)\n",
      "Contigs out: 29 Genome: 21549 N50: 1385 L50: 4\n",
      "Assembled in  0.060516s wall, 0.060000s user + 0.010000s system = 0.070000s CPU (115.7%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 7142\n",
      "Reads for connection: 4830\n",
      "Internal reads: 671\n",
      "Reads cleaned in  0.129343s wall, 0.130000s user + 0.000000s system = 0.130000s CPU (100.5%)\n",
      "\n",
      "Kmer len: 39\n",
      "Raw kmers: 436231 Memory needed (GB): 0.0125635 Memory available (GB): 29.9978 1 cycle(s) will be performed\n",
      "Distinct kmers: 15127\n",
      "Kmer count in  0.177935s wall, 0.170000s user + 0.000000s system = 0.170000s CPU (95.5%)\n",
      "Uniq kmers merging in  0.002184s wall, 0.000000s user + 0.010000s system = 0.010000s CPU (457.8%)\n",
      "Kmers branching in  0.037562s wall, 0.040000s user + 0.000000s system = 0.040000s CPU (106.5%)\n",
      "Kmer: 39 Graph size: 15127 Contigs in: 29\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.014027s wall, 0.020000s user + 0.000000s system = 0.020000s CPU (142.6%)\n",
      "Kmers in multiple/single contigs: 0 11740\n",
      "Fragments before: 59 1960\n",
      "Fragments after: 59 1960\n",
      "New seeds: 0\n",
      "New seeds in  0.020701s wall, 0.020000s user + 0.000000s system = 0.020000s CPU (96.6%)\n",
      "Fragments before: 55 1605\n",
      "Fragments after: 55 1605\n",
      "Connectors: 0 Extenders: 55\n",
      "Connections and extensions in  0.028804s wall, 0.030000s user + 0.000000s system = 0.030000s CPU (104.2%)\n",
      "Contigs out: 29 Genome: 21549 N50: 1385 L50: 4\n",
      "Assembled in  0.063907s wall, 0.070000s user + 0.000000s system = 0.070000s CPU (109.5%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 7140\n",
      "Reads for connection: 4830\n",
      "Internal reads: 671\n",
      "Reads cleaned in  0.121870s wall, 0.120000s user + 0.000000s system = 0.120000s CPU (98.5%)\n",
      "\n",
      "Kmer len: 43\n",
      "Raw kmers: 407617 Memory needed (GB): 0.0117394 Memory available (GB): 29.9974 1 cycle(s) will be performed\n",
      "Distinct kmers: 14636\n",
      "Kmer count in  0.171968s wall, 0.170000s user + 0.010000s system = 0.180000s CPU (104.7%)\n",
      "Uniq kmers merging in  0.002330s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmers branching in  0.030944s wall, 0.030000s user + 0.000000s system = 0.030000s CPU (96.9%)\n",
      "Kmer: 43 Graph size: 14636 Contigs in: 29\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.014395s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (69.5%)\n",
      "Kmers in multiple/single contigs: 0 11425\n",
      "Fragments before: 49 1866\n",
      "Fragments after: 49 1866\n",
      "New seeds: 0\n",
      "New seeds in  0.011375s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (87.9%)\n",
      "Fragments before: 54 1778\n",
      "Fragments after: 48 1778\n",
      "Connectors: 6 Extenders: 42\n",
      "Connections and extensions in  0.025427s wall, 0.030000s user + 0.000000s system = 0.030000s CPU (118.0%)\n",
      "Contigs out: 23 Genome: 21967 N50: 1385 L50: 4\n",
      "Assembled in  0.051566s wall, 0.050000s user + 0.000000s system = 0.050000s CPU (97.0%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 5145\n",
      "Reads for connection: 2448\n",
      "Internal reads: 766\n",
      "Reads cleaned in  0.077901s wall, 0.080000s user + 0.000000s system = 0.080000s CPU (102.7%)\n",
      "\n",
      "Kmer len: 47\n",
      "Raw kmers: 273113 Memory needed (GB): 0.00786565 Memory available (GB): 29.9972 1 cycle(s) will be performed\n",
      "Distinct kmers: 13363\n",
      "Kmer count in  0.121301s wall, 0.110000s user + 0.000000s system = 0.110000s CPU (90.7%)\n",
      "Uniq kmers merging in  0.002147s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (465.7%)\n",
      "Kmers branching in  0.032234s wall, 0.030000s user + 0.000000s system = 0.030000s CPU (93.1%)\n",
      "Kmer: 47 Graph size: 13363 Contigs in: 23\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.015754s wall, 0.020000s user + 0.000000s system = 0.020000s CPU (127.0%)\n",
      "Kmers in multiple/single contigs: 0 11053\n",
      "Fragments before: 41 1054\n",
      "Fragments after: 41 1054\n",
      "New seeds: 0\n",
      "New seeds in  0.014477s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (69.1%)\n",
      "Fragments before: 41 1054\n",
      "Fragments after: 41 1054\n",
      "Connectors: 0 Extenders: 41\n",
      "Connections and extensions in  0.025670s wall, 0.030000s user + 0.000000s system = 0.030000s CPU (116.9%)\n",
      "Contigs out: 23 Genome: 21967 N50: 1385 L50: 4\n",
      "Assembled in  0.056255s wall, 0.060000s user + 0.000000s system = 0.060000s CPU (106.7%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 5145\n",
      "Reads for connection: 2448\n",
      "Internal reads: 765\n",
      "Reads cleaned in  0.072972s wall, 0.070000s user + 0.000000s system = 0.070000s CPU (95.9%)\n",
      "\n",
      "Kmer len: 51\n",
      "Raw kmers: 252533 Memory needed (GB): 0.00727295 Memory available (GB): 29.9968 1 cycle(s) will be performed\n",
      "Distinct kmers: 12825\n",
      "Kmer count in  0.115197s wall, 0.110000s user + 0.000000s system = 0.110000s CPU (95.5%)\n",
      "Uniq kmers merging in  0.002029s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmers branching in  0.030557s wall, 0.030000s user + 0.000000s system = 0.030000s CPU (98.2%)\n",
      "Kmer: 51 Graph size: 12825 Contigs in: 23\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.015413s wall, 0.020000s user + 0.000000s system = 0.020000s CPU (129.8%)\n",
      "Kmers in multiple/single contigs: 0 10692\n",
      "Fragments before: 40 1121\n",
      "Fragments after: 40 1121\n",
      "New seeds: 0\n",
      "New seeds in  0.012385s wall, 0.000000s user + 0.010000s system = 0.010000s CPU (80.7%)\n",
      "Fragments before: 41 1121\n",
      "Fragments after: 40 1121\n",
      "Connectors: 1 Extenders: 39\n",
      "Connections and extensions in  0.023244s wall, 0.020000s user + 0.000000s system = 0.020000s CPU (86.0%)\n",
      "Contigs out: 22 Genome: 22083 N50: 1385 L50: 4\n",
      "Assembled in  0.051403s wall, 0.040000s user + 0.010000s system = 0.050000s CPU (97.3%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 5080\n",
      "Reads for connection: 2340\n",
      "Internal reads: 776\n",
      "Reads cleaned in  0.060440s wall, 0.070000s user + 0.000000s system = 0.070000s CPU (115.8%)\n",
      "\n",
      "Kmer len: 55\n",
      "Raw kmers: 229171 Memory needed (GB): 0.00660012 Memory available (GB): 29.9965 1 cycle(s) will be performed\n",
      "Distinct kmers: 12041\n",
      "Kmer count in  0.099167s wall, 0.090000s user + 0.010000s system = 0.100000s CPU (100.8%)\n",
      "Uniq kmers merging in  0.002011s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmers branching in  0.031672s wall, 0.030000s user + 0.000000s system = 0.030000s CPU (94.7%)\n",
      "Kmer: 55 Graph size: 12041 Contigs in: 22\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.014888s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (67.2%)\n",
      "Kmers in multiple/single contigs: 0 10264\n",
      "Fragments before: 34 898\n",
      "Fragments after: 34 898\n",
      "New seeds: 0\n",
      "New seeds in  0.009834s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (101.7%)\n",
      "Fragments before: 36 898\n",
      "Fragments after: 34 898\n",
      "Connectors: 2 Extenders: 32\n",
      "Connections and extensions in  0.022072s wall, 0.020000s user + 0.000000s system = 0.020000s CPU (90.6%)\n",
      "Contigs out: 20 Genome: 22189 N50: 1826 L50: 3\n",
      "Assembled in  0.047127s wall, 0.040000s user + 0.000000s system = 0.040000s CPU (84.9%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 2977\n",
      "Reads for connection: 1268\n",
      "Internal reads: 474\n",
      "Reads cleaned in  0.055895s wall, 0.060000s user + 0.000000s system = 0.060000s CPU (107.3%)\n",
      "\n",
      "Kmer len: 59\n",
      "Raw kmers: 122326 Memory needed (GB): 0.00352299 Memory available (GB): 29.9963 1 cycle(s) will be performed\n",
      "Distinct kmers: 9166\n",
      "Kmer count in  0.069402s wall, 0.070000s user + 0.000000s system = 0.070000s CPU (100.9%)\n",
      "Uniq kmers merging in  0.001789s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmers branching in  0.026367s wall, 0.030000s user + 0.000000s system = 0.030000s CPU (113.8%)\n",
      "Kmer: 59 Graph size: 9166 Contigs in: 20\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.013868s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (72.1%)\n",
      "Kmers in multiple/single contigs: 0 7800\n",
      "Fragments before: 31 660\n",
      "Fragments after: 31 660\n",
      "New seeds: 0\n",
      "New seeds in  0.007910s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (126.4%)\n",
      "Fragments before: 31 660\n",
      "Fragments after: 31 660\n",
      "Connectors: 0 Extenders: 31\n",
      "Connections and extensions in  0.023247s wall, 0.030000s user + 0.000000s system = 0.030000s CPU (129.0%)\n",
      "Contigs out: 20 Genome: 22189 N50: 1826 L50: 3\n",
      "Assembled in  0.045324s wall, 0.050000s user + 0.000000s system = 0.050000s CPU (110.3%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 2975\n",
      "Reads for connection: 1268\n",
      "Internal reads: 473\n",
      "Reads cleaned in  0.051670s wall, 0.040000s user + 0.010000s system = 0.050000s CPU (96.8%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 21\n",
      "Connected: 361 ambiguously connected: 1 from 634 mate pairs\n",
      "Connect pairs in  0.247733s wall, 0.250000s user + 0.000000s system = 0.250000s CPU (100.9%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 25\n",
      "Connected: 1 ambiguously connected: 0 from 1 mate pairs\n",
      "Connect pairs in  0.006064s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 29\n",
      "Connected: 0 ambiguously connected: 0 from 0 mate pairs\n",
      "Connect pairs in  0.000025s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 31\n",
      "Connected: 0 ambiguously connected: 0 from 0 mate pairs\n",
      "Connect pairs in  0.000024s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 35\n",
      "Connected: 0 ambiguously connected: 0 from 0 mate pairs\n",
      "Connect pairs in  0.000026s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 39\n",
      "Connected: 0 ambiguously connected: 0 from 0 mate pairs\n",
      "Connect pairs in  0.000033s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 43\n",
      "Connected: 0 ambiguously connected: 0 from 0 mate pairs\n",
      "Connect pairs in  0.000027s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 47\n",
      "Connected: 0 ambiguously connected: 0 from 0 mate pairs\n",
      "Connect pairs in  0.000025s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 51\n",
      "Connected: 0 ambiguously connected: 0 from 0 mate pairs\n",
      "Connect pairs in  0.000025s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 55\n",
      "Connected: 0 ambiguously connected: 0 from 0 mate pairs\n",
      "Connect pairs in  0.000025s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 59\n",
      "Connected: 0 ambiguously connected: 0 from 0 mate pairs\n",
      "Connect pairs in  0.000026s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Totally connected: 362\n",
      "Added notconnected: 0\n",
      "\n",
      "Kmer len: 73\n",
      "Raw kmers: 181635 Memory needed (GB): 0.00697478 Memory available (GB): 29.9961 1 cycle(s) will be performed\n",
      "Distinct kmers: 11338\n",
      "Kmer count in  0.093676s wall, 0.090000s user + 0.000000s system = 0.090000s CPU (96.1%)\n",
      "Uniq kmers merging in  0.001920s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmers branching in  0.029568s wall, 0.030000s user + 0.010000s system = 0.040000s CPU (135.3%)\n",
      "Kmer: 73 Graph size: 11338 Contigs in: 20\n",
      "Valley: 13\n",
      "\n",
      "Mark used kmers in  0.005912s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (169.2%)\n",
      "Kmers in multiple/single contigs: 0 9983\n",
      "Fragments before: 8 263\n",
      "Fragments after: 8 263\n",
      "New seeds: 0\n",
      "New seeds in  0.004913s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Fragments before: 40 1141\n",
      "Fragments after: 40 1141\n",
      "Connectors: 0 Extenders: 40\n",
      "Connections and extensions in  0.026348s wall, 0.030000s user + 0.000000s system = 0.030000s CPU (113.9%)\n",
      "Contigs out: 20 Genome: 22189 N50: 1826 L50: 3\n",
      "Assembled in  0.037514s wall, 0.040000s user + 0.000000s system = 0.040000s CPU (106.6%)\n",
      "\n",
      "\n",
      "Kmer len: 167\n",
      "Raw kmers: 108055 Memory needed (GB): 0.0072613 Memory available (GB): 29.9957 1 cycle(s) will be performed\n",
      "Distinct kmers: 8697\n",
      "Kmer count in  0.071768s wall, 0.070000s user + 0.000000s system = 0.070000s CPU (97.5%)\n",
      "Uniq kmers merging in  0.001931s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmers branching in  0.032725s wall, 0.030000s user + 0.000000s system = 0.030000s CPU (91.7%)\n",
      "Kmer: 167 Graph size: 8697 Contigs in: 20\n",
      "Valley: 7\n",
      "\n",
      "Mark used kmers in  0.017003s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (58.8%)\n",
      "Kmers in multiple/single contigs: 0 7655\n",
      "Fragments before: 18 551\n",
      "Fragments after: 18 551\n",
      "New seeds: 0\n",
      "New seeds in  0.006607s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (151.3%)\n",
      "Fragments before: 36 1011\n",
      "Fragments after: 36 1011\n",
      "Connectors: 0 Extenders: 36\n",
      "Connections and extensions in  0.029269s wall, 0.030000s user + 0.000000s system = 0.030000s CPU (102.5%)\n",
      "Contigs out: 18 Genome: 21990 N50: 1826 L50: 3\n",
      "Assembled in  0.053175s wall, 0.050000s user + 0.000000s system = 0.050000s CPU (94.0%)\n",
      "\n",
      "\n",
      "Kmer len: 261\n",
      "Raw kmers: 54924 Memory needed (GB): 0.0052727 Memory available (GB): 29.9952 1 cycle(s) will be performed\n",
      "Distinct kmers: 5733\n",
      "Kmer count in  0.049046s wall, 0.050000s user + 0.000000s system = 0.050000s CPU (101.9%)\n",
      "Uniq kmers merging in  0.001819s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmers branching in  0.026219s wall, 0.030000s user + 0.000000s system = 0.030000s CPU (114.4%)\n",
      "Kmer: 261 Graph size: 5733 Contigs in: 18\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.017483s wall, 0.020000s user + 0.000000s system = 0.020000s CPU (114.4%)\n",
      "Kmers in multiple/single contigs: 0 4990\n",
      "Fragments before: 25 723\n",
      "Fragments after: 25 723\n",
      "New seeds: 0\n",
      "New seeds in  0.007868s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Fragments before: 25 723\n",
      "Fragments after: 25 723\n",
      "Connectors: 0 Extenders: 25\n",
      "Connections and extensions in  0.030172s wall, 0.040000s user + 0.000000s system = 0.040000s CPU (132.6%)\n",
      "Contigs out: 16 Genome: 21577 N50: 1826 L50: 3\n",
      "Assembled in  0.055774s wall, 0.060000s user + 0.000000s system = 0.060000s CPU (107.6%)\n",
      "\n",
      "DONE\n",
      "\tCommand being timed: \"skesa --cores 1 --fastq SRR1952908.kat.in.R1.fastq,SRR1952908.kat.in.R2.fastq --contigs_out SRR1952908.kat.contigs.fasta --vector_percent 1\"\n",
      "\tUser time (seconds): 17.63\n",
      "\tSystem time (seconds): 0.26\n",
      "\tPercent of CPU this job got: 101%\n",
      "\tElapsed (wall clock) time (h:mm:ss or m:ss): 0:17.61\n",
      "\tAverage shared text size (kbytes): 0\n",
      "\tAverage unshared data size (kbytes): 0\n",
      "\tAverage stack size (kbytes): 0\n",
      "\tAverage total size (kbytes): 0\n",
      "\tMaximum resident set size (kbytes): 31000\n",
      "\tAverage resident set size (kbytes): 0\n",
      "\tMajor (requiring I/O) page faults: 0\n",
      "\tMinor (reclaiming a frame) page faults: 32123\n",
      "\tVoluntary context switches: 16031\n",
      "\tInvoluntary context switches: 170\n",
      "\tSwaps: 0\n",
      "\tFile system inputs: 0\n",
      "\tFile system outputs: 48\n",
      "\tSocket messages sent: 0\n",
      "\tSocket messages received: 0\n",
      "\tSignals delivered: 0\n",
      "\tPage size (bytes): 4096\n",
      "\tExit status: 0\n"
     ]
    }
   ],
   "source": [
    "conda run --name skesa /usr/bin/time -v skesa --cores 1 \\\n",
    "    --fastq SRR1952908.kat.in.R1.fastq,SRR1952908.kat.in.R2.fastq \\\n",
    "    --contigs_out SRR1952908.kat.contigs.fasta --vector_percent 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rw-r--r-- 1 apetkau grp_apetkau 22K Jan 16 17:32 SRR1952908.kat.contigs.fasta\n"
     ]
    }
   ],
   "source": [
    "ls -lh *kat*.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.3: Run RGI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "~/workspace/kmer-analysis-test/rgi-kat ~/workspace/kmer-analysis-test\n",
      "\tCommand being timed: \"rgi main -i ../SRR1952908.kat.contigs.fasta -o out -n 1\"\n",
      "\tUser time (seconds): 6.03\n",
      "\tSystem time (seconds): 4.63\n",
      "\tPercent of CPU this job got: 343%\n",
      "\tElapsed (wall clock) time (h:mm:ss or m:ss): 0:03.10\n",
      "\tAverage shared text size (kbytes): 0\n",
      "\tAverage unshared data size (kbytes): 0\n",
      "\tAverage stack size (kbytes): 0\n",
      "\tAverage total size (kbytes): 0\n",
      "\tMaximum resident set size (kbytes): 128820\n",
      "\tAverage resident set size (kbytes): 0\n",
      "\tMajor (requiring I/O) page faults: 0\n",
      "\tMinor (reclaiming a frame) page faults: 164086\n",
      "\tVoluntary context switches: 200\n",
      "\tInvoluntary context switches: 19263\n",
      "\tSwaps: 0\n",
      "\tFile system inputs: 0\n",
      "\tFile system outputs: 11232\n",
      "\tSocket messages sent: 0\n",
      "\tSocket messages received: 0\n",
      "\tSignals delivered: 0\n",
      "\tPage size (bytes): 4096\n",
      "\tExit status: 0\n",
      "~/workspace/kmer-analysis-test\n"
     ]
    }
   ],
   "source": [
    "mkdir rgi-kat\n",
    "pushd rgi-kat\n",
    "conda run --name rgi-4.2.2 /usr/bin/time -v rgi main -i ../SRR1952908.kat.contigs.fasta -o out -n 1\n",
    "popd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Best_Hit_ARO\n",
      "MdtK\n",
      "sul3\n",
      "aadA\n",
      "Salmonella enterica cmlA\n",
      "aadA13\n",
      "qacH\n",
      "emrR\n",
      "mef(B)\n",
      "sdiA\n",
      "TEM-57\n",
      "patA\n",
      "AAC(6')-Ib7\n",
      "cpxA\n",
      "mdtC\n",
      "AAC(6')-Iaa\n",
      "golS\n",
      "acrB\n",
      "mdsC\n",
      "adeF\n",
      "mdsA\n"
     ]
    }
   ],
   "source": [
    "cut -f 9 rgi-kat/out.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Scenario 3: Groot\n",
    "\n",
    "## 3.1 Run Groot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Compare results (detected AMR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Isolate ID\tGenotype\tPredicted Phenotype\n",
      "kat.filter.contigs\taadA1, aadA2, blaTEM-57, cmlA1, sul3, tet(A)\tstreptomycin, ampicillin, chloramphenicol, sulfisoxazole, tetracycline\n"
     ]
    }
   ],
   "source": [
    "cut -f 1,3,4 kat-staramr-out/summary.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Isolate ID\tGenotype\tPredicted Phenotype\n",
      "SRR1952908\taadA1, aadA2, blaTEM-57, cmlA1, sul3, tet(A)\tstreptomycin, ampicillin, chloramphenicol, sulfisoxazole, tetracycline\n"
     ]
    }
   ],
   "source": [
    "cut -f 1,3,4 staramr-out/summary.tsv"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
