{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Kmer analysis tests\n",
    "\n",
    "First let's setup some environment variables."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "PROJECT_DIR=`git rev-parse --show-toplevel`\n",
    "cd $PROJECT_DIR"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# True/false positives code\n",
    "\n",
    "Now, let's setup code to count true/false positives."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "print_positives() {\n",
    "    file1=$1\n",
    "    file2=$2\n",
    "\n",
    "    python -c \"\n",
    "set1 = set(l.strip() for l in open('${file1}'))\n",
    "set2 = set(l.strip() for l in open('${file2}'))\n",
    "\n",
    "true_positives = set1 & set2\n",
    "false_positives = set2 - set1\n",
    "false_negatives = set1 - set2\n",
    "\n",
    "sensitivity = len(true_positives) / (len(true_positives) + len(false_negatives))\n",
    "\n",
    "print(\\\"True Positives: %s: %s\\\" % (len(true_positives), true_positives))\n",
    "print(\\\"False Positives: %s: %s\\\" % (len(false_positives), false_positives))\n",
    "print(\\\"False Negatives: %s: %s\\\" % (len(false_negatives), false_negatives))\n",
    "print(\\\"Sensitivity: %0.2f\\\" % (sensitivity))\n",
    "\"\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Examine input files\n",
    "\n",
    "Let's take a look at the input files.\n",
    "\n",
    "## CARD protein homolog database\n",
    "\n",
    "The CARD protein homolog database looks like."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">gb|GQ343019|+|132-1023|ARO:3002999|CblA-1 [mixed culture bacterium AX_gF3SD01_15] \n",
      "ATGAAAGCATATTTCATCGCCATACTTACCTTATTCACTTGTATAGCTACCGTCGTCCGGGCGCAGCAAATGTCTGAACTTGAAAACCGGATTGACAGTCTGCTCAATGGCAAGAAAGCCACCGTTGGTATAGCCGTATGGACAGACAAAGGAGACATGCTCCGGTATAACGACCATGTACACTTCCCCTTGCTCAGTGTATTCAAATTCCATGTGGCACTGGCCGTACTGGACAAGATGGATAAGCAAAGCATCAGTCTGGACAGCATTGTTTCCATAAAGGCATCCCAAATGCCGCCCAATACCTACAGCCCCCTGCGGAAGAAGTTTCCCGACCAGGATTTCACGATTACGCTTAGGGAACTGATGCAATACAGCATTTCCCAAAGCGACAACAATGCCTGCGACATCTTGATAGAATATGCAGGAGGCATCAAACATATCAACGACTATATCCACCGGTTGAGTATCGACTCCTTCAACCTCTCGGAAACAGAAGACGGCATGCACTCCAGCTTCGAGGCTGTATACCGCAACTGGAGTACTCCTTCCGCTATGGTCCGACTACTGAGAACGGCTGATGAAAAAGAGTTGTTCTCCAACAAGGAGCTGAAAGACTTCTTGTGGCAGACCATGATAGATACTGAAACCGGTGCCAACAAACTGAAAGGTATGTTGCCAGCCAAAACCGTGGTAGGACACAAGACCGGCTCTTCCGACCGCAATGCCGACGGTATGAAAACTGCAGATAATGATGCCGGCCTCGTTATCCTTCCCGACGGCCGGAAATACTACATTGCCGCCTTCGTCATGGACTCATACGAGACGGATGAGGACAATGCGAACATCATCGCCCGCATATCACGCATGGTATATGATGCGATGAGATGA\n",
      ">gb|HQ845196|+|0-861|ARO:3001109|SHV-52 [Klebsiella pneumoniae] \n",
      "ATGCGTTATATTCGCCTGTGTATTATCTCCCTGTTAGCCGCCCTGCCGCTGGCGGTACACGCCAGCCCGCAGCCGCTTGAGCAAATTAAACAAAGCGAAAGCCAGCTGTCGGGCCGCGTAGGCATGATAGAAATGGATCTGGCCAGCGGCCGCACGCTGACCGCCTGGCGCGCCGATGAACGCTTTCCCATGATAAGCACCTTTAAAGTAGTGCTCTGCGGCGCAGTGCTGGCGCGGGTGGATGCCGGTGACGAACAGCTGGAGCGAAAGATCCACTATCGCCAGCAGGATCTGGTGGACTACTCGCCGGTCAGCGAAAAACACCTTGCCGACGGCATGACGGTCGGCGAACTCTGTGCCGCCGCCATTACCATGAGCGATAACAGCGCCGCCAATTTGCTGCTGGCCATCGTCGGCGGCCCCGCAGGATTGACTGCCTTTTTGCGCCAGATCGGCGACAACGTCACCCGCCTTGACCGCTGGGAAACGGAACTGAATGAGGCGCTTCCCGGCGACGCCCGCGACACCACTACCCCGGCCAGCATGGCCGCGACCCTGCGCAAGCTGCTGACCAGCCAGCGTCTGAGCGCCCGTTCGCAACGGCAGCTGCTGCAGTGGATGGTGGACGATCGGGTCGCCGGACCGTTGATCCGCTCCGTGCTGCCGGCGGGCTGGTTTATCGCCGATAAGACCGGAGCTGGCGAACGGGGTGCGCGCGGGATTGTCGCCCTGCTTGGCCCGAATAACAAAGCAGAGCGCATTGTAGTGATTTATCTGCGGGATACCCCGGCGAGCATGGCCGAGCGAAATCAGCAAATCGCCGGGATCGGCGCGGCGCTGATCGAGCACTGGCAACGCTAA\n",
      ">gb|AF028812|+|392-887|ARO:3002867|dfrF [Enterococcus faecalis] \n",
      "ATGATAGGTTTGATTGTTGCGAGGTCAAAGAATAATGTTATAGGCAAGAATGGTAATATACCATGGAAAATAAAGGGAGAACAAAAGCAATTTAGAGAGTTAACAACGGGTAATGTGGTTATTATGGGGCGAAAGTCTTATGAAGAAATCGGTCATCCGTTGCCTAATAGAATGAATATTGTTGTTTCCACCACAACAGAGTATCAAGGAGATAATTTAGTTTCAGTTAAATCATTAGAAGATGCATTATTATTGGCTAAAGGACGAGATGTATACATATCTGGTGGATATGGACTATTTAAGGAAGCTTTGCAAATAGTAGATAAAATGTATATCACAGAAGTAGATTTAAATATTGAAGATGGAGATACATTCTTTCCAGAATTTGATATCAATGATTTTGAAGTTTTGATAGGGGAAACACTTGGTGAGGAAGTGAAATATACGAGAACATTTTATGTAAGGAAAAATGAATTGAGTAGATTTTGGATTTAG\n",
      ">gb|JX017365|+|244-1120|ARO:3001989|CTX-M-130 [Escherichia coli] \n",
      "ATGGTGACAAAGAGAGTGCAACGGATGATGTTCGCGGCGGCGGCGTGCATTCCGCTGCTGCTGGGCAGCGCGCCGCTTTATGCGCAGACGAGTGCGGTGCAGCAAAAGCTGGCGGCGCTGGAGAAAAGCAGCGGAGGGCGGCTGGGCGTCGCGCTCATCGATACCGCAGATAATACGCAGGTGCTTTATCGCGGTGATGAACGCTTTCCAATGTGCAGTACCAGTAAAGTTATGGCGGCCGCGGCGGTGCTTAAGCAGAGTGAAACGCAAAAGCAGCTGCTTAATCAGCCTGTCGAGATCAAGCCTGCCGATCTGGTTAACTACAATCCGATTGCCGAAAAACACGTCAACGGCACAATGACGCTGGCAGAACTGAGCGCGGCCGCGTTGCAGTACAGCGACAATACCGCCATGAACAAATTGATTGCCCAGCTCGGTGGCCCGGGAGGCGTGACGGCTTTTGCCCGCGCGATCGGCGATGAGACGTTTCGTCTGGATCGCACTGAACCTACGCTGAATACCGCCATTCCCGGCGACCCGAGAGACACCACCACGCCGCGGGCGATGGCGCAGACGTTGCGTCAGCTTACGCTGGGTCATGCGCTGGGCGAAACCCAGCGGGCGCAGTTGGTGACGTGGCTCAAAGGCAATACGACCGGCGCAGCCAGCATTCGGGCCGGCTTACCGACGTCGTGGACTGTGGGTGATAAGACCGGCAGCGGCGACTACGGCACCACCAATGATATTGCGGTGATCTGGCCGCAGGGTCGCGCGCCGCTGGTTCTGGTGACCTATTTTACCCAGCCGCAACAGAACGCAGAGAGACGCCACGATGTGCTGGCTTCAGCGGCGAGAATCATCGCCGAAGGGCTGTAA\n",
      ">gb|JN967644|+|0-813|ARO:3002356|NDM-6 [Escherichia coli] \n",
      "ATGGAATTGCCCAATATTATGCACCCGGTCGCGAAGCTGAGCACCGCATTAGCCGCTGCATTGATGCTGAGCGGGTGCATGCCCGGTGAAATCCGCCCGACGATTGGCCAGCAAATGGAAACTGGCGACCAACGGTTTGGCGATCTGGTTTTCCGCCAGCTCGCACCGAATGTCTGGCAGCACACTTCCTATCTCGACATGCCGGGTTTCGGGGCAGTCGCTTCCAACGGTTTGATCGTCAGGGATGGCGGCCGCGTGCTGGTGGTCGATACCGCCTGGACCGATGACCAGACCGCCCAGATCCTCAACTGGATCAAGCAGGAGATCAACCTGCCGGTCGCGCTGGCGGTGGTGACTCACGCGCATCAGGACAAGATGGGCGGTATGGACGCGCTGCATGCGGCGGGGATTGCGACTTATGCCAATGCGTTGTCGAACCAGCTTGCCCCGCAAGAGGGGATGGTTGCGGCGCAACACAGCCTGACTTTCGCCGCCAATGGCTGGGTCGAACCAGCAACCGCGCCCAACTTTGGCCCGCTCAAGGTATTTTACCCCGGCCCCGGCCACACCAGTGACAATATCACCGTTGGGATCGACGGCACCGACATCGCTTTTGGTGGCTGCCTGATCAAGGACAGCAAGGCCAAGTCGCTCGGCAATCTCGGTGATGCCGACACTGAGCACTACGCCGCGTCAGTGCGCGCGTTTGGTGCGGCGTTCCCCAAGGCCAGCATGATCGTGATGAGCCATTCCGCCCCCGATAGCCGCGCCGCAATCACTCATACGGCCCGCATGGCCGACAAGCTGCGCTGA\n"
     ]
    }
   ],
   "source": [
    "head db/nucleotide_fasta_protein_homolog_model.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sequence reads\n",
    "\n",
    "The sequence reads look like:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rwxr-xr-x 1 apetkau grp_apetkau 474M Jan 15 22:42 \u001b[0m\u001b[01;32minput/SRR1952908_1.fastq\u001b[0m\n",
      "-rwxr-xr-x 1 apetkau grp_apetkau 471M Jan 15 22:42 \u001b[01;32minput/SRR1952908_2.fastq\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "ls -lrth input/*.fastq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Scenario 1: Assemble genome with SKESA, run RGI\n",
    "\n",
    "Let's test out assembling the genome with SKESA and running RGI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "skesa --cores 1 --fastq input/SRR1952908_1.fastq,input/SRR1952908_2.fastq --contigs_out SRR1952908-full-assembly.fasta --vector_percent 1 \n",
      "\n",
      "Total mates: 4294092 Paired reads: 2147046\n",
      "Reads acquired in  10.307535s wall, 9.950000s user + 0.350000s system = 10.300000s CPU (99.9%)\n",
      "Adapters clip is disabled\n",
      "\n",
      "Kmer len: 21\n",
      "Raw kmers: 338084078 Memory needed (GB): 6.49121 Memory available (GB): 29.7537 1 cycle(s) will be performed\n",
      "Distinct kmers: 5702791\n",
      "Kmer count in  78.417178s wall, 70.970000s user + 8.790000s system = 79.760000s CPU (101.7%)\n",
      "Uniq kmers merging in  0.403631s wall, 0.210000s user + 0.190000s system = 0.400000s CPU (99.1%)\n",
      "Kmers branching in  11.062922s wall, 11.010000s user + 0.170000s system = 11.180000s CPU (101.1%)\n",
      "\n",
      "Average read length: 99\n",
      "Genome size estimate: 4298544\n",
      "\n",
      "Kmer: 21 Graph size: 5702791 Contigs in: 0\n",
      "Valley: 13\n",
      "\n",
      "Mark used kmers in  0.000014s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmers in multiple/single contigs: 0 0\n",
      "Fragments before: 1519 4753547\n",
      "Fragments after: 1519 4753547\n",
      "New seeds: 1223\n",
      "New seeds in  13.676004s wall, 13.620000s user + 0.190000s system = 13.810000s CPU (101.0%)\n",
      "Fragments before: 2446 51366\n",
      "Fragments after: 2446 51366\n",
      "Connectors: 0 Extenders: 2446\n",
      "Connections and extensions in  1.987397s wall, 1.990000s user + 0.020000s system = 2.010000s CPU (101.1%)\n",
      "Contigs out: 1223 Genome: 4709449 N50: 8488 L50: 172\n",
      "Assembled in  15.690450s wall, 15.640000s user + 0.210000s system = 15.850000s CPU (101.0%)\n",
      "\n",
      "\n",
      "Kmer len: 85\n",
      "Raw kmers: 64986845 Memory needed (GB): 2.49549 Memory available (GB): 29.6567 1 cycle(s) will be performed\n",
      "Distinct kmers: 4981204\n",
      "Kmer count in  20.826940s wall, 19.670000s user + 1.480000s system = 21.150000s CPU (101.6%)\n",
      "Uniq kmers merging in  0.516224s wall, 0.230000s user + 0.290000s system = 0.520000s CPU (100.7%)\n",
      "\n",
      "Average count: 59.7629 Max kmer: 85\n",
      "Connected: 8675 ambiguously connected: 514 from 10000 mate pairs\n",
      "Connect pairs in  44.562795s wall, 44.810000s user + 0.590000s system = 45.400000s CPU (101.9%)\n",
      "\n",
      "N50 for inserts: 259\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 662140\n",
      "Reads for connection: 393720\n",
      "Internal reads: 62222\n",
      "Reads cleaned in  19.772386s wall, 19.710000s user + 0.290000s system = 20.000000s CPU (101.2%)\n",
      "\n",
      "Kmer len: 27\n",
      "Raw kmers: 48108317 Memory needed (GB): 0.92368 Memory available (GB): 29.8691 1 cycle(s) will be performed\n",
      "Distinct kmers: 1369177\n",
      "Kmer count in  10.053764s wall, 9.630000s user + 0.520000s system = 10.150000s CPU (101.0%)\n",
      "Uniq kmers merging in  0.113859s wall, 0.090000s user + 0.030000s system = 0.120000s CPU (105.4%)\n",
      "Kmers branching in  2.055404s wall, 2.070000s user + 0.020000s system = 2.090000s CPU (101.7%)\n",
      "Kmer: 27 Graph size: 1369177 Contigs in: 1223\n",
      "Valley: 28\n",
      "\n",
      "Mark used kmers in  2.044477s wall, 2.050000s user + 0.040000s system = 2.090000s CPU (102.2%)\n",
      "Kmers in multiple/single contigs: 0 1055119\n",
      "Fragments before: 1806 82905\n",
      "Fragments after: 1806 82905\n",
      "New seeds: 52\n",
      "New seeds in  0.930840s wall, 0.920000s user + 0.020000s system = 0.940000s CPU (101.0%)\n",
      "Fragments before: 2550 71053\n",
      "Fragments after: 1776 71053\n",
      "Connectors: 774 Extenders: 1002\n",
      "Connections and extensions in  1.995614s wall, 2.000000s user + 0.010000s system = 2.010000s CPU (100.7%)\n",
      "Contigs out: 501 Genome: 4739479 N50: 27001 L50: 55\n",
      "Assembled in  4.989471s wall, 4.990000s user + 0.070000s system = 5.060000s CPU (101.4%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 350254\n",
      "Reads for connection: 232418\n",
      "Internal reads: 27462\n",
      "Reads cleaned in  9.850151s wall, 9.770000s user + 0.200000s system = 9.970000s CPU (101.2%)\n",
      "\n",
      "Kmer len: 33\n",
      "Raw kmers: 23351255 Memory needed (GB): 0.672516 Memory available (GB): 29.8614 1 cycle(s) will be performed\n",
      "Distinct kmers: 628964\n",
      "Kmer count in  6.142028s wall, 5.950000s user + 0.300000s system = 6.250000s CPU (101.8%)\n",
      "Uniq kmers merging in  0.070658s wall, 0.050000s user + 0.020000s system = 0.070000s CPU (99.1%)\n",
      "Kmers branching in  1.013138s wall, 1.020000s user + 0.010000s system = 1.030000s CPU (101.7%)\n",
      "Kmer: 33 Graph size: 628964 Contigs in: 501\n",
      "Valley: 30\n",
      "\n",
      "Mark used kmers in  1.801062s wall, 1.810000s user + 0.020000s system = 1.830000s CPU (101.6%)\n",
      "Kmers in multiple/single contigs: 0 419191\n",
      "Fragments before: 893 42926\n",
      "Fragments after: 893 42926\n",
      "New seeds: 17\n",
      "New seeds in  0.515022s wall, 0.500000s user + 0.020000s system = 0.520000s CPU (101.0%)\n",
      "Fragments before: 1035 35732\n",
      "Fragments after: 873 35732\n",
      "Connectors: 162 Extenders: 711\n",
      "Connections and extensions in  1.892647s wall, 1.890000s user + 0.040000s system = 1.930000s CPU (102.0%)\n",
      "Contigs out: 356 Genome: 4749844 N50: 40174 L50: 40\n",
      "Assembled in  4.217815s wall, 4.210000s user + 0.080000s system = 4.290000s CPU (101.7%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 277653\n",
      "Reads for connection: 193410\n",
      "Internal reads: 19918\n",
      "Reads cleaned in  9.565238s wall, 9.440000s user + 0.240000s system = 9.680000s CPU (101.2%)\n",
      "\n",
      "Kmer len: 39\n",
      "Raw kmers: 16847346 Memory needed (GB): 0.485204 Memory available (GB): 29.8493 1 cycle(s) will be performed\n",
      "Distinct kmers: 458933\n",
      "Kmer count in  4.317195s wall, 4.150000s user + 0.230000s system = 4.380000s CPU (101.5%)\n",
      "Uniq kmers merging in  0.050522s wall, 0.040000s user + 0.010000s system = 0.050000s CPU (99.0%)\n",
      "Kmers branching in  0.699423s wall, 0.700000s user + 0.000000s system = 0.700000s CPU (100.1%)\n",
      "Kmer: 39 Graph size: 458933 Contigs in: 356\n",
      "Valley: 27\n",
      "\n",
      "Mark used kmers in  1.728018s wall, 1.730000s user + 0.020000s system = 1.750000s CPU (101.3%)\n",
      "Kmers in multiple/single contigs: 0 278888\n",
      "Fragments before: 631 34665\n",
      "Fragments after: 631 34665\n",
      "New seeds: 12\n",
      "New seeds in  0.361222s wall, 0.360000s user + 0.010000s system = 0.370000s CPU (102.4%)\n",
      "Fragments before: 736 29035\n",
      "Fragments after: 622 29035\n",
      "Connectors: 114 Extenders: 508\n",
      "Connections and extensions in  1.889419s wall, 1.890000s user + 0.020000s system = 1.910000s CPU (101.1%)\n",
      "Contigs out: 254 Genome: 4758206 N50: 61729 L50: 26\n",
      "Assembled in  3.985346s wall, 3.980000s user + 0.050000s system = 4.030000s CPU (101.1%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 226872\n",
      "Reads for connection: 165128\n",
      "Internal reads: 14896\n",
      "Reads cleaned in  8.202592s wall, 8.080000s user + 0.210000s system = 8.290000s CPU (101.1%)\n",
      "\n",
      "Kmer len: 47\n",
      "Raw kmers: 11954675 Memory needed (GB): 0.344295 Memory available (GB): 29.8404 1 cycle(s) will be performed\n",
      "Distinct kmers: 330405\n",
      "Kmer count in  3.002826s wall, 2.820000s user + 0.220000s system = 3.040000s CPU (101.2%)\n",
      "Uniq kmers merging in  0.032089s wall, 0.010000s user + 0.020000s system = 0.030000s CPU (93.5%)\n",
      "Kmers branching in  0.490308s wall, 0.490000s user + 0.010000s system = 0.500000s CPU (102.0%)\n",
      "Kmer: 47 Graph size: 330405 Contigs in: 254\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  1.559271s wall, 1.560000s user + 0.020000s system = 1.580000s CPU (101.3%)\n",
      "Kmers in multiple/single contigs: 0 177717\n",
      "Fragments before: 491 29610\n",
      "Fragments after: 491 29610\n",
      "New seeds: 7\n",
      "New seeds in  0.481796s wall, 0.480000s user + 0.010000s system = 0.490000s CPU (101.7%)\n",
      "Fragments before: 522 24922\n",
      "Fragments after: 464 24922\n",
      "Connectors: 58 Extenders: 406\n",
      "Connections and extensions in  1.860577s wall, 1.860000s user + 0.010000s system = 1.870000s CPU (100.5%)\n",
      "Contigs out: 204 Genome: 4764532 N50: 67769 L50: 22\n",
      "Assembled in  3.906492s wall, 3.910000s user + 0.040000s system = 3.950000s CPU (101.1%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 200982\n",
      "Reads for connection: 145774\n",
      "Internal reads: 13164\n",
      "Reads cleaned in  7.178059s wall, 7.050000s user + 0.220000s system = 7.270000s CPU (101.3%)\n",
      "\n",
      "Kmer len: 53\n",
      "Raw kmers: 9378489 Memory needed (GB): 0.2701 Memory available (GB): 29.8335 1 cycle(s) will be performed\n",
      "Distinct kmers: 277865\n",
      "Kmer count in  2.476944s wall, 2.410000s user + 0.110000s system = 2.520000s CPU (101.7%)\n",
      "Uniq kmers merging in  0.031725s wall, 0.030000s user + 0.000000s system = 0.030000s CPU (94.6%)\n",
      "Kmers branching in  0.401401s wall, 0.410000s user + 0.000000s system = 0.410000s CPU (102.1%)\n",
      "Kmer: 53 Graph size: 277865 Contigs in: 204\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  1.542821s wall, 1.530000s user + 0.040000s system = 1.570000s CPU (101.8%)\n",
      "Kmers in multiple/single contigs: 0 146855\n",
      "Fragments before: 411 23403\n",
      "Fragments after: 411 23403\n",
      "New seeds: 2\n",
      "New seeds in  0.353511s wall, 0.360000s user + 0.000000s system = 0.360000s CPU (101.8%)\n",
      "Fragments before: 410 20205\n",
      "Fragments after: 391 20205\n",
      "Connectors: 19 Extenders: 372\n",
      "Connections and extensions in  1.862755s wall, 1.850000s user + 0.030000s system = 1.880000s CPU (100.9%)\n",
      "Contigs out: 187 Genome: 4767010 N50: 67769 L50: 22\n",
      "Assembled in  3.760893s wall, 3.740000s user + 0.070000s system = 3.810000s CPU (101.3%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 186291\n",
      "Reads for connection: 135410\n",
      "Internal reads: 12139\n",
      "Reads cleaned in  7.042954s wall, 6.980000s user + 0.170000s system = 7.150000s CPU (101.5%)\n",
      "\n",
      "Kmer len: 59\n",
      "Raw kmers: 7593238 Memory needed (GB): 0.218685 Memory available (GB): 29.8273 1 cycle(s) will be performed\n",
      "Distinct kmers: 248823\n",
      "Kmer count in  2.035456s wall, 1.870000s user + 0.190000s system = 2.060000s CPU (101.2%)\n",
      "Uniq kmers merging in  0.031370s wall, 0.020000s user + 0.010000s system = 0.030000s CPU (95.6%)\n",
      "Kmers branching in  0.355163s wall, 0.360000s user + 0.000000s system = 0.360000s CPU (101.4%)\n",
      "Kmer: 59 Graph size: 248823 Contigs in: 187\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  1.461680s wall, 1.470000s user + 0.010000s system = 1.480000s CPU (101.3%)\n",
      "Kmers in multiple/single contigs: 0 132240\n",
      "Fragments before: 379 23211\n",
      "Fragments after: 379 23211\n",
      "New seeds: 0\n",
      "New seeds in  0.313900s wall, 0.310000s user + 0.010000s system = 0.320000s CPU (101.9%)\n",
      "Fragments before: 372 20155\n",
      "Fragments after: 356 20155\n",
      "Connectors: 16 Extenders: 340\n",
      "Connections and extensions in  1.862764s wall, 1.850000s user + 0.030000s system = 1.880000s CPU (100.9%)\n",
      "Contigs out: 171 Genome: 4769325 N50: 73973 L50: 20\n",
      "Assembled in  3.642103s wall, 3.640000s user + 0.050000s system = 3.690000s CPU (101.3%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 169175\n",
      "Reads for connection: 121706\n",
      "Internal reads: 11332\n",
      "Reads cleaned in  6.176589s wall, 6.100000s user + 0.160000s system = 6.260000s CPU (101.4%)\n",
      "\n",
      "Kmer len: 65\n",
      "Raw kmers: 5913547 Memory needed (GB): 0.22708 Memory available (GB): 29.8219 1 cycle(s) will be performed\n",
      "Distinct kmers: 218600\n",
      "Kmer count in  1.911287s wall, 1.840000s user + 0.100000s system = 1.940000s CPU (101.5%)\n",
      "Uniq kmers merging in  0.017964s wall, 0.000000s user + 0.010000s system = 0.010000s CPU (55.7%)\n",
      "Kmers branching in  0.372212s wall, 0.380000s user + 0.000000s system = 0.380000s CPU (102.1%)\n",
      "Kmer: 65 Graph size: 218600 Contigs in: 171\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  1.654260s wall, 1.660000s user + 0.020000s system = 1.680000s CPU (101.6%)\n",
      "Kmers in multiple/single contigs: 0 118001\n",
      "Fragments before: 355 22426\n",
      "Fragments after: 355 22426\n",
      "New seeds: 0\n",
      "New seeds in  0.267556s wall, 0.270000s user + 0.000000s system = 0.270000s CPU (100.9%)\n",
      "Fragments before: 339 18977\n",
      "Fragments after: 327 18977\n",
      "Connectors: 12 Extenders: 315\n",
      "Connections and extensions in  1.911049s wall, 1.910000s user + 0.010000s system = 1.920000s CPU (100.5%)\n",
      "Contigs out: 159 Genome: 4770801 N50: 105396 L50: 17\n",
      "Assembled in  3.836121s wall, 3.850000s user + 0.030000s system = 3.880000s CPU (101.1%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 144165\n",
      "Reads for connection: 99386\n",
      "Internal reads: 10902\n",
      "Reads cleaned in  5.937692s wall, 5.820000s user + 0.180000s system = 6.000000s CPU (101.0%)\n",
      "\n",
      "Kmer len: 71\n",
      "Raw kmers: 4194107 Memory needed (GB): 0.161054 Memory available (GB): 29.8161 1 cycle(s) will be performed\n",
      "Distinct kmers: 187227\n",
      "Kmer count in  1.368708s wall, 1.350000s user + 0.030000s system = 1.380000s CPU (100.8%)\n",
      "Uniq kmers merging in  0.014919s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (67.0%)\n",
      "Kmers branching in  0.315496s wall, 0.320000s user + 0.000000s system = 0.320000s CPU (101.4%)\n",
      "Kmer: 71 Graph size: 187227 Contigs in: 159\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  1.672312s wall, 1.690000s user + 0.020000s system = 1.710000s CPU (102.3%)\n",
      "Kmers in multiple/single contigs: 0 103845\n",
      "Fragments before: 323 21822\n",
      "Fragments after: 323 21822\n",
      "New seeds: 0\n",
      "New seeds in  0.231468s wall, 0.230000s user + 0.000000s system = 0.230000s CPU (99.4%)\n",
      "Fragments before: 306 18184\n",
      "Fragments after: 293 18184\n",
      "Connectors: 13 Extenders: 280\n",
      "Connections and extensions in  1.887723s wall, 1.890000s user + 0.020000s system = 1.910000s CPU (101.2%)\n",
      "Contigs out: 142 Genome: 4771879 N50: 105396 L50: 17\n",
      "Assembled in  3.794440s wall, 3.810000s user + 0.040000s system = 3.850000s CPU (101.5%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 119202\n",
      "Reads for connection: 77522\n",
      "Internal reads: 10292\n",
      "Reads cleaned in  5.361536s wall, 5.310000s user + 0.120000s system = 5.430000s CPU (101.3%)\n",
      "\n",
      "Kmer len: 79\n",
      "Raw kmers: 2527329 Memory needed (GB): 0.0970494 Memory available (GB): 29.8113 1 cycle(s) will be performed\n",
      "Distinct kmers: 145899\n",
      "Kmer count in  0.950861s wall, 0.950000s user + 0.020000s system = 0.970000s CPU (102.0%)\n",
      "Uniq kmers merging in  0.012870s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (77.7%)\n",
      "Kmers branching in  0.232158s wall, 0.220000s user + 0.010000s system = 0.230000s CPU (99.1%)\n",
      "Kmer: 79 Graph size: 145899 Contigs in: 142\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  1.560887s wall, 1.560000s user + 0.030000s system = 1.590000s CPU (101.9%)\n",
      "Kmers in multiple/single contigs: 0 85502\n",
      "Fragments before: 286 22269\n",
      "Fragments after: 286 22269\n",
      "New seeds: 2\n",
      "New seeds in  0.188167s wall, 0.180000s user + 0.000000s system = 0.180000s CPU (95.7%)\n",
      "Fragments before: 279 18666\n",
      "Fragments after: 266 18666\n",
      "Connectors: 13 Extenders: 253\n",
      "Connections and extensions in  1.920256s wall, 1.890000s user + 0.060000s system = 1.950000s CPU (101.5%)\n",
      "Contigs out: 130 Genome: 4774327 N50: 106040 L50: 17\n",
      "Assembled in  3.671486s wall, 3.640000s user + 0.090000s system = 3.730000s CPU (101.6%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 109127\n",
      "Reads for connection: 69894\n",
      "Internal reads: 10034\n",
      "Reads cleaned in  5.069015s wall, 5.030000s user + 0.120000s system = 5.150000s CPU (101.6%)\n",
      "\n",
      "Kmer len: 85\n",
      "Raw kmers: 1676505 Memory needed (GB): 0.0643778 Memory available (GB): 29.807 1 cycle(s) will be performed\n",
      "Distinct kmers: 118107\n",
      "Kmer count in  0.709077s wall, 0.700000s user + 0.020000s system = 0.720000s CPU (101.5%)\n",
      "Uniq kmers merging in  0.008472s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (118.0%)\n",
      "Kmers branching in  0.196295s wall, 0.200000s user + 0.000000s system = 0.200000s CPU (101.9%)\n",
      "Kmer: 85 Graph size: 118107 Contigs in: 130\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  1.496529s wall, 1.510000s user + 0.010000s system = 1.520000s CPU (101.6%)\n",
      "Kmers in multiple/single contigs: 0 71912\n",
      "Fragments before: 246 20242\n",
      "Fragments after: 246 20242\n",
      "New seeds: 1\n",
      "New seeds in  0.143438s wall, 0.130000s user + 0.010000s system = 0.140000s CPU (97.6%)\n",
      "Fragments before: 251 17077\n",
      "Fragments after: 234 17077\n",
      "Connectors: 17 Extenders: 217\n",
      "Connections and extensions in  1.916697s wall, 1.900000s user + 0.040000s system = 1.940000s CPU (101.2%)\n",
      "Contigs out: 113 Genome: 4777330 N50: 107855 L50: 16\n",
      "Assembled in  3.558556s wall, 3.550000s user + 0.060000s system = 3.610000s CPU (101.4%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 101153\n",
      "Reads for connection: 64926\n",
      "Internal reads: 9304\n",
      "Reads cleaned in  4.253377s wall, 4.170000s user + 0.130000s system = 4.300000s CPU (101.1%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 21\n",
      "Connected: 17787 ambiguously connected: 10195 from 32463 mate pairs\n",
      "Connect pairs in  75.460771s wall, 75.690000s user + 0.450000s system = 76.140000s CPU (100.9%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 27\n",
      "Connected: 1263 ambiguously connected: 8835 from 10195 mate pairs\n",
      "Connect pairs in  14.446111s wall, 14.570000s user + 0.130000s system = 14.700000s CPU (101.8%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 33\n",
      "Connected: 646 ambiguously connected: 8132 from 8835 mate pairs\n",
      "Connect pairs in  9.511031s wall, 9.560000s user + 0.110000s system = 9.670000s CPU (101.7%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 39\n",
      "Connected: 520 ambiguously connected: 7582 from 8132 mate pairs\n",
      "Connect pairs in  8.141888s wall, 8.190000s user + 0.080000s system = 8.270000s CPU (101.6%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 47\n",
      "Connected: 959 ambiguously connected: 6571 from 7582 mate pairs\n",
      "Connect pairs in  9.572837s wall, 9.580000s user + 0.090000s system = 9.670000s CPU (101.0%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 53\n",
      "Connected: 150 ambiguously connected: 6325 from 6571 mate pairs\n",
      "Connect pairs in  6.484823s wall, 6.530000s user + 0.070000s system = 6.600000s CPU (101.8%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 59\n",
      "Connected: 376 ambiguously connected: 5867 from 6325 mate pairs\n",
      "Connect pairs in  6.982131s wall, 7.000000s user + 0.060000s system = 7.060000s CPU (101.1%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 65\n",
      "Connected: 216 ambiguously connected: 5519 from 5867 mate pairs\n",
      "Connect pairs in  7.030612s wall, 7.100000s user + 0.070000s system = 7.170000s CPU (102.0%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 71\n",
      "Connected: 91 ambiguously connected: 5193 from 5519 mate pairs\n",
      "Connect pairs in  6.865769s wall, 6.910000s user + 0.030000s system = 6.940000s CPU (101.1%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 79\n",
      "Connected: 277 ambiguously connected: 4648 from 5193 mate pairs\n",
      "Connect pairs in  6.730569s wall, 6.750000s user + 0.040000s system = 6.790000s CPU (100.9%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 85\n",
      "Connected: 399 ambiguously connected: 3242 from 4648 mate pairs\n",
      "Connect pairs in  4.954720s wall, 5.000000s user + 0.050000s system = 5.050000s CPU (101.9%)\n",
      "Totally connected: 22684\n",
      "Added notconnected: 6483\n",
      "\n",
      "Kmer len: 105\n",
      "Raw kmers: 7033945 Memory needed (GB): 0.337629 Memory available (GB): 29.8025 1 cycle(s) will be performed\n",
      "Distinct kmers: 133347\n",
      "Kmer count in  2.299254s wall, 2.230000s user + 0.100000s system = 2.330000s CPU (101.3%)\n",
      "Uniq kmers merging in  0.021230s wall, 0.020000s user + 0.010000s system = 0.030000s CPU (141.3%)\n",
      "Kmers branching in  0.244101s wall, 0.240000s user + 0.000000s system = 0.240000s CPU (98.3%)\n",
      "Kmer: 105 Graph size: 133347 Contigs in: 113\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  1.608115s wall, 1.610000s user + 0.030000s system = 1.640000s CPU (102.0%)\n",
      "Kmers in multiple/single contigs: 0 104726\n",
      "Fragments before: 194 21067\n",
      "Fragments after: 194 21067\n",
      "New seeds: 1\n",
      "New seeds in  0.113110s wall, 0.100000s user + 0.010000s system = 0.110000s CPU (97.3%)\n",
      "Fragments before: 212 19285\n",
      "Fragments after: 189 19285\n",
      "Connectors: 23 Extenders: 166\n",
      "Connections and extensions in  1.967022s wall, 1.970000s user + 0.010000s system = 1.980000s CPU (100.7%)\n",
      "Contigs out: 86 Genome: 4780559 N50: 119488 L50: 14\n",
      "Assembled in  3.690453s wall, 3.690000s user + 0.050000s system = 3.740000s CPU (101.3%)\n",
      "\n",
      "\n",
      "Kmer len: 181\n",
      "Raw kmers: 4618655 Memory needed (GB): 0.310374 Memory available (GB): 29.797 1 cycle(s) will be performed\n",
      "Distinct kmers: 126177\n",
      "Kmer count in  1.817539s wall, 1.710000s user + 0.130000s system = 1.840000s CPU (101.2%)\n",
      "Uniq kmers merging in  0.013246s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (75.5%)\n",
      "Kmers branching in  0.262403s wall, 0.260000s user + 0.010000s system = 0.270000s CPU (102.9%)\n",
      "Kmer: 181 Graph size: 126177 Contigs in: 86\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  1.838370s wall, 1.840000s user + 0.030000s system = 1.870000s CPU (101.7%)\n",
      "Kmers in multiple/single contigs: 0 95503\n",
      "Fragments before: 132 20662\n",
      "Fragments after: 132 20662\n",
      "New seeds: 0\n",
      "New seeds in  0.117787s wall, 0.110000s user + 0.000000s system = 0.110000s CPU (93.4%)\n",
      "Fragments before: 151 19684\n",
      "Fragments after: 128 19684\n",
      "Connectors: 23 Extenders: 105\n",
      "Connections and extensions in  2.124773s wall, 2.130000s user + 0.020000s system = 2.150000s CPU (101.2%)\n",
      "Contigs out: 56 Genome: 4783416 N50: 133113 L50: 11\n",
      "Assembled in  4.083048s wall, 4.080000s user + 0.050000s system = 4.130000s CPU (101.1%)\n",
      "\n",
      "\n",
      "Kmer len: 259\n",
      "Raw kmers: 2955718 Memory needed (GB): 0.283749 Memory available (GB): 29.7898 1 cycle(s) will be performed\n",
      "Distinct kmers: 115683\n",
      "Kmer count in  1.484669s wall, 1.390000s user + 0.120000s system = 1.510000s CPU (101.7%)\n",
      "Uniq kmers merging in  0.037306s wall, 0.020000s user + 0.020000s system = 0.040000s CPU (107.2%)\n",
      "Kmers branching in  0.264569s wall, 0.260000s user + 0.000000s system = 0.260000s CPU (98.3%)\n",
      "Kmer: 259 Graph size: 115683 Contigs in: 56\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  2.027958s wall, 2.020000s user + 0.050000s system = 2.070000s CPU (102.1%)\n",
      "Kmers in multiple/single contigs: 0 89688\n",
      "Fragments before: 100 16279\n",
      "Fragments after: 100 16279\n",
      "New seeds: 0\n",
      "New seeds in  0.114529s wall, 0.110000s user + 0.010000s system = 0.120000s CPU (104.8%)\n",
      "Fragments before: 100 14986\n",
      "Fragments after: 97 14986\n",
      "Connectors: 3 Extenders: 94\n",
      "Connections and extensions in  2.125644s wall, 2.090000s user + 0.050000s system = 2.140000s CPU (100.7%)\n",
      "Contigs out: 53 Genome: 4784264 N50: 143835 L50: 9\n",
      "Assembled in  4.269999s wall, 4.220000s user + 0.110000s system = 4.330000s CPU (101.4%)\n",
      "\n",
      "DONE\n",
      "\tCommand being timed: \"skesa --cores 1 --fastq input/SRR1952908_1.fastq,input/SRR1952908_2.fastq --contigs_out SRR1952908-full-assembly.fasta --vector_percent 1\"\n",
      "\tUser time (seconds): 518.95\n",
      "\tSystem time (seconds): 18.53\n",
      "\tPercent of CPU this job got: 101%\n",
      "\tElapsed (wall clock) time (h:mm:ss or m:ss): 8:50.47\n",
      "\tAverage shared text size (kbytes): 0\n",
      "\tAverage unshared data size (kbytes): 0\n",
      "\tAverage stack size (kbytes): 0\n",
      "\tAverage total size (kbytes): 0\n",
      "\tMaximum resident set size (kbytes): 5550612\n",
      "\tAverage resident set size (kbytes): 0\n",
      "\tMajor (requiring I/O) page faults: 0\n",
      "\tMinor (reclaiming a frame) page faults: 9773415\n",
      "\tVoluntary context switches: 467623\n",
      "\tInvoluntary context switches: 611\n",
      "\tSwaps: 0\n",
      "\tFile system inputs: 0\n",
      "\tFile system outputs: 9352\n",
      "\tSocket messages sent: 0\n",
      "\tSocket messages received: 0\n",
      "\tSignals delivered: 0\n",
      "\tPage size (bytes): 4096\n",
      "\tExit status: 0\n"
     ]
    }
   ],
   "source": [
    "conda run --name skesa /usr/bin/time -v skesa --cores 1 \\\n",
    "    --fastq input/SRR1952908_1.fastq,input/SRR1952908_2.fastq \\\n",
    "    --contigs_out SRR1952908-full-assembly.fasta --vector_percent 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Great. Let's look at the assembled genome and run RGI on it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rw-r--r-- 1 apetkau grp_apetkau 4.6M Jan 16 20:18 SRR1952908-full-assembly.fasta\n"
     ]
    }
   ],
   "source": [
    "ls -lrth SRR1952908-full-assembly.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "~/workspace/kmer-analysis-test/rgi-full-assembly-out ~/workspace/kmer-analysis-test\n",
      "\tCommand being timed: \"rgi main -i ../SRR1952908-full-assembly.fasta -o out -n 1\"\n",
      "\tUser time (seconds): 236.39\n",
      "\tSystem time (seconds): 4.84\n",
      "\tPercent of CPU this job got: 109%\n",
      "\tElapsed (wall clock) time (h:mm:ss or m:ss): 3:40.13\n",
      "\tAverage shared text size (kbytes): 0\n",
      "\tAverage unshared data size (kbytes): 0\n",
      "\tAverage stack size (kbytes): 0\n",
      "\tAverage total size (kbytes): 0\n",
      "\tMaximum resident set size (kbytes): 141668\n",
      "\tAverage resident set size (kbytes): 0\n",
      "\tMajor (requiring I/O) page faults: 0\n",
      "\tMinor (reclaiming a frame) page faults: 254807\n",
      "\tVoluntary context switches: 370\n",
      "\tInvoluntary context switches: 217658\n",
      "\tSwaps: 0\n",
      "\tFile system inputs: 0\n",
      "\tFile system outputs: 204216\n",
      "\tSocket messages sent: 0\n",
      "\tSocket messages received: 0\n",
      "\tSignals delivered: 0\n",
      "\tPage size (bytes): 4096\n",
      "\tExit status: 0\n",
      "~/workspace/kmer-analysis-test\n"
     ]
    }
   ],
   "source": [
    "mkdir rgi-full-assembly-out\n",
    "pushd rgi-full-assembly-out\n",
    "conda run --name rgi-4.2.2 /usr/bin/time -v rgi main -i ../SRR1952908-full-assembly.fasta -o out -n 1\n",
    "popd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ugd\n",
      "basS\n",
      "MdtK\n",
      "acrD\n",
      "PmrF\n",
      "TEM-57\n",
      "tet(C)\n",
      "kdpE\n",
      "patA\n",
      "bacA\n",
      "Escherichia coli acrA\n",
      "acrB\n",
      "golS\n",
      "mdsA\n",
      "adeF\n",
      "mdsC\n",
      "cpxA\n",
      "emrR\n",
      "emrA\n",
      "emrB\n",
      "CRP\n",
      "mdtC\n",
      "sdiA\n",
      "AAC(6')-Iaa\n",
      "marA\n",
      "OKP-B-10\n",
      "mdtM\n",
      "Escherichia coli mdfA\n",
      "msbA\n",
      "mef(B)\n",
      "sul3\n",
      "qacH\n",
      "aadA13\n",
      "Salmonella enterica cmlA\n",
      "aadA\n",
      "aadA9\n",
      "mdtB\n",
      "mdtC\n",
      "baeR\n"
     ]
    }
   ],
   "source": [
    "grep 'protein homolog model' rgi-full-assembly-out/out.txt | cut -f 9 | tee rgi-full-assembly-out/genes.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Scenario 2: Prefilter reads with kat, assemble with SKESA, run RGI\n",
    "\n",
    "## 2.1: Prefilter reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Kmer Analysis Toolkit (KAT) V2.4.2\n",
      "\n",
      "Running KAT in filter sequence mode\n",
      "-----------------------------------\n",
      "\n",
      "Input 50320 is a sequence file.  Counting kmers for input 50320 (db/nucleotide_fasta_protein_homolog_model.fasta) ... done.  Time taken: 0.6s\n",
      "\n",
      "Filtering sequences ...\n",
      "Processed 100000 pairs\n",
      "Processed 200000 pairs\n",
      "Processed 300000 pairs\n",
      "Processed 400000 pairs\n",
      "Processed 500000 pairs\n",
      "Processed 600000 pairs\n",
      "Processed 700000 pairs\n",
      "Processed 800000 pairs\n",
      "Processed 900000 pairs\n",
      "Processed 1000000 pairs\n",
      "Processed 1100000 pairs\n",
      "Processed 1200000 pairs\n",
      "Processed 1300000 pairs\n",
      "Processed 1400000 pairs\n",
      "Processed 1500000 pairs\n",
      "Processed 1600000 pairs\n",
      "Processed 1700000 pairs\n",
      "Processed 1800000 pairs\n",
      "Processed 1900000 pairs\n",
      "Processed 2000000 pairs\n",
      "Processed 2100000 pairs\n",
      "Finished filtering.  Time taken: 109.4s\n",
      "\n",
      "Found 10204 / 2147046 to keep\n",
      "\n",
      "KAT filter seq completed.\n",
      "Total runtime: 110.0s\n",
      "\n",
      "\tCommand being timed: \"kat filter seq -t 1 -o SRR1952908.kat -m 31 --seq input/SRR1952908_1.fastq --seq2 input/SRR1952908_2.fastq db/nucleotide_fasta_protein_homolog_model.fasta\"\n",
      "\tUser time (seconds): 108.71\n",
      "\tSystem time (seconds): 1.70\n",
      "\tPercent of CPU this job got: 100%\n",
      "\tElapsed (wall clock) time (h:mm:ss or m:ss): 1:50.04\n",
      "\tAverage shared text size (kbytes): 0\n",
      "\tAverage unshared data size (kbytes): 0\n",
      "\tAverage stack size (kbytes): 0\n",
      "\tAverage total size (kbytes): 0\n",
      "\tMaximum resident set size (kbytes): 843704\n",
      "\tAverage resident set size (kbytes): 0\n",
      "\tMajor (requiring I/O) page faults: 0\n",
      "\tMinor (reclaiming a frame) page faults: 314859\n",
      "\tVoluntary context switches: 6\n",
      "\tInvoluntary context switches: 141\n",
      "\tSwaps: 0\n",
      "\tFile system inputs: 0\n",
      "\tFile system outputs: 9472\n",
      "\tSocket messages sent: 0\n",
      "\tSocket messages received: 0\n",
      "\tSignals delivered: 0\n",
      "\tPage size (bytes): 4096\n",
      "\tExit status: 0\n"
     ]
    }
   ],
   "source": [
    "conda run --name kat /usr/bin/time -v kat filter seq -t 1 -o SRR1952908.kat -m 31 \\\n",
    "    --seq input/SRR1952908_1.fastq --seq2 input/SRR1952908_2.fastq \\\n",
    "    db/nucleotide_fasta_protein_homolog_model.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rw-r--r-- 1 apetkau grp_apetkau 2.3M Jan 16 17:30 SRR1952908.kat.in.R2.fastq\n",
      "-rw-r--r-- 1 apetkau grp_apetkau 2.3M Jan 16 17:30 SRR1952908.kat.in.R1.fastq\n"
     ]
    }
   ],
   "source": [
    "ls -lrth *kat*.fastq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.2 Assemble extracted reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "skesa --cores 1 --fastq SRR1952908.kat.in.R1.fastq,SRR1952908.kat.in.R2.fastq --contigs_out SRR1952908.kat.contigs.fasta --vector_percent 1 \n",
      "\n",
      "Total mates: 20408 Paired reads: 10204\n",
      "Reads acquired in  0.063515s wall, 0.050000s user + 0.000000s system = 0.050000s CPU (78.7%)\n",
      "Adapters clip is disabled\n",
      "\n",
      "Kmer len: 21\n",
      "Raw kmers: 1607679 Memory needed (GB): 0.0308674 Memory available (GB): 29.9988 1 cycle(s) will be performed\n",
      "Distinct kmers: 33311\n",
      "Kmer count in  0.421162s wall, 0.400000s user + 0.020000s system = 0.420000s CPU (99.7%)\n",
      "Uniq kmers merging in  0.004130s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (242.2%)\n",
      "WARNING: --min_count changed from 2 to 4 because of high coverage for genome size 9698\n",
      "WARNING: --max_kmer_count 10 to 21 because of high coverage for genome size 9698\n",
      "Kmers branching in  0.044778s wall, 0.040000s user + 0.010000s system = 0.050000s CPU (111.7%)\n",
      "\n",
      "Average read length: 99\n",
      "Genome size estimate: 9918\n",
      "\n",
      "Kmer: 21 Graph size: 25504 Contigs in: 0\n",
      "Valley: 19\n",
      "\n",
      "Mark used kmers in  0.000007s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmers in multiple/single contigs: 0 0\n",
      "Fragments before: 33 22459\n",
      "Fragments after: 33 22459\n",
      "New seeds: 29\n",
      "New seeds in  0.043043s wall, 0.050000s user + 0.000000s system = 0.050000s CPU (116.2%)\n",
      "Fragments before: 58 1218\n",
      "Fragments after: 58 1218\n",
      "Connectors: 0 Extenders: 58\n",
      "Connections and extensions in  0.011847s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (84.4%)\n",
      "Contigs out: 29 Genome: 21281 N50: 1385 L50: 4\n",
      "Assembled in  0.055422s wall, 0.060000s user + 0.000000s system = 0.060000s CPU (108.3%)\n",
      "\n",
      "\n",
      "Kmer len: 59\n",
      "Raw kmers: 832819 Memory needed (GB): 0.0239852 Memory available (GB): 29.9983 1 cycle(s) will be performed\n",
      "Distinct kmers: 22029\n",
      "Kmer count in  0.305944s wall, 0.290000s user + 0.010000s system = 0.300000s CPU (98.1%)\n",
      "Uniq kmers merging in  0.002918s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "\n",
      "Average count: 41.2572 Max kmer: 59\n",
      "Connected: 7966 ambiguously connected: 1804 from 10204 mate pairs\n",
      "Connect pairs in  12.401576s wall, 12.500000s user + 0.120000s system = 12.620000s CPU (101.8%)\n",
      "\n",
      "N50 for inserts: 261\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 7368\n",
      "Reads for connection: 5364\n",
      "Internal reads: 548\n",
      "Reads cleaned in  0.149493s wall, 0.150000s user + 0.000000s system = 0.150000s CPU (100.3%)\n",
      "\n",
      "Kmer len: 25\n",
      "Raw kmers: 553307 Memory needed (GB): 0.0106235 Memory available (GB): 29.999 1 cycle(s) will be performed\n",
      "Distinct kmers: 16730\n",
      "Kmer count in  0.178685s wall, 0.180000s user + 0.010000s system = 0.190000s CPU (106.3%)\n",
      "Uniq kmers merging in  0.002294s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmers branching in  0.036637s wall, 0.040000s user + 0.000000s system = 0.040000s CPU (109.2%)\n",
      "Kmer: 25 Graph size: 16730 Contigs in: 29\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.009038s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (110.6%)\n",
      "Kmers in multiple/single contigs: 0 12551\n",
      "Fragments before: 79 3281\n",
      "Fragments after: 79 3281\n",
      "New seeds: 2\n",
      "New seeds in  0.024604s wall, 0.030000s user + 0.000000s system = 0.030000s CPU (121.9%)\n",
      "Fragments before: 62 1581\n",
      "Fragments after: 62 1581\n",
      "Connectors: 0 Extenders: 62\n",
      "Connections and extensions in  0.027103s wall, 0.020000s user + 0.000000s system = 0.020000s CPU (73.8%)\n",
      "Contigs out: 31 Genome: 21471 N50: 1385 L50: 4\n",
      "Assembled in  0.061150s wall, 0.060000s user + 0.000000s system = 0.060000s CPU (98.1%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 7368\n",
      "Reads for connection: 5364\n",
      "Internal reads: 548\n",
      "Reads cleaned in  0.140535s wall, 0.150000s user + 0.000000s system = 0.150000s CPU (106.7%)\n",
      "\n",
      "Kmer len: 29\n",
      "Raw kmers: 523835 Memory needed (GB): 0.0100576 Memory available (GB): 29.9987 1 cycle(s) will be performed\n",
      "Distinct kmers: 16358\n",
      "Kmer count in  0.177127s wall, 0.170000s user + 0.010000s system = 0.180000s CPU (101.6%)\n",
      "Uniq kmers merging in  0.002500s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmers branching in  0.030366s wall, 0.030000s user + 0.000000s system = 0.030000s CPU (98.8%)\n",
      "Kmer: 29 Graph size: 16358 Contigs in: 31\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.013278s wall, 0.020000s user + 0.000000s system = 0.020000s CPU (150.6%)\n",
      "Kmers in multiple/single contigs: 0 12455\n",
      "Fragments before: 79 3091\n",
      "Fragments after: 79 3091\n",
      "New seeds: 0\n",
      "New seeds in  0.023398s wall, 0.020000s user + 0.000000s system = 0.020000s CPU (85.5%)\n",
      "Fragments before: 62 1680\n",
      "Fragments after: 61 1680\n",
      "Connectors: 1 Extenders: 60\n",
      "Connections and extensions in  0.026612s wall, 0.030000s user + 0.000000s system = 0.030000s CPU (112.7%)\n",
      "Contigs out: 30 Genome: 21514 N50: 1385 L50: 4\n",
      "Assembled in  0.063688s wall, 0.070000s user + 0.000000s system = 0.070000s CPU (109.9%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 7142\n",
      "Reads for connection: 5104\n",
      "Internal reads: 580\n",
      "Reads cleaned in  0.108943s wall, 0.110000s user + 0.000000s system = 0.110000s CPU (101.0%)\n",
      "\n",
      "Kmer len: 31\n",
      "Raw kmers: 493367 Memory needed (GB): 0.00947265 Memory available (GB): 29.9985 1 cycle(s) will be performed\n",
      "Distinct kmers: 15972\n",
      "Kmer count in  0.171107s wall, 0.170000s user + 0.000000s system = 0.170000s CPU (99.4%)\n",
      "Uniq kmers merging in  0.002453s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmers branching in  0.031414s wall, 0.040000s user + 0.000000s system = 0.040000s CPU (127.3%)\n",
      "Kmer: 31 Graph size: 15972 Contigs in: 30\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.013786s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (72.5%)\n",
      "Kmers in multiple/single contigs: 0 12207\n",
      "Fragments before: 77 2974\n",
      "Fragments after: 77 2974\n",
      "New seeds: 0\n",
      "New seeds in  0.022755s wall, 0.030000s user + 0.000000s system = 0.030000s CPU (131.8%)\n",
      "Fragments before: 60 1649\n",
      "Fragments after: 60 1649\n",
      "Connectors: 0 Extenders: 60\n",
      "Connections and extensions in  0.024681s wall, 0.020000s user + 0.000000s system = 0.020000s CPU (81.0%)\n",
      "Contigs out: 30 Genome: 21514 N50: 1385 L50: 4\n",
      "Assembled in  0.061616s wall, 0.060000s user + 0.000000s system = 0.060000s CPU (97.4%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 7142\n",
      "Reads for connection: 5104\n",
      "Internal reads: 580\n",
      "Reads cleaned in  0.125667s wall, 0.130000s user + 0.000000s system = 0.130000s CPU (103.4%)\n",
      "\n",
      "Kmer len: 35\n",
      "Raw kmers: 464799 Memory needed (GB): 0.0133862 Memory available (GB): 29.9982 1 cycle(s) will be performed\n",
      "Distinct kmers: 15577\n",
      "Kmer count in  0.198249s wall, 0.200000s user + 0.010000s system = 0.210000s CPU (105.9%)\n",
      "Uniq kmers merging in  0.002944s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmers branching in  0.036919s wall, 0.040000s user + 0.000000s system = 0.040000s CPU (108.3%)\n",
      "Kmer: 35 Graph size: 15577 Contigs in: 30\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.014080s wall, 0.010000s user + 0.010000s system = 0.020000s CPU (142.0%)\n",
      "Kmers in multiple/single contigs: 0 11952\n",
      "Fragments before: 66 2393\n",
      "Fragments after: 66 2393\n",
      "New seeds: 0\n",
      "New seeds in  0.017143s wall, 0.020000s user + 0.000000s system = 0.020000s CPU (116.7%)\n",
      "Fragments before: 58 1652\n",
      "Fragments after: 57 1652\n",
      "Connectors: 1 Extenders: 56\n",
      "Connections and extensions in  0.028915s wall, 0.030000s user + 0.000000s system = 0.030000s CPU (103.8%)\n",
      "Contigs out: 29 Genome: 21549 N50: 1385 L50: 4\n",
      "Assembled in  0.060516s wall, 0.060000s user + 0.010000s system = 0.070000s CPU (115.7%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 7142\n",
      "Reads for connection: 4830\n",
      "Internal reads: 671\n",
      "Reads cleaned in  0.129343s wall, 0.130000s user + 0.000000s system = 0.130000s CPU (100.5%)\n",
      "\n",
      "Kmer len: 39\n",
      "Raw kmers: 436231 Memory needed (GB): 0.0125635 Memory available (GB): 29.9978 1 cycle(s) will be performed\n",
      "Distinct kmers: 15127\n",
      "Kmer count in  0.177935s wall, 0.170000s user + 0.000000s system = 0.170000s CPU (95.5%)\n",
      "Uniq kmers merging in  0.002184s wall, 0.000000s user + 0.010000s system = 0.010000s CPU (457.8%)\n",
      "Kmers branching in  0.037562s wall, 0.040000s user + 0.000000s system = 0.040000s CPU (106.5%)\n",
      "Kmer: 39 Graph size: 15127 Contigs in: 29\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.014027s wall, 0.020000s user + 0.000000s system = 0.020000s CPU (142.6%)\n",
      "Kmers in multiple/single contigs: 0 11740\n",
      "Fragments before: 59 1960\n",
      "Fragments after: 59 1960\n",
      "New seeds: 0\n",
      "New seeds in  0.020701s wall, 0.020000s user + 0.000000s system = 0.020000s CPU (96.6%)\n",
      "Fragments before: 55 1605\n",
      "Fragments after: 55 1605\n",
      "Connectors: 0 Extenders: 55\n",
      "Connections and extensions in  0.028804s wall, 0.030000s user + 0.000000s system = 0.030000s CPU (104.2%)\n",
      "Contigs out: 29 Genome: 21549 N50: 1385 L50: 4\n",
      "Assembled in  0.063907s wall, 0.070000s user + 0.000000s system = 0.070000s CPU (109.5%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 7140\n",
      "Reads for connection: 4830\n",
      "Internal reads: 671\n",
      "Reads cleaned in  0.121870s wall, 0.120000s user + 0.000000s system = 0.120000s CPU (98.5%)\n",
      "\n",
      "Kmer len: 43\n",
      "Raw kmers: 407617 Memory needed (GB): 0.0117394 Memory available (GB): 29.9974 1 cycle(s) will be performed\n",
      "Distinct kmers: 14636\n",
      "Kmer count in  0.171968s wall, 0.170000s user + 0.010000s system = 0.180000s CPU (104.7%)\n",
      "Uniq kmers merging in  0.002330s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmers branching in  0.030944s wall, 0.030000s user + 0.000000s system = 0.030000s CPU (96.9%)\n",
      "Kmer: 43 Graph size: 14636 Contigs in: 29\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.014395s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (69.5%)\n",
      "Kmers in multiple/single contigs: 0 11425\n",
      "Fragments before: 49 1866\n",
      "Fragments after: 49 1866\n",
      "New seeds: 0\n",
      "New seeds in  0.011375s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (87.9%)\n",
      "Fragments before: 54 1778\n",
      "Fragments after: 48 1778\n",
      "Connectors: 6 Extenders: 42\n",
      "Connections and extensions in  0.025427s wall, 0.030000s user + 0.000000s system = 0.030000s CPU (118.0%)\n",
      "Contigs out: 23 Genome: 21967 N50: 1385 L50: 4\n",
      "Assembled in  0.051566s wall, 0.050000s user + 0.000000s system = 0.050000s CPU (97.0%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 5145\n",
      "Reads for connection: 2448\n",
      "Internal reads: 766\n",
      "Reads cleaned in  0.077901s wall, 0.080000s user + 0.000000s system = 0.080000s CPU (102.7%)\n",
      "\n",
      "Kmer len: 47\n",
      "Raw kmers: 273113 Memory needed (GB): 0.00786565 Memory available (GB): 29.9972 1 cycle(s) will be performed\n",
      "Distinct kmers: 13363\n",
      "Kmer count in  0.121301s wall, 0.110000s user + 0.000000s system = 0.110000s CPU (90.7%)\n",
      "Uniq kmers merging in  0.002147s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (465.7%)\n",
      "Kmers branching in  0.032234s wall, 0.030000s user + 0.000000s system = 0.030000s CPU (93.1%)\n",
      "Kmer: 47 Graph size: 13363 Contigs in: 23\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.015754s wall, 0.020000s user + 0.000000s system = 0.020000s CPU (127.0%)\n",
      "Kmers in multiple/single contigs: 0 11053\n",
      "Fragments before: 41 1054\n",
      "Fragments after: 41 1054\n",
      "New seeds: 0\n",
      "New seeds in  0.014477s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (69.1%)\n",
      "Fragments before: 41 1054\n",
      "Fragments after: 41 1054\n",
      "Connectors: 0 Extenders: 41\n",
      "Connections and extensions in  0.025670s wall, 0.030000s user + 0.000000s system = 0.030000s CPU (116.9%)\n",
      "Contigs out: 23 Genome: 21967 N50: 1385 L50: 4\n",
      "Assembled in  0.056255s wall, 0.060000s user + 0.000000s system = 0.060000s CPU (106.7%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 5145\n",
      "Reads for connection: 2448\n",
      "Internal reads: 765\n",
      "Reads cleaned in  0.072972s wall, 0.070000s user + 0.000000s system = 0.070000s CPU (95.9%)\n",
      "\n",
      "Kmer len: 51\n",
      "Raw kmers: 252533 Memory needed (GB): 0.00727295 Memory available (GB): 29.9968 1 cycle(s) will be performed\n",
      "Distinct kmers: 12825\n",
      "Kmer count in  0.115197s wall, 0.110000s user + 0.000000s system = 0.110000s CPU (95.5%)\n",
      "Uniq kmers merging in  0.002029s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmers branching in  0.030557s wall, 0.030000s user + 0.000000s system = 0.030000s CPU (98.2%)\n",
      "Kmer: 51 Graph size: 12825 Contigs in: 23\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.015413s wall, 0.020000s user + 0.000000s system = 0.020000s CPU (129.8%)\n",
      "Kmers in multiple/single contigs: 0 10692\n",
      "Fragments before: 40 1121\n",
      "Fragments after: 40 1121\n",
      "New seeds: 0\n",
      "New seeds in  0.012385s wall, 0.000000s user + 0.010000s system = 0.010000s CPU (80.7%)\n",
      "Fragments before: 41 1121\n",
      "Fragments after: 40 1121\n",
      "Connectors: 1 Extenders: 39\n",
      "Connections and extensions in  0.023244s wall, 0.020000s user + 0.000000s system = 0.020000s CPU (86.0%)\n",
      "Contigs out: 22 Genome: 22083 N50: 1385 L50: 4\n",
      "Assembled in  0.051403s wall, 0.040000s user + 0.010000s system = 0.050000s CPU (97.3%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 5080\n",
      "Reads for connection: 2340\n",
      "Internal reads: 776\n",
      "Reads cleaned in  0.060440s wall, 0.070000s user + 0.000000s system = 0.070000s CPU (115.8%)\n",
      "\n",
      "Kmer len: 55\n",
      "Raw kmers: 229171 Memory needed (GB): 0.00660012 Memory available (GB): 29.9965 1 cycle(s) will be performed\n",
      "Distinct kmers: 12041\n",
      "Kmer count in  0.099167s wall, 0.090000s user + 0.010000s system = 0.100000s CPU (100.8%)\n",
      "Uniq kmers merging in  0.002011s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmers branching in  0.031672s wall, 0.030000s user + 0.000000s system = 0.030000s CPU (94.7%)\n",
      "Kmer: 55 Graph size: 12041 Contigs in: 22\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.014888s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (67.2%)\n",
      "Kmers in multiple/single contigs: 0 10264\n",
      "Fragments before: 34 898\n",
      "Fragments after: 34 898\n",
      "New seeds: 0\n",
      "New seeds in  0.009834s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (101.7%)\n",
      "Fragments before: 36 898\n",
      "Fragments after: 34 898\n",
      "Connectors: 2 Extenders: 32\n",
      "Connections and extensions in  0.022072s wall, 0.020000s user + 0.000000s system = 0.020000s CPU (90.6%)\n",
      "Contigs out: 20 Genome: 22189 N50: 1826 L50: 3\n",
      "Assembled in  0.047127s wall, 0.040000s user + 0.000000s system = 0.040000s CPU (84.9%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 2977\n",
      "Reads for connection: 1268\n",
      "Internal reads: 474\n",
      "Reads cleaned in  0.055895s wall, 0.060000s user + 0.000000s system = 0.060000s CPU (107.3%)\n",
      "\n",
      "Kmer len: 59\n",
      "Raw kmers: 122326 Memory needed (GB): 0.00352299 Memory available (GB): 29.9963 1 cycle(s) will be performed\n",
      "Distinct kmers: 9166\n",
      "Kmer count in  0.069402s wall, 0.070000s user + 0.000000s system = 0.070000s CPU (100.9%)\n",
      "Uniq kmers merging in  0.001789s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmers branching in  0.026367s wall, 0.030000s user + 0.000000s system = 0.030000s CPU (113.8%)\n",
      "Kmer: 59 Graph size: 9166 Contigs in: 20\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.013868s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (72.1%)\n",
      "Kmers in multiple/single contigs: 0 7800\n",
      "Fragments before: 31 660\n",
      "Fragments after: 31 660\n",
      "New seeds: 0\n",
      "New seeds in  0.007910s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (126.4%)\n",
      "Fragments before: 31 660\n",
      "Fragments after: 31 660\n",
      "Connectors: 0 Extenders: 31\n",
      "Connections and extensions in  0.023247s wall, 0.030000s user + 0.000000s system = 0.030000s CPU (129.0%)\n",
      "Contigs out: 20 Genome: 22189 N50: 1826 L50: 3\n",
      "Assembled in  0.045324s wall, 0.050000s user + 0.000000s system = 0.050000s CPU (110.3%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 2975\n",
      "Reads for connection: 1268\n",
      "Internal reads: 473\n",
      "Reads cleaned in  0.051670s wall, 0.040000s user + 0.010000s system = 0.050000s CPU (96.8%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 21\n",
      "Connected: 361 ambiguously connected: 1 from 634 mate pairs\n",
      "Connect pairs in  0.247733s wall, 0.250000s user + 0.000000s system = 0.250000s CPU (100.9%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 25\n",
      "Connected: 1 ambiguously connected: 0 from 1 mate pairs\n",
      "Connect pairs in  0.006064s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 29\n",
      "Connected: 0 ambiguously connected: 0 from 0 mate pairs\n",
      "Connect pairs in  0.000025s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 31\n",
      "Connected: 0 ambiguously connected: 0 from 0 mate pairs\n",
      "Connect pairs in  0.000024s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 35\n",
      "Connected: 0 ambiguously connected: 0 from 0 mate pairs\n",
      "Connect pairs in  0.000026s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 39\n",
      "Connected: 0 ambiguously connected: 0 from 0 mate pairs\n",
      "Connect pairs in  0.000033s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 43\n",
      "Connected: 0 ambiguously connected: 0 from 0 mate pairs\n",
      "Connect pairs in  0.000027s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 47\n",
      "Connected: 0 ambiguously connected: 0 from 0 mate pairs\n",
      "Connect pairs in  0.000025s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 51\n",
      "Connected: 0 ambiguously connected: 0 from 0 mate pairs\n",
      "Connect pairs in  0.000025s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 55\n",
      "Connected: 0 ambiguously connected: 0 from 0 mate pairs\n",
      "Connect pairs in  0.000025s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 59\n",
      "Connected: 0 ambiguously connected: 0 from 0 mate pairs\n",
      "Connect pairs in  0.000026s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Totally connected: 362\n",
      "Added notconnected: 0\n",
      "\n",
      "Kmer len: 73\n",
      "Raw kmers: 181635 Memory needed (GB): 0.00697478 Memory available (GB): 29.9961 1 cycle(s) will be performed\n",
      "Distinct kmers: 11338\n",
      "Kmer count in  0.093676s wall, 0.090000s user + 0.000000s system = 0.090000s CPU (96.1%)\n",
      "Uniq kmers merging in  0.001920s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmers branching in  0.029568s wall, 0.030000s user + 0.010000s system = 0.040000s CPU (135.3%)\n",
      "Kmer: 73 Graph size: 11338 Contigs in: 20\n",
      "Valley: 13\n",
      "\n",
      "Mark used kmers in  0.005912s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (169.2%)\n",
      "Kmers in multiple/single contigs: 0 9983\n",
      "Fragments before: 8 263\n",
      "Fragments after: 8 263\n",
      "New seeds: 0\n",
      "New seeds in  0.004913s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Fragments before: 40 1141\n",
      "Fragments after: 40 1141\n",
      "Connectors: 0 Extenders: 40\n",
      "Connections and extensions in  0.026348s wall, 0.030000s user + 0.000000s system = 0.030000s CPU (113.9%)\n",
      "Contigs out: 20 Genome: 22189 N50: 1826 L50: 3\n",
      "Assembled in  0.037514s wall, 0.040000s user + 0.000000s system = 0.040000s CPU (106.6%)\n",
      "\n",
      "\n",
      "Kmer len: 167\n",
      "Raw kmers: 108055 Memory needed (GB): 0.0072613 Memory available (GB): 29.9957 1 cycle(s) will be performed\n",
      "Distinct kmers: 8697\n",
      "Kmer count in  0.071768s wall, 0.070000s user + 0.000000s system = 0.070000s CPU (97.5%)\n",
      "Uniq kmers merging in  0.001931s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmers branching in  0.032725s wall, 0.030000s user + 0.000000s system = 0.030000s CPU (91.7%)\n",
      "Kmer: 167 Graph size: 8697 Contigs in: 20\n",
      "Valley: 7\n",
      "\n",
      "Mark used kmers in  0.017003s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (58.8%)\n",
      "Kmers in multiple/single contigs: 0 7655\n",
      "Fragments before: 18 551\n",
      "Fragments after: 18 551\n",
      "New seeds: 0\n",
      "New seeds in  0.006607s wall, 0.010000s user + 0.000000s system = 0.010000s CPU (151.3%)\n",
      "Fragments before: 36 1011\n",
      "Fragments after: 36 1011\n",
      "Connectors: 0 Extenders: 36\n",
      "Connections and extensions in  0.029269s wall, 0.030000s user + 0.000000s system = 0.030000s CPU (102.5%)\n",
      "Contigs out: 18 Genome: 21990 N50: 1826 L50: 3\n",
      "Assembled in  0.053175s wall, 0.050000s user + 0.000000s system = 0.050000s CPU (94.0%)\n",
      "\n",
      "\n",
      "Kmer len: 261\n",
      "Raw kmers: 54924 Memory needed (GB): 0.0052727 Memory available (GB): 29.9952 1 cycle(s) will be performed\n",
      "Distinct kmers: 5733\n",
      "Kmer count in  0.049046s wall, 0.050000s user + 0.000000s system = 0.050000s CPU (101.9%)\n",
      "Uniq kmers merging in  0.001819s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmers branching in  0.026219s wall, 0.030000s user + 0.000000s system = 0.030000s CPU (114.4%)\n",
      "Kmer: 261 Graph size: 5733 Contigs in: 18\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.017483s wall, 0.020000s user + 0.000000s system = 0.020000s CPU (114.4%)\n",
      "Kmers in multiple/single contigs: 0 4990\n",
      "Fragments before: 25 723\n",
      "Fragments after: 25 723\n",
      "New seeds: 0\n",
      "New seeds in  0.007868s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Fragments before: 25 723\n",
      "Fragments after: 25 723\n",
      "Connectors: 0 Extenders: 25\n",
      "Connections and extensions in  0.030172s wall, 0.040000s user + 0.000000s system = 0.040000s CPU (132.6%)\n",
      "Contigs out: 16 Genome: 21577 N50: 1826 L50: 3\n",
      "Assembled in  0.055774s wall, 0.060000s user + 0.000000s system = 0.060000s CPU (107.6%)\n",
      "\n",
      "DONE\n",
      "\tCommand being timed: \"skesa --cores 1 --fastq SRR1952908.kat.in.R1.fastq,SRR1952908.kat.in.R2.fastq --contigs_out SRR1952908.kat.contigs.fasta --vector_percent 1\"\n",
      "\tUser time (seconds): 17.63\n",
      "\tSystem time (seconds): 0.26\n",
      "\tPercent of CPU this job got: 101%\n",
      "\tElapsed (wall clock) time (h:mm:ss or m:ss): 0:17.61\n",
      "\tAverage shared text size (kbytes): 0\n",
      "\tAverage unshared data size (kbytes): 0\n",
      "\tAverage stack size (kbytes): 0\n",
      "\tAverage total size (kbytes): 0\n",
      "\tMaximum resident set size (kbytes): 31000\n",
      "\tAverage resident set size (kbytes): 0\n",
      "\tMajor (requiring I/O) page faults: 0\n",
      "\tMinor (reclaiming a frame) page faults: 32123\n",
      "\tVoluntary context switches: 16031\n",
      "\tInvoluntary context switches: 170\n",
      "\tSwaps: 0\n",
      "\tFile system inputs: 0\n",
      "\tFile system outputs: 48\n",
      "\tSocket messages sent: 0\n",
      "\tSocket messages received: 0\n",
      "\tSignals delivered: 0\n",
      "\tPage size (bytes): 4096\n",
      "\tExit status: 0\n"
     ]
    }
   ],
   "source": [
    "conda run --name skesa /usr/bin/time -v skesa --cores 1 \\\n",
    "    --fastq SRR1952908.kat.in.R1.fastq,SRR1952908.kat.in.R2.fastq \\\n",
    "    --contigs_out SRR1952908.kat.contigs.fasta --vector_percent 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rw-r--r-- 1 apetkau grp_apetkau 22K Jan 16 17:32 SRR1952908.kat.contigs.fasta\n"
     ]
    }
   ],
   "source": [
    "ls -lh *kat*.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.3: Run RGI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "~/workspace/kmer-analysis-test/rgi-kat ~/workspace/kmer-analysis-test\n",
      "\tCommand being timed: \"rgi main -i ../SRR1952908.kat.contigs.fasta -o out -n 1\"\n",
      "\tUser time (seconds): 6.03\n",
      "\tSystem time (seconds): 4.63\n",
      "\tPercent of CPU this job got: 343%\n",
      "\tElapsed (wall clock) time (h:mm:ss or m:ss): 0:03.10\n",
      "\tAverage shared text size (kbytes): 0\n",
      "\tAverage unshared data size (kbytes): 0\n",
      "\tAverage stack size (kbytes): 0\n",
      "\tAverage total size (kbytes): 0\n",
      "\tMaximum resident set size (kbytes): 128820\n",
      "\tAverage resident set size (kbytes): 0\n",
      "\tMajor (requiring I/O) page faults: 0\n",
      "\tMinor (reclaiming a frame) page faults: 164086\n",
      "\tVoluntary context switches: 200\n",
      "\tInvoluntary context switches: 19263\n",
      "\tSwaps: 0\n",
      "\tFile system inputs: 0\n",
      "\tFile system outputs: 11232\n",
      "\tSocket messages sent: 0\n",
      "\tSocket messages received: 0\n",
      "\tSignals delivered: 0\n",
      "\tPage size (bytes): 4096\n",
      "\tExit status: 0\n",
      "~/workspace/kmer-analysis-test\n"
     ]
    }
   ],
   "source": [
    "mkdir rgi-kat\n",
    "pushd rgi-kat\n",
    "conda run --name rgi-4.2.2 /usr/bin/time -v rgi main -i ../SRR1952908.kat.contigs.fasta -o out -n 1\n",
    "popd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MdtK\n",
      "sul3\n",
      "aadA\n",
      "Salmonella enterica cmlA\n",
      "aadA13\n",
      "qacH\n",
      "emrR\n",
      "mef(B)\n",
      "sdiA\n",
      "TEM-57\n",
      "patA\n",
      "AAC(6')-Ib7\n",
      "cpxA\n",
      "mdtC\n",
      "AAC(6')-Iaa\n",
      "golS\n",
      "acrB\n",
      "mdsC\n",
      "adeF\n",
      "mdsA\n"
     ]
    }
   ],
   "source": [
    "grep 'protein homolog model' rgi-kat/out.txt | cut -f 9 | tee rgi-kat/genes.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's compare to scenario 1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True Positives: 19: {'aadA13', 'mdtC', 'patA', 'cpxA', 'acrB', 'adeF', 'sul3', 'qacH', 'Salmonella enterica cmlA', 'emrR', \"AAC(6')-Iaa\", 'TEM-57', 'golS', 'sdiA', 'mdsA', 'MdtK', 'aadA', 'mdsC', 'mef(B)'}\n",
      "False Positives: 1: {\"AAC(6')-Ib7\"}\n",
      "False Negatives: 19: {'msbA', 'CRP', 'baeR', 'OKP-B-10', 'basS', 'kdpE', 'mdtM', 'Escherichia coli mdfA', 'acrD', 'ugd', 'PmrF', 'marA', 'emrB', 'aadA9', 'tet(C)', 'Escherichia coli acrA', 'emrA', 'bacA', 'mdtB'}\n",
      "Sensitivity: 0.50\n"
     ]
    }
   ],
   "source": [
    "print_positives \"rgi-full-assembly-out/genes.txt\" \"rgi-kat/genes.txt\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Scenario 3: Break RGI database to kmers first\n",
    "\n",
    "## 3.1: Run jellyfish"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">gb|GQ343019|+|132-1023|ARO:3002999|CblA-1 [mixed culture bacterium AX_gF3SD01_15] \n",
      "ATGAAAGCATATTTCATCGCCATACTTACCTTATTCACTTGTATAGCTACCGTCGTCCGGGCGCAGCAAATGTCTGAACTTGAAAACCGGATTGACAGTCTGCTCAATGGCAAGAAAGCCACCGTTGGTATAGCCGTATGGACAGACAAAGGAGACATGCTCCGGTATAACGACCATGTACACTTCCCCTTGCTCAGTGTATTCAAATTCCATGTGGCACTGGCCGTACTGGACAAGATGGATAAGCAAAGCATCAGTCTGGACAGCATTGTTTCCATAAAGGCATCCCAAATGCCGCCCAATACCTACAGCCCCCTGCGGAAGAAGTTTCCCGACCAGGATTTCACGATTACGCTTAGGGAACTGATGCAATACAGCATTTCCCAAAGCGACAACAATGCCTGCGACATCTTGATAGAATATGCAGGAGGCATCAAACATATCAACGACTATATCCACCGGTTGAGTATCGACTCCTTCAACCTCTCGGAAACAGAAGACGGCATGCACTCCAGCTTCGAGGCTGTATACCGCAACTGGAGTACTCCTTCCGCTATGGTCCGACTACTGAGAACGGCTGATGAAAAAGAGTTGTTCTCCAACAAGGAGCTGAAAGACTTCTTGTGGCAGACCATGATAGATACTGAAACCGGTGCCAACAAACTGAAAGGTATGTTGCCAGCCAAAACCGTGGTAGGACACAAGACCGGCTCTTCCGACCGCAATGCCGACGGTATGAAAACTGCAGATAATGATGCCGGCCTCGTTATCCTTCCCGACGGCCGGAAATACTACATTGCCGCCTTCGTCATGGACTCATACGAGACGGATGAGGACAATGCGAACATCATCGCCCGCATATCACGCATGGTATATGATGCGATGAGATGA\n"
     ]
    }
   ],
   "source": [
    "head db/nucleotide_fasta_protein_homolog_model.fasta -n 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Scenario 3: RGI BWT\n",
    "\n",
    "## 3.1 Run RGI BWT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "~/workspace/kmer-analysis-test/rgi-bwt-out ~/workspace/kmer-analysis-test\n",
      "[bwa_idx_build] fail to open file '/home/CSCScience.ca/apetkau/miniconda3/envs/rgi-4.2.2/lib/python3.6/site-packages/app/_data/card_reference.fasta' : No such file or directory\n",
      "[E::bwa_idx_load_from_disk] fail to locate the index files\n",
      "\tCommand being timed: \"rgi bwt -1 SRR1952908_1.fastq -2 SRR1952908_2.fastq -a bwa -n 1 -o out-bwt\"\n",
      "\tUser time (seconds): 3.86\n",
      "\tSystem time (seconds): 4.40\n",
      "\tPercent of CPU this job got: 465%\n",
      "\tElapsed (wall clock) time (h:mm:ss or m:ss): 0:01.77\n",
      "\tAverage shared text size (kbytes): 0\n",
      "\tAverage unshared data size (kbytes): 0\n",
      "\tAverage stack size (kbytes): 0\n",
      "\tAverage total size (kbytes): 0\n",
      "\tMaximum resident set size (kbytes): 184568\n",
      "\tAverage resident set size (kbytes): 0\n",
      "\tMajor (requiring I/O) page faults: 32\n",
      "\tMinor (reclaiming a frame) page faults: 87474\n",
      "\tVoluntary context switches: 288\n",
      "\tInvoluntary context switches: 48445\n",
      "\tSwaps: 0\n",
      "\tFile system inputs: 8760\n",
      "\tFile system outputs: 64\n",
      "\tSocket messages sent: 0\n",
      "\tSocket messages received: 0\n",
      "\tSignals delivered: 0\n",
      "\tPage size (bytes): 4096\n",
      "\tExit status: 0\n",
      "~/workspace/kmer-analysis-test\n"
     ]
    }
   ],
   "source": [
    "mkdir rgi-bwt-out\n",
    "pushd rgi-bwt-out\n",
    "conda run --name rgi-4.2.2 /usr/bin/time -v rgi bwt -1 SRR1952908_1.fastq -2 SRR1952908_2.fastq -a bwa -n 1 -o out-bwt\n",
    "popd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Compare results (detected AMR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Isolate ID\tGenotype\tPredicted Phenotype\n",
      "kat.filter.contigs\taadA1, aadA2, blaTEM-57, cmlA1, sul3, tet(A)\tstreptomycin, ampicillin, chloramphenicol, sulfisoxazole, tetracycline\n"
     ]
    }
   ],
   "source": [
    "cut -f 1,3,4 kat-staramr-out/summary.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Isolate ID\tGenotype\tPredicted Phenotype\n",
      "SRR1952908\taadA1, aadA2, blaTEM-57, cmlA1, sul3, tet(A)\tstreptomycin, ampicillin, chloramphenicol, sulfisoxazole, tetracycline\n"
     ]
    }
   ],
   "source": [
    "cut -f 1,3,4 staramr-out/summary.tsv"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
