{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SNP signature\n",
    "\n",
    "* Assumption: a set of alleles from an MLST scheme which has already been aligned with clustalo.\n",
    "\n",
    "## Get only a kmer which happens to overlap a SNP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "aspA_1\n",
      "\tAATGTTTTTGAACCAGTTGCA\n",
      "\tAATTAATTCCCGTTCCAATAG\n",
      "\tATGATAGGTGAAGATATACAA\n",
      "\tCATTAAGTTGTAATTGTCCAC\n",
      "\tCTAGAAAATTGATTTTAGAGA\n",
      "\tCTGAGGATTTGATCGAGGCGA\n",
      "\tGACTTTTAAGCAGTGGTCCAA\n",
      "\tGTAGTTCTATCATGCCAGGTA\n",
      "\tGTTCAAAAACATTAAGTTGTA\n",
      "\tTTACAGTTACATCTGCTCCAA\n",
      "aspA_2\n",
      "\tAATGTTTTTGAACCAGTTGTA\n",
      "\tAATTAATTCCTGTTCCAATAG\n",
      "\tATGATAGGTGAAGATATACAA\n",
      "\tCATTAAGTTGTAATTGTCCAC\n",
      "\tCTAGAAAATTGATTTTAGAGA\n",
      "\tCTGAGGATTTAATCGAGGCGA\n",
      "\tGACTTTTAAGTAGTGGTCCAA\n",
      "\tGTAGTTCTATCATGCCAGGTA\n",
      "\tGTTCAAAAACATTAAGTTGTA\n",
      "\tTTACAGTTACGTCTGCTCCAA\n",
      "aspA_1: \t['173_G', '278_C', '341_C', '413_T', '449_A', '44_G', '458_T', '475_C', '83_G', '8_T']\n",
      "aspA_10: \t['173_G', '278_C', '341_C', '413_T', '449_A', '44_A', '458_T', '475_C', '83_G', '8_C']\n",
      "aspA_2: \t['173_A', '278_T', '341_C', '413_C', '449_A', '44_G', '458_T', '475_T', '83_A', '8_T']\n",
      "aspA_3: \t['173_G', '278_C', '341_C', '413_T', '449_A', '44_G', '458_T', '475_C', '83_A', '8_T']\n",
      "aspA_4: \t['173_G', '278_C', '341_T', '413_T', '449_A', '44_A', '458_T', '475_C', '83_G', '8_C']\n",
      "aspA_5: \t['173_G', '278_C', '341_T', '413_T', '449_A', '44_A', '458_C', '475_C', '83_G', '8_C']\n",
      "aspA_6: \t['173_G', '278_T', '341_C', '413_T', '449_G', '44_G', '458_T', '475_T', '83_A', '8_T']\n",
      "aspA_7: \t['173_A', '278_T', '341_C', '413_T', '449_A', '44_G', '458_T', '475_T', '83_A', '8_T']\n",
      "aspA_8: \t['173_G', '278_T', '341_C', '413_C', '449_A', '44_G', '458_T', '475_T', '83_G', '8_T']\n",
      "aspA_9: \t['173_G', '278_T', '341_C', '413_C', '449_G', '44_G', '458_T', '475_T', '83_G', '8_T']\n"
     ]
    }
   ],
   "source": [
    "from Bio import AlignIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Seq import reverse_complement\n",
    "import pandas as pd\n",
    "import math\n",
    "\n",
    "align = AlignIO.read(\"aspA-10.aligned.fasta\", \"fasta\")\n",
    "\n",
    "kmer_size = 21\n",
    "\n",
    "def canonical_kmer(kmer):\n",
    "    kmer_upper = kmer.upper()\n",
    "    kmer_rc = reverse_complement(kmer_upper)\n",
    "    if kmer_upper < kmer_rc:\n",
    "        return kmer_upper\n",
    "    else:\n",
    "        return kmer_rc\n",
    "\n",
    "def check_freq(str):\n",
    "    freqcy = {}\n",
    "    min_v=1000\n",
    "    min_char = None\n",
    "    for c in set(str):\n",
    "       freqcy[c] = str.count(c)\n",
    "       if freqcy[c] < min_v:\n",
    "           min_v = freqcy[c]\n",
    "           min_char = c\n",
    "    return (freqcy, min_char)\n",
    "\n",
    "# Stores table like {'allele': set(kmers)}\n",
    "table_of_kmers = {}\n",
    "\n",
    "# Stores table like {'allele': set(position_snp)}\n",
    "snp_table = {}\n",
    "\n",
    "for i in range(align.get_alignment_length()):\n",
    "    col_align = align[:, i:(i+1)]\n",
    "    col = align[:, i]\n",
    "    (cf, min_char) = check_freq(col)\n",
    "    \n",
    "    # If a SNP (e.g., more than one different nucleotide)\n",
    "    if len(cf) > 1:\n",
    "        for ca in col_align:\n",
    "            snp_table.setdefault(ca.id, set()).add(f\"{i}_{str(ca.seq)}\")\n",
    "        left_edge = max(0, i - math.floor(kmer_size/2))\n",
    "        right_edge = left_edge + kmer_size\n",
    "        if right_edge > align.get_alignment_length():\n",
    "            right_edge = align.get_alignment_length()\n",
    "            left_edge = right_edge - kmer_size\n",
    "            \n",
    "        kmer_slice = align[:, left_edge:right_edge]\n",
    "        \n",
    "        for allele in kmer_slice:\n",
    "            kmer_string = canonical_kmer(str(allele.seq))\n",
    "            table_of_kmers.setdefault(allele.id, set()).add(kmer_string)\n",
    "\n",
    "# Only print first 2 alleles\n",
    "count = 0\n",
    "for id in table_of_kmers:\n",
    "    if count >= 2:\n",
    "        break\n",
    "    else:\n",
    "        print(f\"{id}\")\n",
    "        for kmer in sorted(table_of_kmers[id]):\n",
    "            print(f\"\\t{kmer}\")\n",
    "    count += 1\n",
    "    \n",
    "# Print SNPs\n",
    "for allele in sorted(snp_table.keys()):\n",
    "    print(f\"{allele}: \\t{sorted(snp_table[allele])}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Print kmers that uniquely differentiate one allele from the rest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>aspA_1</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_10</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_2</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_3</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_4</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_5</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_6</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_7</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_8</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_9</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         0\n",
       "aspA_1   0\n",
       "aspA_10  0\n",
       "aspA_2   0\n",
       "aspA_3   0\n",
       "aspA_4   0\n",
       "aspA_5   3\n",
       "aspA_6   0\n",
       "aspA_7   0\n",
       "aspA_8   0\n",
       "aspA_9   0"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ordered_labels = sorted(table_of_kmers.keys())\n",
    "\n",
    "data = []\n",
    "for allele1 in ordered_labels:\n",
    "    unique_kmers1 = table_of_kmers[allele1]\n",
    "    \n",
    "    for allele2 in ordered_labels:\n",
    "        if allele2 != allele1 and len(unique_kmers1) > 0:\n",
    "            unique_kmers1 = unique_kmers1 - table_of_kmers[allele2]\n",
    "            \n",
    "    data.append(len(unique_kmers1))\n",
    "    \n",
    "df = pd.DataFrame(data, index=ordered_labels)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Print SNPs that uniquely differentiate one allele from the rest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>aspA_1</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_10</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_2</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_3</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_4</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_5</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_6</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_7</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_8</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aspA_9</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         0\n",
       "aspA_1   0\n",
       "aspA_10  0\n",
       "aspA_2   0\n",
       "aspA_3   0\n",
       "aspA_4   0\n",
       "aspA_5   1\n",
       "aspA_6   0\n",
       "aspA_7   0\n",
       "aspA_8   0\n",
       "aspA_9   0"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ordered_labels = sorted(snp_table.keys())\n",
    "\n",
    "data = []\n",
    "for allele1 in ordered_labels:\n",
    "    unique_snps1 = snp_table[allele1]\n",
    "    \n",
    "    for allele2 in ordered_labels:\n",
    "        if allele2 != allele1 and len(unique_snps1) > 0:\n",
    "            unique_snps1 = unique_snps1 - snp_table[allele2]\n",
    "            \n",
    "    data.append(len(unique_snps1))\n",
    "    \n",
    "df2 = pd.DataFrame(data, index=ordered_labels)\n",
    "df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
